{
 "cells": [
  {
   "cell_type": "markdown",
   "id": "elder-workshop",
   "metadata": {},
   "source": [
    "## Goal of the notebook: calculate fitness effect of insertion mutations in DM25 and write DFE data to file\n",
    "\n",
    "Input Data:\n",
    "- mutant trajectories: number of counts at each TA site at every time point for each population. \n",
    "\n",
    "### PLEASE DOWNLOAD DATA FROM ZENODO INTO DATA FOLDER, FILES ARE TOO LARGE FOR GITHUB\n",
    "\n",
    "Analysis Approach:\n",
    "- Calculate fitness effects of disrupting genes by linear fit to log frequency vs time. Average over TA sites in the interior of the gene. (require at least x number of TA sites for which we calculate fitness)\n",
    "- If no fitness is calculated, set value as -1\n",
    "- Correct the fitness effect of each gene with an internal control: the average effect of disrupting pseudogenes\n",
    "- Do a pairwise comparison of the ancestor vs each population: if there is no fitness calculated in one of the strains, repeat fitness calculation with a less strict set of thresholds.\n",
    "- Calculate the means and variance of the DFEs for each population, also estimate error in the mean and variance.\n",
    "\n",
    "Output:\n",
    "- fitness_gene_corrected: (genes x libraries x 2 replicates) inverse-variance weighted average of fitness effect of disrupting gene at insertion sites in the interior\n",
    "- fitness_relaxed: (genes x libraries x 2 replicates) same as above except I calculate fitness effects for genes using relaxed thresholds\n",
    "- means/variance/errors in estimates of DFE for each population\n",
    "\n",
    "#### NOTE: I have commented out all the places where I write images or data to file, please uncomment and rewrite data if you'd like to mess with parameters and redo the analysis"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 1,
   "id": "dying-asset",
   "metadata": {},
   "outputs": [],
   "source": [
    "import numpy as np\n",
    "from scipy import stats\n",
    "import matplotlib.pyplot as plt\n",
    "from Bio.SeqIO.FastaIO import SimpleFastaParser\n",
    "import re\n",
    "import pandas as pd\n",
    "import seaborn as sns"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "id": "sacred-morning",
   "metadata": {},
   "outputs": [],
   "source": [
    "from matplotlib import cm\n",
    "from statsmodels.stats.multitest import multipletests\n",
    "from scipy.stats import beta, gamma,lognorm, poisson\n",
    "import matplotlib.patches as mpatches"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "id": "outer-lindsay",
   "metadata": {},
   "outputs": [],
   "source": [
    "import pathlib\n",
    "import os"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "id": "revised-diameter",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "/Users/anuraglimdi/Desktop/TnSeq_Paper/LTEE-TnSeq_Paper/Analysis/Part_3_TnSeq_analysis/Fitness_estimation\n"
     ]
    }
   ],
   "source": [
    "#current working directory\n",
    "cwd = os.getcwd()\n",
    "print(cwd)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 5,
   "id": "broad-traffic",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "/Users/anuraglimdi/Desktop/TnSeq_Paper/LTEE-TnSeq_Paper\n"
     ]
    }
   ],
   "source": [
    "#use the pathlib.Path function to get the parent directories-> goal is to navigate to directory with the metadata\n",
    "# and the breseq output data\n",
    "path = pathlib.Path(cwd)\n",
    "repo = str(path.parents[2])\n",
    "print(path.parents[2]) #this should be the base directory for the github repository: the exact path will differ for \n",
    "#each unique user"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "affected-teaching",
   "metadata": {},
   "source": [
    "### Please make sure that the output of the previous cell is the path where you have cloned the github repository"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 6,
   "id": "excellent-beverage",
   "metadata": {},
   "outputs": [],
   "source": [
    "metadata_path = repo+'/Metadata/'\n",
    "#data_path = repo+'/Data/Mutant_Trajectories/'"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 7,
   "id": "eastern-button",
   "metadata": {},
   "outputs": [],
   "source": [
    "#input directory with all the counts data\n",
    "directory = repo+'/Data/Mutant_Trajectories/'\n",
    "#names of libraries\n",
    "libraries = ['REL606', 'REL607', 'REL11330', 'REL11333', 'REL11364', 'REL11336', 'REL11339', 'REL11389', 'REL11392', 'REL11342', 'REL11345', 'REL11348', 'REL11367', 'REL11370']\n",
    "#more interpretable names for the figures in the paper\n",
    "libraries2 = ['REL606', 'REL607', 'Ara-1', 'Ara-2', 'Ara-3', 'Ara-4', 'Ara-5', 'Ara-6', 'Ara+1', 'Ara+2', 'Ara+3', 'Ara+4', 'Ara+5', 'Ara+6']"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 8,
   "id": "unlimited-atmosphere",
   "metadata": {},
   "outputs": [],
   "source": [
    "#opening the pandas file with all the metadata!\n",
    "all_data = pd.read_csv(metadata_path+\"all_metadata_REL606.txt\", sep=\"\\t\")\n",
    "names = all_data.iloc[:,0]\n",
    "gene_start = all_data.iloc[:,3]\n",
    "gene_end = all_data.iloc[:,4]\n",
    "strand = all_data.iloc[:,5]\n",
    "locations = np.transpose(np.vstack([gene_start,gene_end,strand]))\n",
    "k12_tags = all_data.iloc[:,2]\n",
    "uniprot_rel606 = all_data.iloc[:,6]\n",
    "locations_pseudogenes = np.loadtxt(metadata_path+'pseudogenes_locations_REL606.txt')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 9,
   "id": "informational-globe",
   "metadata": {},
   "outputs": [],
   "source": [
    "#list of genes to be excluded from analysis as they lie within large deletions\n",
    "exclude_genes = np.loadtxt(repo+\"/Analysis/Part_2_WGS_analysis/output/Deleted_genes_REL606_k12annotated.txt\")\n",
    "\n",
    "exclude_pseudogenes = np.loadtxt(repo+\"/Analysis/Part_2_WGS_analysis/output/Deleted_pseudogenes_REL606_k12annotated.txt\")\n",
    "\n",
    "n_pseudo = exclude_pseudogenes.shape[1]\n",
    "\n",
    "#fractions of the gene at the 5' and 3' ends to be excluded from analysis because they insertions there may not actually\n",
    "#be disruptive to protein function\n",
    "frac5p = 0.1\n",
    "frac3p = 0.25"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 10,
   "id": "coastal-poetry",
   "metadata": {},
   "outputs": [],
   "source": [
    "savepath = repo+'/Analysis/Part_3_TnSeq_analysis/Processed_data_for_plotting/'#path where I'm saving data such as fitness, fraction, etc. "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 11,
   "id": "optional-elephant",
   "metadata": {},
   "outputs": [],
   "source": [
    "with open(metadata_path+\"rel606_reference.fasta\") as in_handle:\n",
    "    for title, seq in SimpleFastaParser(in_handle):\n",
    "        ta_sites = [m.start(0) for m in re.finditer('TA', seq)]\n",
    "ta_sites = np.array(ta_sites)\n",
    "\n",
    "#counting how many TA sites are present in each gene\n",
    "ta_gene = np.zeros(len(names))\n",
    "for i in range(0,len(names)):\n",
    "    start = locations[i, 0]\n",
    "    end = locations[i, 1]\n",
    "    length = end - start\n",
    "    #if the gene is on the forward strand\n",
    "    if locations[i,2]==1:\n",
    "        #counting sites only in the middle 80% of the gene, excluding 10% at each end\n",
    "        ta_gene[i] = np.sum((ta_sites > start+length*frac5p)&(ta_sites < end - length*frac3p))\n",
    "    elif locations[i,2]==-1:\n",
    "        ta_gene[i] = np.sum((ta_sites < start+length*frac5p)&(ta_sites > end - length*frac3p))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 12,
   "id": "velvet-official",
   "metadata": {},
   "outputs": [],
   "source": [
    "#all the information from the fitness assay condensed into a couple of matrices\n",
    "counts_all_green = np.zeros([len(libraries), 5, len(ta_sites)])\n",
    "counts_all_red = np.zeros([len(libraries), 5, len(ta_sites)])\n",
    "\n",
    "for k in range(len(libraries)):\n",
    "    #loading the new file where I have the counts for each TA site for all time points\n",
    "    gname = directory+'green_'+libraries[k]+'_merged_all_TAsites.txt'\n",
    "    greendata = np.loadtxt(gname)\n",
    "    rname = directory+'red_'+libraries[k]+'_merged_all_TAsites.txt'\n",
    "    reddata = np.loadtxt(rname)\n",
    "    #now for extracting the UMI corrected coutns\n",
    "    counts_all_green[k,:,:] = greendata[2:11:2,:]\n",
    "    counts_all_red[k,:,:] = reddata[2:11:2,:]"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "joint-immigration",
   "metadata": {},
   "source": [
    "### Part 1: Calculating fitness effects in all populations"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 13,
   "id": "perceived-small",
   "metadata": {},
   "outputs": [],
   "source": [
    "#normalized counts data for each sample:\n",
    "greensum = np.sum(counts_all_green,axis=2)\n",
    "redsum = np.sum(counts_all_red, axis=2)\n",
    "#fraction of sites above threshold\n",
    "fraction_t0 = np.zeros([len(names),len(libraries)])-1"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 17,
   "id": "reserved-johnston",
   "metadata": {},
   "outputs": [],
   "source": [
    "#timepoint where we look at the fraction of sites that are present in the dataset\n",
    "t_comp = 0\n",
    "site_thresh = 5\n",
    "ta_min = 5\n",
    "#not using this data directly, but will be needed later when we are inferring gene essentiality"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 18,
   "id": "indie-weekly",
   "metadata": {},
   "outputs": [],
   "source": [
    "#first step, counting the number of TA sites above the threshold at the comparison timepoint for all the populations\n",
    "for k in range(len(libraries)):\n",
    "    for i in range(len(names)):\n",
    "        #if the gene is present in this genetic background, and there are at least 5 TA sites in the gene\n",
    "        if exclude_genes[k,i] == 0 and ta_gene[i] >=ta_min:\n",
    "            start = locations[i, 0]\n",
    "            end = locations[i, 1]\n",
    "            length = end - start\n",
    "            #if the gene is on the forward strand\n",
    "            if locations[i,2]==1:\n",
    "                search_area = (ta_sites > start+length*frac5p)&(ta_sites < end - length*frac3p)\n",
    "            #if the gene is on the reverse strand\n",
    "            elif locations[i,2]==-1:\n",
    "                search_area = (ta_sites < start+length*frac5p)&(ta_sites > end - length*frac3p)\n",
    "            green = np.sum(counts_all_green[k,t_comp,search_area]/greensum[k,t_comp]*10**7 > site_thresh)\n",
    "            red = np.sum(counts_all_red[k,t_comp,search_area]/redsum[k,t_comp]*10**7 > site_thresh)\n",
    "            fraction_t0[i,k] = (green+red)/np.abs(ta_gene[i])/2"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "noted-leader",
   "metadata": {},
   "source": [
    "### Saving data: fraction of sites above coverage threshold, used later for gene essentiality estimation and more"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 19,
   "id": "electronic-message",
   "metadata": {},
   "outputs": [],
   "source": [
    "fname = savepath+f'fraction_t0_site_thresh_{site_thresh}_ta_min_{ta_min}.txt'\n",
    "# np.savetxt(savepath+f'fraction_t0_site_thresh_{site_thresh}_ta_min_{ta_min}.txt', fraction_t0)"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "first-cover",
   "metadata": {},
   "source": [
    "### Important Functions for fitness estimation:\n",
    "- search_gene: returns the coordinates of TA sites that lie within the interior of any gene\n",
    "- inverse_var_weight: returns the inverse variance weights a given N_0, N_1\n",
    "- fitness_estimate: calculates fitness estimate and measurement error for a gene"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 17,
   "id": "joined-hybrid",
   "metadata": {},
   "outputs": [],
   "source": [
    "def search_gene(locations,ta_sites,i):\n",
    "    start = locations[i, 0]\n",
    "    end = locations[i, 1]\n",
    "    length = end - start\n",
    "    #if the gene is on the forward strand\n",
    "    if locations[i,2]==1:\n",
    "        search_area = (ta_sites > start+length*frac5p)&(ta_sites < end - length*frac3p)\n",
    "    #if the gene is on the reverse strand\n",
    "    elif locations[i,2]==-1:\n",
    "        search_area = (ta_sites < start+length*frac5p)&(ta_sites > end - length*frac3p)\n",
    "    return search_area\n",
    "\n",
    "#following the weights definition from \n",
    "def inverse_var_weight(n0,n1):\n",
    "    \n",
    "    return (((1+n0)**-1 + (1+n1)**-1)/np.log(2)**2)**-1"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "c6ad9362",
   "metadata": {},
   "source": [
    "#### Quick summary of the function:\n",
    "\n",
    "- Restrict analysis to genes that are not lost, and have at least 5 TA sites in the interior\n",
    "- For every interior TA site, calculate the fitness effect by fitting a line to log(frequency) vs time\n",
    "- If trajectory dies out (i.e. counts = 0), then set that timepoint to 1 (a pseudocount), and do a linear fit of log(frequency) vs time, upto that specific time-point\n",
    "- If there are at least 2 surviving trajectories in each replicate, then use an inverse variance weighting approach to get the average effect of disrupting the gene.\n",
    "- Estimate uncertainty in the fitness measurement using a few different metrics (discussed below): for the paper, we use an inverse variance weighted SEM\n",
    "\n",
    "\n",
    "#### Note about convention: if no fitness is calculated for a gene, its value is set to -1: this is simply so it allows me to keep track of where thresholds were met or not"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 21,
   "id": "polished-amount",
   "metadata": {},
   "outputs": [],
   "source": [
    "#calculating fitnesses\n",
    "def fitness_estimate(counts_red, counts_green, locs, genes_lost, initial_depth, min_sites, min_frac, max_reads, t_start, t_end):\n",
    "    \"\"\"\n",
    "    data:\n",
    "    - counts_red, counts_green: trajectories for both replicates\n",
    "    - exclude: filtering out genes that are lost/have too few sites from analysis as they're indicated as -1\n",
    "    params:\n",
    "    - locs: start and end points of genes\n",
    "    - initial_depth: exclude sites below a certain threshold in fitness estimation\n",
    "    - min_sites: minimum number of sites required for estimating fitness for each replicate\n",
    "    - min_frac: fraction of TA sites that need to considered in estimating fitness\n",
    "    - site_thresh: minimum number of reads for a trajectory to be included (strictly greater than)\n",
    "    - max_reads: maximum reads that can be used for inverse variance estimate\n",
    "    \n",
    "    \"\"\"\n",
    "    green = np.sum(counts_green,axis=1)\n",
    "    red = np.sum(counts_red, axis=1)\n",
    "    #fitness effect (best estimate)\n",
    "    fitness_inverse_var = np.zeros([len(genes_lost),2])-1\n",
    "    #three different metrics of measurement error\n",
    "    stderr_fitness = np.zeros([len(genes_lost)])-1\n",
    "    stderr_fitness_inverse_var = np.zeros([len(genes_lost)])-1\n",
    "    stderr_fitness_hybrid = np.zeros([len(genes_lost)])-1\n",
    "    #defining time interval based on gene essentiality status:\n",
    "    time_gens = np.linspace(t_start,t_end,t_end-t_start+1)*6.64\n",
    "    \n",
    "    #iterating over all genes that are not lost over evolution\n",
    "    for gene in np.where(genes_lost==0)[0] :\n",
    "        #TA sites within this gene\n",
    "        search_area = search_gene(locations=locs, ta_sites=ta_sites, i=gene)\n",
    "        #weights\n",
    "        weights_green = []\n",
    "        weights_red = []\n",
    "        #fitnesses\n",
    "        s_green = []\n",
    "        s_red = []\n",
    "        \n",
    "        ta = np.sum(search_area==True)\n",
    "        #now, iterating over sites with at least tm1_depth reads at t-1:\n",
    "        for j in np.where((search_area==True))[0]:\n",
    "            if (counts_green[t_start,j]/green[t_start]*10**7>initial_depth) & (counts_red[t_start,j]/red[t_start]*10**7>initial_depth):\n",
    "                #extracting the trajectories\n",
    "                traj_green = counts_green[t_start:t_end+1,j]/green[t_start:t_end+1]*10**7\n",
    "                traj_red = counts_red[t_start:t_end+1,j]/red[t_start:t_end+1]*10**7\n",
    "                n_0 = traj_green[0]\n",
    "                n_after_g = traj_green[1]\n",
    "                n_after_r = traj_red[1]\n",
    "                #this part is different:\n",
    "                #i'll find the first occurrence of zero if at all, and then fit a trajectory upto that point\n",
    "                #green replicate first\n",
    "                if 0 not in traj_green:\n",
    "                    sg = np.polyfit(time_gens, np.log(traj_green), 1)[0]\n",
    "                    wg = min(inverse_var_weight(n_0, n_after_g), inverse_var_weight(max_reads, max_reads))\n",
    "                    #sotring the data\n",
    "                    weights_green.append(wg)\n",
    "                    s_green.append(sg)\n",
    "                else:\n",
    "                    #first zero in the gren dataset:\n",
    "                    ind_g = int(np.where(traj_green==0)[0][0])\n",
    "                    counts_g = traj_green[:ind_g+1].copy()\n",
    "                    #add a pseudocount to the first occurrence of zero\n",
    "                    counts_g[-1] = 1\n",
    "                    sg = np.polyfit(time_gens[:ind_g+1], np.log(counts_g), 1)[0]\n",
    "                    wg = min(inverse_var_weight(n_0, n_after_g), inverse_var_weight(max_reads, max_reads))\n",
    "                    #storing the data\n",
    "                    weights_green.append(wg)\n",
    "                    s_green.append(sg)\n",
    "                \n",
    "                #now the red replicate\n",
    "                if 0 not in traj_red:\n",
    "                    sg = np.polyfit(time_gens, np.log(traj_red), 1)[0]\n",
    "                    wg = min(inverse_var_weight(n_0, n_after_r), inverse_var_weight(max_reads, max_reads))\n",
    "                    #sotring the data\n",
    "                    weights_red.append(wg)\n",
    "                    s_red.append(sg)\n",
    "                else:\n",
    "                    #first zero in the red dataset:\n",
    "                    ind_g = int(np.where(traj_red==0)[0][0])\n",
    "                    counts_g = traj_red[:ind_g+1].copy()\n",
    "                    #add a pseudocount to the first occurrence of zero\n",
    "                    counts_g[-1] = 1\n",
    "                    sg = np.polyfit(time_gens[:ind_g+1], np.log(counts_g), 1)[0]\n",
    "                    wg = min(inverse_var_weight(n_0, n_after_r), inverse_var_weight(max_reads, max_reads))\n",
    "                    #storing the data\n",
    "                    weights_red.append(wg)\n",
    "                    s_red.append(sg)\n",
    "                \n",
    "        #converting to numpy arrays\n",
    "        s_green = np.array(s_green)\n",
    "        s_red = np.array(s_red)\n",
    "        weights_green = np.array(weights_green)\n",
    "        weights_red = np.array(weights_red)\n",
    "        \n",
    "        #pooling over all sites, require minimum number of sites, and fraction of sites represented\n",
    "        if min(len(s_green),len(s_red))>=min_sites and min(len(s_green),len(s_red))/ta>=min_frac:\n",
    "\n",
    "            #weighing based on inverse variacne:\n",
    "            fitness_inverse_var[gene,0] = np.average(s_green, weights=weights_green)\n",
    "            fitness_inverse_var[gene,1] = np.average(s_red, weights=weights_red)\n",
    "            \n",
    "            #standard error in fitness estimates:\n",
    "            stderr_fitness[gene] = np.std(np.hstack([s_green,s_red]))/np.sqrt(len(s_green)+len(s_red)-1)\n",
    "            #-1 is for the unbiased estimator\n",
    "            \n",
    "            #inverse variance weighted standard error measurement\n",
    "            stacked_s = np.hstack([s_green,s_red])\n",
    "            stacked_w = np.hstack([weights_green,weights_red])\n",
    "            \n",
    "            average = np.average(stacked_s, weights=stacked_w)\n",
    "            sem_inv = np.sqrt(np.average((stacked_s-average)**2, weights=stacked_w)/(len(stacked_s)-1))\n",
    "            sem_hybrid = np.sqrt(np.average((stacked_s-average)**2)/(len(stacked_s)-1))\n",
    "            \n",
    "            #saving the inverse variance wieghted measurements.\n",
    "            stderr_fitness_inverse_var[gene] = sem_inv\n",
    "            #saving the hyrid measurement errors\n",
    "            stderr_fitness_hybrid[gene] = sem_hybrid\n",
    "            \n",
    "    return fitness_inverse_var, stderr_fitness, stderr_fitness_inverse_var, stderr_fitness_hybrid"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 22,
   "id": "precise-bruce",
   "metadata": {},
   "outputs": [],
   "source": [
    "#parameters for filtering out noisy data and selecting which time points to use in analysis\n",
    "min_sites = 2\n",
    "min_frac = 0.2\n",
    "#timepoints to compare:\n",
    "# change this depending on the analysis you want to do!\n",
    "t0 = 0\n",
    "t1 = 4\n",
    "time_range = np.linspace(t0,t1,t1-t0+1)*6.64\n",
    "tm1_depth = 5"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "a8c864c7",
   "metadata": {},
   "source": [
    "#### Please note that I experimented with three different metrics of error in the fitness estimates.\n",
    "\n",
    "Note that the mean fitness of a gene is calculated as an inverse variance weighted average ($\\mu_{inv} = \\sum(w_{i}s_{i})$)\n",
    "\n",
    "- option 1: the standard error (using the builtin numpy function): this is the least precise, as it'll automatically calculate the mean without weighting in the sem calculation. $\\sum(x_{i} - \\mu_{uncorrected})^{2}$\n",
    "- option 2: the inverse variance weighted standard error: in the sem calculation, weigh replicates by the same amount as they were in calculating the mean (this is the one that is the most appropriate in our analysis). $\\sum(s_{i} - \\mu_{uncorrected})^{2}w_{i}$, where $w_{i}$ are the weights assigned to the different TA sites within the gene\n",
    "- option 3: hybrid standard error: no weights, but instead of $\\sum(s_{i} - \\mu_{uncorrected})^{2}$, we use $\\sum(s_{i} - \\mu_{inv})^{2}$\n",
    "\n",
    "We went with option 2 as it takes into account the fact that we don't believe all our fitness measurements within a gene equally, and places less value on the inherently noisier ones in estimating uncertainty in the true fitness effect. Also note that this choice does not impact our results in a meaningful way."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 23,
   "id": "growing-scratch",
   "metadata": {},
   "outputs": [],
   "source": [
    "s_inverse_var = np.zeros([len(names),len(libraries),2])-1\n",
    "s_inverse_pseudo = np.zeros([n_pseudo,len(libraries),2])-1\n",
    "\n",
    "#the three different error metrics\n",
    "standard_error = np.zeros([len(names),len(libraries)])-1\n",
    "standard_error_inv = np.zeros([len(names),len(libraries)])-1\n",
    "standard_error_hybrid = np.zeros([len(names),len(libraries)])-1\n",
    "\n",
    "#the three different error metrics\n",
    "standard_error_pseudo = np.zeros([n_pseudo,len(libraries)])-1\n",
    "standard_error_pseudo_inv = np.zeros([len(names),len(libraries)])-1\n",
    "standard_error_pseudo_hybrid = np.zeros([len(names),len(libraries)])-1\n",
    "\n",
    "for k in range(0,len(libraries)):\n",
    "    #now I'm going to call the function which actually runs the code for estimating fitness\n",
    "    fitness_inverse_var, pop_error, pop_error_inv, pop_error_hybrid = fitness_estimate(counts_all_green[k,:,:], counts_all_red[k,:,:], \n",
    "                                                           genes_lost=exclude_genes[k,:], locs=locations, initial_depth=tm1_depth,\n",
    "                                                           min_frac=min_frac,max_reads=100,\n",
    "                                                           min_sites=min_sites, t_start=t0, t_end=t1)\n",
    "    #runnning the function for the pseudogenes\n",
    "    fitness_inverse_var_pseudogenes, pop_error_pseudo, pop_error_pseudo_inv, pop_error_pseudo_hybrid = fitness_estimate(counts_all_green[k,:,:], counts_all_red[k,:,:], \n",
    "                                                            genes_lost=exclude_pseudogenes[k,:],\n",
    "                                                           locs=locations_pseudogenes,initial_depth=tm1_depth,\n",
    "                                                           min_frac=min_frac,max_reads=100,\n",
    "                                                           min_sites=min_sites, t_start=t0, t_end=t1)\n",
    "        \n",
    "    #genes in the analysis:\n",
    "    genes_analyzed = np.where(fitness_inverse_var[:,0]>-1)[0]\n",
    "    \n",
    "    #pseudogenes in the analysis\n",
    "    pseudogenes_analyzed = np.where(fitness_inverse_var_pseudogenes[:,0]>-1)[0]\n",
    "    \n",
    "    #calculating the mean correction factor for each populaiton, not excluding outliers:\n",
    "    mean_correction_inv_var = np.mean(fitness_inverse_var_pseudogenes[pseudogenes_analyzed,:], axis=0)\n",
    "\n",
    "    #saving the fitness estimates for the genes\n",
    "    s_inverse_var[genes_analyzed, k, 0] = fitness_inverse_var[genes_analyzed, 0]\n",
    "    s_inverse_var[genes_analyzed, k, 1] = fitness_inverse_var[genes_analyzed, 1]\n",
    "    \n",
    "#     s_inverse_var[np.where(np.abs(pvals_fdr[:,k]<ess_thresh)), k, :] = -2\n",
    "    \n",
    "    #saving the pseudogenes fitness estimates:\n",
    "    s_inverse_pseudo[:,k,:] = fitness_inverse_var_pseudogenes\n",
    "    #unweighted error estimates\n",
    "    standard_error[genes_analyzed, k] = pop_error[genes_analyzed]\n",
    "    standard_error_pseudo[pseudogenes_analyzed, k] = pop_error_pseudo[pseudogenes_analyzed]\n",
    "    #inverse variance weighted error estimates\n",
    "    standard_error_inv[genes_analyzed, k] = pop_error_inv[genes_analyzed]\n",
    "    standard_error_pseudo_inv[pseudogenes_analyzed, k] = pop_error_pseudo_inv[pseudogenes_analyzed]\n",
    "    #hybrid error estimates\n",
    "    standard_error_hybrid[genes_analyzed, k] = pop_error_hybrid[genes_analyzed]\n",
    "    standard_error_pseudo_hybrid[pseudogenes_analyzed, k] = pop_error_pseudo_hybrid[pseudogenes_analyzed]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 24,
   "id": "large-portuguese",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "True\n"
     ]
    }
   ],
   "source": [
    "s_inverse_var_copy = np.copy(s_inverse_var)\n",
    "print(np.min(s_inverse_var==s_inverse_var_copy))"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "26fb5037",
   "metadata": {},
   "source": [
    "#### Correcting the fitness estimates by subtracting the mean effect of disrupting pseudogenes that have fitness effects near neutrality.\n",
    "\n",
    "- Note that some pseudogenes can have highly deleterious effects, potentially due to polar effects.\n",
    "- For Ara-2, all the fitness estimates are biased due to a few mutants taking over and the bulk of the uncorrected DFE is concentrated around -0.2, so there are no pseudogenes with an uncorrected fitness of +/- 0.05"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 25,
   "id": "growing-massage",
   "metadata": {},
   "outputs": [],
   "source": [
    "run_once = 0"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 26,
   "id": "periodic-bruce",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "[-0.0087086  -0.00636075] REL606\n",
      "[-0.01023422 -0.0100438 ] REL607\n",
      "[-0.00725029 -0.00512881] REL11330\n",
      "[-0.00495299 -0.00657039] REL11364\n",
      "[-0.00751923 -0.00780591] REL11336\n",
      "[-0.00699322 -0.00589254] REL11339\n",
      "[-0.01184721 -0.01206626] REL11389\n",
      "[-0.00337866 -0.00263297] REL11392\n",
      "[-0.00553487 -0.00680097] REL11342\n",
      "[-0.0077014  -0.00696454] REL11345\n",
      "[-0.01199336 -0.00900709] REL11348\n",
      "[-0.00712573 -0.00686453] REL11367\n",
      "[-0.01362303 -0.02072889] REL11370\n"
     ]
    }
   ],
   "source": [
    "while run_once==0:\n",
    "    for k in range(0, len(libraries)):\n",
    "        #handling all populations except ara-2\n",
    "        if k not in [3]:\n",
    "            genes_analyzed = np.where(s_inverse_var[:,k,0]>-1)[0]\n",
    "            pseudogenes_analyzed = np.where((s_inverse_pseudo[:,k,0]>-0.05)&(s_inverse_pseudo[:,k,1]>-0.05))[0]\n",
    "            corrections = np.mean(s_inverse_pseudo[pseudogenes_analyzed,k,:],axis=0)\n",
    "            print(corrections, libraries[k])\n",
    "            #correcting the estimates by replicate\n",
    "            s_inverse_var[genes_analyzed, k, 0] = s_inverse_var[genes_analyzed, k, 0] - corrections[0]\n",
    "            s_inverse_var[genes_analyzed, k, 1] = s_inverse_var[genes_analyzed, k, 1] - corrections[1]\n",
    "        #because a-2 has some mutations fix, pseudogenes are around -19%\n",
    "        #so I;m requiring that the fitnesses > -0.05\n",
    "        else:\n",
    "            genes_analyzed = np.where(s_inverse_var[:,k,0]>-1)[0]\n",
    "            #pseudogenes in the analysis\n",
    "            pseudogenes_analyzed = np.where(s_inverse_pseudo[:,k,0]>-1)[0]\n",
    "            corrections = np.mean(s_inverse_pseudo[pseudogenes_analyzed,k,:],axis=0)\n",
    "            #correcting the estimates by replicate\n",
    "            s_inverse_var[genes_analyzed, k, 0] = s_inverse_var[genes_analyzed, k, 0] - corrections[0]\n",
    "            s_inverse_var[genes_analyzed, k, 1] = s_inverse_var[genes_analyzed, k, 1] - corrections[1]\n",
    "    run_once=1"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "sorted-police",
   "metadata": {},
   "source": [
    "### Saving data: This is the fitness estimates for all genes/pseudogenes, and the errors in the fitness estimates (which are computed slightly differently)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 27,
   "id": "organized-population",
   "metadata": {},
   "outputs": [],
   "source": [
    "fname1 = savepath+f'fitness_corrected_genes'\n",
    "fname2 = savepath+f'errors_genes'\n",
    "fname3 = savepath+f'fitness_pseudogenes'\n",
    "fname4 = savepath+f'errors_pseudogenes'\n",
    "\n",
    "fname5 = savepath+f'errors_genes_inv'\n",
    "fname6 = savepath+f'errors_pseudogenes_inv'\n",
    "\n",
    "fname7 = savepath+f'errors_genes_hybrid'\n",
    "fname8 = savepath+f'errors_pseudogenes_hybrid'\n",
    "\n",
    "\n",
    "#saving the files now:\n",
    "# np.save(fname1, s_inverse_var)\n",
    "# np.save(fname2, standard_error)\n",
    "# np.save(fname3, s_inverse_pseudo)\n",
    "# np.save(fname4, standard_error_pseudo)\n",
    "# np.save(fname5, standard_error_inv)\n",
    "# np.save(fname6, standard_error_pseudo_inv)\n",
    "# np.save(fname7, standard_error_hybrid)\n",
    "# np.save(fname8, standard_error_pseudo_hybrid)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 235,
   "id": "orange-level",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "array([[-0.32309008, -0.26881478],\n",
       "       [-0.24791682, -0.34827508],\n",
       "       [-0.13660168, -0.02533479],\n",
       "       [-0.03299209, -0.01883153],\n",
       "       [-0.18254041, -0.16316946],\n",
       "       [-0.07390178, -0.08912543],\n",
       "       [-0.10089448, -0.07464315],\n",
       "       [-0.15432662, -0.17637027],\n",
       "       [-0.06192539, -0.06884853],\n",
       "       [-0.08027262, -0.09543254],\n",
       "       [-0.08000855, -0.04871196],\n",
       "       [-0.2952767 , -0.16504088],\n",
       "       [-0.05312648, -0.0402353 ],\n",
       "       [-0.18650211, -0.17968245]])"
      ]
     },
     "execution_count": 235,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "#just sanity checking some data\n",
    "s_inverse_var[list(names).index('sspA')]"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "super-voluntary",
   "metadata": {},
   "source": [
    "### Data for Figure 2: DFE does not change despite frequent epistatic interactions between genetic background and insertion mutations \n",
    "\n",
    "Save the data in a dataframe format so it is easy to use seaborn to make a ridgeplot"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 29,
   "id": "regional-browser",
   "metadata": {},
   "outputs": [],
   "source": [
    "#creating the consolidated pandas dataframe for all the data and the metadata\n",
    "dfe_data_all_inverse_var= pd.DataFrame()\n",
    "#mutator status of each population\n",
    "mutator = ['non-mutator', 'non-mutator', 'mutator', 'mutator', 'mutator', 'mutator', 'non-mutator', 'non-mutator'\n",
    "           , 'non-mutator', 'non-mutator', 'mutator', 'non-mutator', 'non-mutator', 'mutator']\n",
    "#ara phenotype\n",
    "ara_minus = [0,2,3,4,5,6,7]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 30,
   "id": "conservative-decision",
   "metadata": {
    "scrolled": true
   },
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "/var/folders/f8/92wybs9d7wqfthlqj1drht_c0000gn/T/ipykernel_20642/3217383995.py:48: FutureWarning: The frame.append method is deprecated and will be removed from pandas in a future version. Use pandas.concat instead.\n",
      "  dfe_data_all_inverse_var = dfe_data_all_inverse_var.append(next1, ignore_index=True)\n",
      "/var/folders/f8/92wybs9d7wqfthlqj1drht_c0000gn/T/ipykernel_20642/3217383995.py:49: FutureWarning: The frame.append method is deprecated and will be removed from pandas in a future version. Use pandas.concat instead.\n",
      "  dfe_data_all_inverse_var = dfe_data_all_inverse_var.append(next2, ignore_index=True)\n",
      "/var/folders/f8/92wybs9d7wqfthlqj1drht_c0000gn/T/ipykernel_20642/3217383995.py:48: FutureWarning: The frame.append method is deprecated and will be removed from pandas in a future version. Use pandas.concat instead.\n",
      "  dfe_data_all_inverse_var = dfe_data_all_inverse_var.append(next1, ignore_index=True)\n",
      "/var/folders/f8/92wybs9d7wqfthlqj1drht_c0000gn/T/ipykernel_20642/3217383995.py:49: FutureWarning: The frame.append method is deprecated and will be removed from pandas in a future version. Use pandas.concat instead.\n",
      "  dfe_data_all_inverse_var = dfe_data_all_inverse_var.append(next2, ignore_index=True)\n",
      "/var/folders/f8/92wybs9d7wqfthlqj1drht_c0000gn/T/ipykernel_20642/3217383995.py:48: FutureWarning: The frame.append method is deprecated and will be removed from pandas in a future version. Use pandas.concat instead.\n",
      "  dfe_data_all_inverse_var = dfe_data_all_inverse_var.append(next1, ignore_index=True)\n",
      "/var/folders/f8/92wybs9d7wqfthlqj1drht_c0000gn/T/ipykernel_20642/3217383995.py:49: FutureWarning: The frame.append method is deprecated and will be removed from pandas in a future version. Use pandas.concat instead.\n",
      "  dfe_data_all_inverse_var = dfe_data_all_inverse_var.append(next2, ignore_index=True)\n",
      "/var/folders/f8/92wybs9d7wqfthlqj1drht_c0000gn/T/ipykernel_20642/3217383995.py:48: FutureWarning: The frame.append method is deprecated and will be removed from pandas in a future version. Use pandas.concat instead.\n",
      "  dfe_data_all_inverse_var = dfe_data_all_inverse_var.append(next1, ignore_index=True)\n",
      "/var/folders/f8/92wybs9d7wqfthlqj1drht_c0000gn/T/ipykernel_20642/3217383995.py:49: FutureWarning: The frame.append method is deprecated and will be removed from pandas in a future version. Use pandas.concat instead.\n",
      "  dfe_data_all_inverse_var = dfe_data_all_inverse_var.append(next2, ignore_index=True)\n",
      "/var/folders/f8/92wybs9d7wqfthlqj1drht_c0000gn/T/ipykernel_20642/3217383995.py:48: FutureWarning: The frame.append method is deprecated and will be removed from pandas in a future version. Use pandas.concat instead.\n",
      "  dfe_data_all_inverse_var = dfe_data_all_inverse_var.append(next1, ignore_index=True)\n",
      "/var/folders/f8/92wybs9d7wqfthlqj1drht_c0000gn/T/ipykernel_20642/3217383995.py:49: FutureWarning: The frame.append method is deprecated and will be removed from pandas in a future version. Use pandas.concat instead.\n",
      "  dfe_data_all_inverse_var = dfe_data_all_inverse_var.append(next2, ignore_index=True)\n",
      "/var/folders/f8/92wybs9d7wqfthlqj1drht_c0000gn/T/ipykernel_20642/3217383995.py:48: FutureWarning: The frame.append method is deprecated and will be removed from pandas in a future version. Use pandas.concat instead.\n",
      "  dfe_data_all_inverse_var = dfe_data_all_inverse_var.append(next1, ignore_index=True)\n",
      "/var/folders/f8/92wybs9d7wqfthlqj1drht_c0000gn/T/ipykernel_20642/3217383995.py:49: FutureWarning: The frame.append method is deprecated and will be removed from pandas in a future version. Use pandas.concat instead.\n",
      "  dfe_data_all_inverse_var = dfe_data_all_inverse_var.append(next2, ignore_index=True)\n",
      "/var/folders/f8/92wybs9d7wqfthlqj1drht_c0000gn/T/ipykernel_20642/3217383995.py:48: FutureWarning: The frame.append method is deprecated and will be removed from pandas in a future version. Use pandas.concat instead.\n",
      "  dfe_data_all_inverse_var = dfe_data_all_inverse_var.append(next1, ignore_index=True)\n",
      "/var/folders/f8/92wybs9d7wqfthlqj1drht_c0000gn/T/ipykernel_20642/3217383995.py:49: FutureWarning: The frame.append method is deprecated and will be removed from pandas in a future version. Use pandas.concat instead.\n",
      "  dfe_data_all_inverse_var = dfe_data_all_inverse_var.append(next2, ignore_index=True)\n",
      "/var/folders/f8/92wybs9d7wqfthlqj1drht_c0000gn/T/ipykernel_20642/3217383995.py:48: FutureWarning: The frame.append method is deprecated and will be removed from pandas in a future version. Use pandas.concat instead.\n",
      "  dfe_data_all_inverse_var = dfe_data_all_inverse_var.append(next1, ignore_index=True)\n",
      "/var/folders/f8/92wybs9d7wqfthlqj1drht_c0000gn/T/ipykernel_20642/3217383995.py:49: FutureWarning: The frame.append method is deprecated and will be removed from pandas in a future version. Use pandas.concat instead.\n",
      "  dfe_data_all_inverse_var = dfe_data_all_inverse_var.append(next2, ignore_index=True)\n",
      "/var/folders/f8/92wybs9d7wqfthlqj1drht_c0000gn/T/ipykernel_20642/3217383995.py:48: FutureWarning: The frame.append method is deprecated and will be removed from pandas in a future version. Use pandas.concat instead.\n",
      "  dfe_data_all_inverse_var = dfe_data_all_inverse_var.append(next1, ignore_index=True)\n",
      "/var/folders/f8/92wybs9d7wqfthlqj1drht_c0000gn/T/ipykernel_20642/3217383995.py:49: FutureWarning: The frame.append method is deprecated and will be removed from pandas in a future version. Use pandas.concat instead.\n",
      "  dfe_data_all_inverse_var = dfe_data_all_inverse_var.append(next2, ignore_index=True)\n",
      "/var/folders/f8/92wybs9d7wqfthlqj1drht_c0000gn/T/ipykernel_20642/3217383995.py:48: FutureWarning: The frame.append method is deprecated and will be removed from pandas in a future version. Use pandas.concat instead.\n",
      "  dfe_data_all_inverse_var = dfe_data_all_inverse_var.append(next1, ignore_index=True)\n",
      "/var/folders/f8/92wybs9d7wqfthlqj1drht_c0000gn/T/ipykernel_20642/3217383995.py:49: FutureWarning: The frame.append method is deprecated and will be removed from pandas in a future version. Use pandas.concat instead.\n",
      "  dfe_data_all_inverse_var = dfe_data_all_inverse_var.append(next2, ignore_index=True)\n",
      "/var/folders/f8/92wybs9d7wqfthlqj1drht_c0000gn/T/ipykernel_20642/3217383995.py:48: FutureWarning: The frame.append method is deprecated and will be removed from pandas in a future version. Use pandas.concat instead.\n",
      "  dfe_data_all_inverse_var = dfe_data_all_inverse_var.append(next1, ignore_index=True)\n",
      "/var/folders/f8/92wybs9d7wqfthlqj1drht_c0000gn/T/ipykernel_20642/3217383995.py:49: FutureWarning: The frame.append method is deprecated and will be removed from pandas in a future version. Use pandas.concat instead.\n",
      "  dfe_data_all_inverse_var = dfe_data_all_inverse_var.append(next2, ignore_index=True)\n",
      "/var/folders/f8/92wybs9d7wqfthlqj1drht_c0000gn/T/ipykernel_20642/3217383995.py:48: FutureWarning: The frame.append method is deprecated and will be removed from pandas in a future version. Use pandas.concat instead.\n",
      "  dfe_data_all_inverse_var = dfe_data_all_inverse_var.append(next1, ignore_index=True)\n",
      "/var/folders/f8/92wybs9d7wqfthlqj1drht_c0000gn/T/ipykernel_20642/3217383995.py:49: FutureWarning: The frame.append method is deprecated and will be removed from pandas in a future version. Use pandas.concat instead.\n",
      "  dfe_data_all_inverse_var = dfe_data_all_inverse_var.append(next2, ignore_index=True)\n",
      "/var/folders/f8/92wybs9d7wqfthlqj1drht_c0000gn/T/ipykernel_20642/3217383995.py:48: FutureWarning: The frame.append method is deprecated and will be removed from pandas in a future version. Use pandas.concat instead.\n",
      "  dfe_data_all_inverse_var = dfe_data_all_inverse_var.append(next1, ignore_index=True)\n",
      "/var/folders/f8/92wybs9d7wqfthlqj1drht_c0000gn/T/ipykernel_20642/3217383995.py:49: FutureWarning: The frame.append method is deprecated and will be removed from pandas in a future version. Use pandas.concat instead.\n",
      "  dfe_data_all_inverse_var = dfe_data_all_inverse_var.append(next2, ignore_index=True)\n",
      "/var/folders/f8/92wybs9d7wqfthlqj1drht_c0000gn/T/ipykernel_20642/3217383995.py:48: FutureWarning: The frame.append method is deprecated and will be removed from pandas in a future version. Use pandas.concat instead.\n",
      "  dfe_data_all_inverse_var = dfe_data_all_inverse_var.append(next1, ignore_index=True)\n",
      "/var/folders/f8/92wybs9d7wqfthlqj1drht_c0000gn/T/ipykernel_20642/3217383995.py:49: FutureWarning: The frame.append method is deprecated and will be removed from pandas in a future version. Use pandas.concat instead.\n",
      "  dfe_data_all_inverse_var = dfe_data_all_inverse_var.append(next2, ignore_index=True)\n"
     ]
    }
   ],
   "source": [
    "for k in range(len(libraries)):\n",
    "    #had this stupid looking if condition to remove Ara-2 and Ara+4 from analysis very quickly without changing much\n",
    "    if k not in [103,1011]:\n",
    "        next1=pd.DataFrame()\n",
    "        next2=pd.DataFrame()\n",
    "        noness_pop = np.where(s_inverse_var[:,k,0]>-1)[0]\n",
    "        if k<2:\n",
    "            #replicate 1\n",
    "            #here, I'm correcting for the average fitness estimate for a pseudogene in that technical replicate\n",
    "            next1[\"Fitness estimate\"] = s_inverse_var[noness_pop,k,0]\n",
    "            next1[\"Population\"] = libraries2[k]\n",
    "            next1[\"Evolutionary History\"] = \"Ancestral\"\n",
    "            next1[\"Replicate\"] = \"Green\"\n",
    "            next1[\"Genes\"] = names[noness_pop]\n",
    "            next1[\"Mutator\"] = mutator[k]\n",
    "            \n",
    "            #replicate 2\n",
    "            next2[\"Fitness estimate\"] = s_inverse_var[noness_pop,k,1]\n",
    "            next2[\"Population\"] = libraries2[k]\n",
    "            next2[\"Evolutionary History\"] = \"Ancestral\"\n",
    "            next2[\"Replicate\"] = \"Red\"\n",
    "            next2[\"Genes\"] = names[noness_pop]\n",
    "            next2[\"Mutator\"] = mutator[k]\n",
    "\n",
    "        else:\n",
    "            #replicate 1\n",
    "            next1[\"Fitness estimate\"] = s_inverse_var[noness_pop,k,0]\n",
    "            next1[\"Population\"] = libraries2[k]\n",
    "            next1[\"Evolutionary History\"] = \"Evolved\"\n",
    "            next1[\"Replicate\"] = \"Green\"\n",
    "            next1[\"Genes\"] = names[noness_pop]\n",
    "            next1[\"Mutator\"] = mutator[k]\n",
    "\n",
    "            #replicate 2\n",
    "            next2[\"Fitness estimate\"] = s_inverse_var[noness_pop,k,1]\n",
    "            next2[\"Population\"] = libraries2[k]\n",
    "            next2[\"Evolutionary History\"] = \"Evolved\"\n",
    "            next2[\"Replicate\"] = \"Red\"\n",
    "            next2[\"Genes\"] = names[noness_pop]\n",
    "            next2[\"Mutator\"] = mutator[k]\n",
    "\n",
    "        if k in ara_minus:\n",
    "            next1[\"Ara Phenotype\"] = \"Ara-\"\n",
    "            next2[\"Ara Phenotype\"] = \"Ara-\"\n",
    "        else:\n",
    "            next1[\"Ara Phenotype\"] = \"Ara+\"\n",
    "            next2[\"Ara Phenotype\"] = \"Ara+\"\n",
    "        dfe_data_all_inverse_var = dfe_data_all_inverse_var.append(next1, ignore_index=True)\n",
    "        dfe_data_all_inverse_var = dfe_data_all_inverse_var.append(next2, ignore_index=True)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 31,
   "id": "hollow-muscle",
   "metadata": {},
   "outputs": [],
   "source": [
    "fname = savepath+f'dfe_data_pandas.csv'\n",
    "# dfe_data_all_inverse_var.to_csv(fname)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 32,
   "id": "elect-creature",
   "metadata": {},
   "outputs": [],
   "source": [
    "means_dfe = np.zeros(len(libraries))\n",
    "var_dfe = np.zeros(len(libraries))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 33,
   "id": "alpine-protest",
   "metadata": {},
   "outputs": [],
   "source": [
    "for k in range(len(libraries)):\n",
    "    rep1 = s_inverse_var[:,k,0][s_inverse_var[:,k,0]>-1]\n",
    "    rep2 = s_inverse_var[:,k,1][s_inverse_var[:,k,1]>-1]\n",
    "    means_dfe[k] = np.mean([rep1,rep2])\n",
    "    var_dfe[k] = np.var([rep1,rep2])"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "9de82fb3",
   "metadata": {},
   "source": [
    "### Now that we have the means and variances of the DFE, I will calculate the uncertainty in the mean and variances, given the measurement noise in fitness effect of each gene. Please see the methods section of the paper for the exact formulae."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 179,
   "id": "hired-canal",
   "metadata": {},
   "outputs": [],
   "source": [
    "def error_mean_dfe(gene_errors, pseudogene_errors):\n",
    "    error = np.zeros(len(libraries))\n",
    "    for k in range(len(libraries)):\n",
    "        genes_analyzed = gene_errors[:,k]>-1\n",
    "        pseudogenes_analyzed = pseudogene_errors[:,k]>-1\n",
    "        gene_contribution = np.sum(gene_errors[genes_analyzed,k]**2)/sum(genes_analyzed)**2\n",
    "        pseudogene_contribution = np.sum(pseudogene_errors[pseudogenes_analyzed,k]**2)/sum(pseudogenes_analyzed)**2\n",
    "        error[k] = np.sqrt(gene_contribution+pseudogene_contribution)\n",
    "    \n",
    "    return error"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 180,
   "id": "intensive-thesis",
   "metadata": {},
   "outputs": [],
   "source": [
    "error_mean = error_mean_dfe(standard_error, standard_error_pseudo)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 181,
   "id": "expired-creator",
   "metadata": {},
   "outputs": [],
   "source": [
    "error_mean_inv = error_mean_dfe(standard_error_inv, standard_error_pseudo_inv)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 182,
   "id": "continuous-exchange",
   "metadata": {},
   "outputs": [],
   "source": [
    "error_mean_hybrid = error_mean_dfe(standard_error_hybrid, standard_error_pseudo_hybrid)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 230,
   "id": "75796ead",
   "metadata": {},
   "outputs": [],
   "source": [
    "def error_var_dfe(gene_errors, pseudo_errors, fitness, fitness_pseudo, means_DFE, error_mean):\n",
    "    error = np.zeros(len(libraries))\n",
    "    for k in range(len(libraries)):\n",
    "        #pseudogenes\n",
    "        pseudo_analyzed = np.where(pseudo_errors[:,k]>-0.1)[0]\n",
    "        M = len(pseudo_analyzed)\n",
    "        pseudo_mean = np.mean(fitness_pseudo[pseudo_analyzed, k, :],axis=1)\n",
    "        #genes analyzed\n",
    "        genes_analyzed = np.where(gene_errors[:,k]>-1)[0]\n",
    "        N = len(genes_analyzed)\n",
    "        fitness_mean = np.mean(fitness[genes_analyzed, k, :],axis=1)\n",
    "        #coefficients\n",
    "        coeff = 2/N*(fitness_mean - means_DFE[k])\n",
    "        #errors\n",
    "        term1 = np.sum(gene_errors[genes_analyzed,k]**2*coeff**2)\n",
    "        \n",
    "        error[k] = np.sqrt(term1)\n",
    "    return error"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 231,
   "id": "going-chassis",
   "metadata": {},
   "outputs": [],
   "source": [
    "error_var = error_var_dfe(standard_error,standard_error_pseudo, s_inverse_var, s_inverse_pseudo, means_dfe, error_mean)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 232,
   "id": "infrared-passport",
   "metadata": {},
   "outputs": [],
   "source": [
    "error_var_inv = error_var_dfe(standard_error_inv,standard_error_pseudo_inv, s_inverse_var, s_inverse_pseudo, means_dfe, error_mean_inv)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 233,
   "id": "satisfactory-peter",
   "metadata": {},
   "outputs": [],
   "source": [
    "error_var_hybrid = error_var_dfe(standard_error_hybrid,standard_error_pseudo_hybrid, s_inverse_var, s_inverse_pseudo, means_dfe, error_mean_hybrid)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 185,
   "id": "widespread-station",
   "metadata": {},
   "outputs": [],
   "source": [
    "fname1 = savepath+f'mean_dfe.txt'\n",
    "fname2 = savepath+f'variance_dfe.txt'\n",
    "fname3 = savepath+f'error_mean.txt'\n",
    "fname4 = savepath+f'error_variance.txt'\n",
    "\n",
    "#saving the files now:\n",
    "# np.savetxt(fname1, means_dfe)\n",
    "# np.savetxt(fname2, var_dfe)\n",
    "# np.savetxt(fname3, error_mean)\n",
    "# np.savetxt(fname4, error_var)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 186,
   "id": "running-representation",
   "metadata": {},
   "outputs": [],
   "source": [
    "fname5 = savepath+f'error_mean_inv.txt'\n",
    "fname6 = savepath+f'error_variance_inv.txt'\n",
    "\n",
    "# np.savetxt(fname5, error_mean_inv)\n",
    "# np.savetxt(fname6, error_var_inv)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 187,
   "id": "southwest-picking",
   "metadata": {},
   "outputs": [],
   "source": [
    "fname7 = savepath+f'error_mean_hybrid.txt'\n",
    "fname8 = savepath+f'error_variance_hybrid.txt'\n",
    "\n",
    "# np.savetxt(fname7, error_mean_hybrid)\n",
    "# np.savetxt(fname8, error_var_hybrid)"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "minor-nickel",
   "metadata": {},
   "source": [
    "## Part 2: Calculating fitness effects using relaxed thresholds\n",
    "- It is possible that differences in populations (whether a gene is calculated or not) are due to arbitrariness of the thresholds. \n",
    "- Now, for every gene that has a fitness calculated with the strict thresholds in at least one population, I'll recalculate fitness with more relaxed requirements for how many TA sites are needed per gene.\n",
    "- If a fitness effect is not calculated even with relaxed thresholds, only then will I try to make an assessment of whether there is differential essentiality in LB\n",
    "\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 78,
   "id": "reasonable-turning",
   "metadata": {},
   "outputs": [],
   "source": [
    "def s_gene_pairwise(counts_red,counts_green,pop1,pop2,gene):\n",
    "    \"\"\"\n",
    "    pop1: where gene is not calculated\n",
    "    pop2: where gene is calculated\n",
    "    \"\"\"\n",
    "    green = np.sum(counts_green,axis=1)\n",
    "    red = np.sum(counts_red, axis=1)\n",
    "    \n",
    "    #doing this only for one gene\n",
    "    fitness_inverse_var = np.zeros(2)-1\n",
    "    stderr_fitness = -1\n",
    "    #TA sites within this gene\n",
    "    search_area = search_gene(locations=locations, ta_sites=ta_sites, i=gene)\n",
    "    #weights\n",
    "    weights_green = []\n",
    "    weights_red = []\n",
    "    #fitnesses\n",
    "    s_green = []\n",
    "    s_red = []\n",
    "    #defining time interval based on gene essentiality status:\n",
    "    t_start = 0\n",
    "    t_end = 4\n",
    "    time_gens = np.linspace(t_start,t_end,t_end-t_start+1)*6.64\n",
    "    max_reads=100\n",
    "    ta = np.sum(search_area==True)\n",
    "    #now, iterating over sites with at least tm1_depth reads at t-1:\n",
    "    for j in np.where((search_area==True))[0]:\n",
    "        if counts_green[0,j]/green[0]*10**7>tm1_depth:\n",
    "            #extracting the trajectories\n",
    "            traj_green = counts_green[t_start:t_end+1,j]/green[t_start:t_end+1]*10**7\n",
    "            traj_red = counts_red[t_start:t_end+1,j]/red[t_start:t_end+1]*10**7\n",
    "            #\n",
    "            n_0 = traj_green[0]\n",
    "            n_after_g = traj_green[1]\n",
    "            n_after_r = traj_red[1]\n",
    "            #this part is different:\n",
    "            #i'll find the first occurrence of zero if at all, and then fit a trajectory upto that point\n",
    "            #green replicate first\n",
    "            if 0 not in traj_green:\n",
    "                sg = np.polyfit(time_gens, np.log(traj_green), 1)[0]\n",
    "                wg = min(inverse_var_weight(n_0, n_after_g), inverse_var_weight(max_reads, max_reads))\n",
    "                #sotring the data\n",
    "                weights_green.append(wg)\n",
    "                s_green.append(sg)\n",
    "            else:\n",
    "                #first zero in the red dataset:\n",
    "                ind_g = int(np.where(traj_green==0)[0][0])\n",
    "                counts_g = traj_green[:ind_g+1].copy()\n",
    "                #add a pseudocount to the first occurrence of zero\n",
    "                counts_g[-1] = 1\n",
    "                sg = np.polyfit(time_gens[:ind_g+1], np.log(counts_g), 1)[0]\n",
    "                wg = min(inverse_var_weight(n_0, n_after_g), inverse_var_weight(max_reads, max_reads))\n",
    "                #storing the data\n",
    "                weights_green.append(wg)\n",
    "                s_green.append(sg)\n",
    "            #now the red replicate\n",
    "            if 0 not in traj_red:\n",
    "                sg = np.polyfit(time_gens, np.log(traj_red), 1)[0]\n",
    "                wg = min(inverse_var_weight(n_0, n_after_r), inverse_var_weight(max_reads, max_reads))\n",
    "                #sotring the data\n",
    "                weights_red.append(wg)\n",
    "                s_red.append(sg)\n",
    "            else:\n",
    "                #first zero in the red dataset:\n",
    "                ind_g = int(np.where(traj_red==0)[0][0])\n",
    "                counts_g = traj_red[:ind_g+1].copy()\n",
    "                #add a pseudocount to the first occurrence of zero\n",
    "                counts_g[-1] = 1\n",
    "                sg = np.polyfit(time_gens[:ind_g+1], np.log(counts_g), 1)[0]\n",
    "                wg = min(inverse_var_weight(n_0, n_after_r), inverse_var_weight(max_reads, max_reads))\n",
    "                #storing the data\n",
    "                weights_red.append(wg)\n",
    "                s_red.append(sg)\n",
    "\n",
    "    #converting to numpy arrays\n",
    "    s_green = np.array(s_green)\n",
    "    s_red = np.array(s_red)\n",
    "    weights_green = np.array(weights_green)\n",
    "    weights_red = np.array(weights_red)\n",
    "\n",
    "    #pooling over all sites, note the relaxed thresholds\n",
    "    if min(len(s_green),len(s_red))>=1 and min(len(s_green),len(s_red))/ta>=0.1:\n",
    "\n",
    "        #weighing based on inverse variacne:\n",
    "        fitness_inverse_var[0] = np.average(s_green, weights=weights_green)\n",
    "        fitness_inverse_var[1] = np.average(s_red, weights=weights_red)\n",
    "\n",
    "        #standard error in fitness estimates:\n",
    "        stderr_fitness = np.std(np.hstack([s_green,s_red]))/np.sqrt(len(s_green)+len(s_red))\n",
    "\n",
    "    return fitness_inverse_var, stderr_fitness"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 79,
   "id": "ignored-dubai",
   "metadata": {},
   "outputs": [],
   "source": [
    "#making a copy of the fitness estimates\n",
    "s_inverse_var_relaxed = np.copy(s_inverse_var)"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "stable-essence",
   "metadata": {},
   "source": [
    "First, for every gene that has a fitness calculated in the evolved clone, try and calculate a fitness effect in the ancestor REL606 using the relaxed thresholds"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 81,
   "id": "august-fashion",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "REL607\n",
      "anc -1.0 REL607 0.016 913\n",
      "anc -1.0 REL607 -0.003 1054\n",
      "anc -1.0 REL607 -0.095 1791\n",
      "27 Number of genes not calculated in anc but calculated in REL607 24\n",
      "REL11330\n",
      "anc -1.0 REL11330 0.014 295\n",
      "anc -1.0 REL11330 -0.114 357\n",
      "anc -1.0 REL11330 -0.108 358\n",
      "anc -1.0 REL11330 -0.2 848\n",
      "anc -1.0 REL11330 -0.06 913\n",
      "anc -0.378 REL11330 -0.059 1837\n",
      "anc -1.0 REL11330 -0.092 1926\n",
      "anc -1.0 REL11330 -0.192 2139\n",
      "anc -1.0 REL11330 -0.0 2242\n",
      "anc -1.0 REL11330 -0.063 3134\n",
      "anc -1.0 REL11330 -0.063 3244\n",
      "40 Number of genes not calculated in anc but calculated in REL11330 29\n",
      "REL11364\n",
      "anc -1.0 REL11364 -0.044 547\n",
      "anc -1.0 REL11364 -0.197 622\n",
      "anc -1.0 REL11364 -0.074 924\n",
      "anc -1.0 REL11364 -0.15 1054\n",
      "anc -1.0 REL11364 -0.177 1120\n",
      "anc -1.0 REL11364 -0.166 1228\n",
      "anc -1.0 REL11364 -0.005 1315\n",
      "anc -1.0 REL11364 -0.057 1366\n",
      "anc -1.0 REL11364 -0.014 1926\n",
      "anc -0.55 REL11364 -0.162 3588\n",
      "40 Number of genes not calculated in anc but calculated in REL11364 30\n",
      "REL11336\n",
      "anc -1.0 REL11336 -0.01 1054\n",
      "anc -1.0 REL11336 -0.127 1120\n",
      "anc -0.378 REL11336 -0.026 1837\n",
      "anc -1.0 REL11336 -0.015 1926\n",
      "anc -0.366 REL11336 -0.167 2226\n",
      "anc -1.0 REL11336 -0.06 3090\n",
      "anc -1.0 REL11336 -0.029 3134\n",
      "anc -1.0 REL11336 0.005 3381\n",
      "anc -1.0 REL11336 -0.004 3382\n",
      "anc -0.39 REL11336 0.002 3384\n",
      "anc -0.55 REL11336 -0.166 3588\n",
      "40 Number of genes not calculated in anc but calculated in REL11336 29\n",
      "REL11339\n",
      "anc -1.0 REL11339 -0.094 357\n",
      "anc -1.0 REL11339 -0.189 358\n",
      "anc -1.0 REL11339 -0.001 913\n",
      "anc -1.0 REL11339 -0.187 1120\n",
      "anc -1.0 REL11339 -0.085 1315\n",
      "anc -0.429 REL11339 -0.175 1818\n",
      "anc -1.0 REL11339 -0.033 1926\n",
      "anc -1.0 REL11339 -0.006 3155\n",
      "anc -1.0 REL11339 -0.071 3244\n",
      "anc -0.55 REL11339 -0.163 3588\n",
      "anc -1.0 REL11339 -0.113 3840\n",
      "42 Number of genes not calculated in anc but calculated in REL11339 31\n",
      "REL11389\n",
      "anc -1.0 REL11389 -0.092 154\n",
      "anc -1.0 REL11389 -0.079 1639\n",
      "anc -1.0 REL11389 -0.032 1697\n",
      "anc -1.0 REL11389 -0.044 1707\n",
      "anc -1.0 REL11389 -0.075 1737\n",
      "anc -0.429 REL11389 -0.04 1818\n",
      "anc -0.378 REL11389 -0.02 1837\n",
      "anc -1.0 REL11389 -0.081 1855\n",
      "anc -1.0 REL11389 -0.037 1878\n",
      "anc -1.0 REL11389 -0.049 1916\n",
      "anc -1.0 REL11389 -0.013 1949\n",
      "anc -1.0 REL11389 -0.076 1951\n",
      "anc -1.0 REL11389 -0.04 1988\n",
      "anc -1.0 REL11389 -0.024 1989\n",
      "anc -1.0 REL11389 -0.032 1991\n",
      "anc -1.0 REL11389 -0.029 1992\n",
      "anc -1.0 REL11389 -0.044 2128\n",
      "anc -1.0 REL11389 -0.054 2133\n",
      "anc -1.0 REL11389 -0.06 2134\n",
      "anc -0.46 REL11389 -0.087 2137\n",
      "anc -0.33 REL11389 -0.065 2138\n",
      "anc -1.0 REL11389 -0.03 2139\n",
      "anc -1.0 REL11389 -0.048 2184\n",
      "anc -1.0 REL11389 -0.049 2191\n",
      "anc -1.0 REL11389 -0.049 2225\n",
      "anc -0.366 REL11389 -0.049 2226\n",
      "anc -1.0 REL11389 -0.046 2228\n",
      "anc -1.0 REL11389 -0.049 2229\n",
      "anc -1.0 REL11389 -0.069 2242\n",
      "anc -1.0 REL11389 -0.074 2243\n",
      "anc -1.0 REL11389 -0.034 2246\n",
      "anc -1.0 REL11389 -0.066 2276\n",
      "anc -1.0 REL11389 -0.045 2279\n",
      "anc -1.0 REL11389 -0.041 2281\n",
      "anc -1.0 REL11389 -0.052 2286\n",
      "anc -1.0 REL11389 -0.033 2298\n",
      "anc -0.55 REL11389 -0.167 3588\n",
      "61 Number of genes not calculated in anc but calculated in REL11389 24\n",
      "REL11392\n",
      "anc -1.0 REL11392 -0.184 621\n",
      "anc -1.0 REL11392 -0.184 622\n",
      "anc -1.0 REL11392 -0.008 1054\n",
      "anc -1.0 REL11392 -0.186 1120\n",
      "anc -1.0 REL11392 -0.049 1565\n",
      "anc -0.378 REL11392 0.008 1837\n",
      "anc -1.0 REL11392 -0.013 1926\n",
      "anc -1.0 REL11392 -0.187 2062\n",
      "anc -1.0 REL11392 -0.096 2242\n",
      "anc -1.0 REL11392 -0.057 2837\n",
      "anc -1.0 REL11392 -0.191 3382\n",
      "anc -1.0 REL11392 -0.036 3840\n",
      "47 Number of genes not calculated in anc but calculated in REL11392 35\n",
      "REL11342\n",
      "anc -1.0 REL11342 -0.082 621\n",
      "anc -1.0 REL11342 -0.091 622\n",
      "anc -0.326 REL11342 -0.147 812\n",
      "anc -1.0 REL11342 -0.063 913\n",
      "anc -1.0 REL11342 -0.03 1049\n",
      "anc -1.0 REL11342 -0.131 1120\n",
      "anc -1.0 REL11342 0.015 1315\n",
      "anc -1.0 REL11342 -0.002 1926\n",
      "anc -1.0 REL11342 -0.011 2362\n",
      "anc -1.0 REL11342 0.002 2363\n",
      "anc -1.0 REL11342 -0.002 2413\n",
      "anc -1.0 REL11342 -0.002 2414\n",
      "anc -0.382 REL11342 0.009 2415\n",
      "anc -1.0 REL11342 -0.109 2837\n",
      "anc -1.0 REL11342 -0.098 3244\n",
      "anc -0.39 REL11342 -0.132 3384\n",
      "anc -0.55 REL11342 -0.153 3588\n",
      "44 Number of genes not calculated in anc but calculated in REL11342 27\n",
      "REL11345\n",
      "anc -1.0 REL11345 -0.092 621\n",
      "anc -1.0 REL11345 -0.145 622\n",
      "anc -0.326 REL11345 -0.142 812\n",
      "anc -1.0 REL11345 -0.178 1120\n",
      "anc -1.0 REL11345 0.012 1344\n",
      "anc -1.0 REL11345 -0.193 3382\n",
      "anc -0.39 REL11345 -0.149 3384\n",
      "25 Number of genes not calculated in anc but calculated in REL11345 18\n",
      "REL11348\n",
      "anc -1.0 REL11348 -0.11 357\n",
      "anc -1.0 REL11348 -0.122 358\n",
      "anc -1.0 REL11348 0.013 913\n",
      "anc -1.0 REL11348 -0.038 2413\n",
      "anc -1.0 REL11348 0.023 2414\n",
      "anc -0.382 REL11348 -0.001 2415\n",
      "anc -1.0 REL11348 -0.059 3382\n",
      "anc -0.39 REL11348 -0.067 3384\n",
      "anc -0.55 REL11348 -0.176 3588\n",
      "17 Number of genes not calculated in anc but calculated in REL11348 8\n",
      "REL11367\n",
      "anc -1.0 REL11367 -0.015 621\n",
      "anc -1.0 REL11367 -0.128 1120\n",
      "16 Number of genes not calculated in anc but calculated in REL11367 14\n",
      "REL11370\n",
      "anc -1.0 REL11370 -0.194 325\n",
      "anc -1.0 REL11370 0.014 1120\n",
      "anc -1.0 REL11370 -0.046 1926\n",
      "anc -1.0 REL11370 -0.192 2139\n",
      "anc -1.0 REL11370 -0.002 2242\n",
      "anc -1.0 REL11370 -0.178 2315\n",
      "anc -1.0 REL11370 -0.141 2837\n",
      "anc -1.0 REL11370 -0.027 3381\n",
      "anc -1.0 REL11370 -0.032 3382\n",
      "anc -0.39 REL11370 -0.019 3384\n",
      "anc -1.0 REL11370 -0.108 3585\n",
      "anc -0.55 REL11370 -0.187 3588\n",
      "anc -1.0 REL11370 0.037 3677\n",
      "30 Number of genes not calculated in anc but calculated in REL11370 17\n"
     ]
    }
   ],
   "source": [
    "non_ess_cutoff = -0.2\n",
    "cutoffs = np.ones(14)*-0.3\n",
    "for k in range(1,len(libraries)):\n",
    "    #only considering pseudogenes which have a fitness of -0.05 or higher. SOme appear deleterious, presumably\n",
    "    #from polar effects\n",
    "    anc_num = 0\n",
    "    #correction factors for fitness estimation.\n",
    "    pseudogenes_analyzed = np.where((s_inverse_pseudo[:,anc_num,0]>-0.05)&(s_inverse_pseudo[:,anc_num,1]>-0.05))[0]\n",
    "    corrections_anc = np.mean(s_inverse_pseudo[pseudogenes_analyzed,anc_num,:],axis=0)\n",
    "\n",
    "    #exclude ara-2 from analysis\n",
    "    if k!=3:\n",
    "        #counter variables\n",
    "        count=0\n",
    "        print(libraries[k])\n",
    "        #extracting genes that are -1 in ancestor and near neutral in the evolved population\n",
    "        ess_anc = np.where((np.mean(s_inverse_var[:,anc_num,:],axis=1)==-1)&(np.mean(s_inverse_var[:,k,:],axis=1)>non_ess_cutoff))[0]\n",
    "        \n",
    "        #relaxed thresholds\n",
    "        for gene in ess_anc:\n",
    "            s_relaxed, std_relaxed = s_gene_pairwise(counts_all_green[anc_num,:,:], counts_all_red[anc_num,:,:],anc_num,k,gene)\n",
    "            #updating the relaxed fitness estimates array for the ancestor (gene was -1 in ancestor):\n",
    "            s_inverse_var_relaxed[gene,anc_num,:] = s_relaxed - corrections_anc*(np.mean(s_relaxed)!=-1)\n",
    "            #updating the counter\n",
    "            if np.mean(s_relaxed)>cutoffs[anc_num]:\n",
    "                count+=1\n",
    "            else:\n",
    "                print('anc',np.round(np.mean(s_relaxed),3),libraries[k],np.round(np.mean(s_inverse_var[gene,k,:]),3),gene)\n",
    "                \n",
    "        print(np.size(ess_anc),f'Number of genes not calculated in anc but calculated in {libraries[k]}',\n",
    "              count)\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 82,
   "id": "beneficial-banana",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "REL607\n",
      "REL607 -0.306 anc -0.077 379\n",
      "REL607 -1.0 anc -0.039 922\n",
      "REL607 -1.0 anc -0.12 1026\n",
      "REL607 -1.0 anc -0.004 1138\n",
      "REL607 -1.0 anc -0.082 1782\n",
      "REL607 -1.0 anc -0.044 1958\n",
      "REL607 -1.0 anc -0.001 3202\n",
      "REL607 -1.0 anc -0.034 3460\n",
      "REL607 -1.0 anc 0.023 3811\n",
      "REL607 -1.0 anc -0.003 3911\n",
      "57 Number of genes  calculated in anc but not calculated in REL607 47\n",
      "REL11330\n",
      "REL11330 -1.0 anc -0.083 764\n",
      "REL11330 -1.0 anc -0.151 835\n",
      "REL11330 -1.0 anc -0.094 1850\n",
      "REL11330 -1.0 anc -0.044 1958\n",
      "REL11330 -1.0 anc -0.066 2750\n",
      "REL11330 -1.0 anc -0.001 3202\n",
      "REL11330 -0.302 anc -0.004 3899\n",
      "REL11330 -0.313 anc -0.003 3911\n",
      "40 Number of genes  calculated in anc but not calculated in REL11330 32\n",
      "REL11364\n",
      "REL11364 -1.0 anc -0.094 86\n",
      "REL11364 -0.345 anc 0.015 235\n",
      "REL11364 -1.0 anc 0.007 302\n",
      "REL11364 -1.0 anc -0.107 390\n",
      "REL11364 -1.0 anc -0.11 751\n",
      "REL11364 -1.0 anc 0.002 843\n",
      "REL11364 -1.0 anc 0.013 1293\n",
      "REL11364 -1.0 anc -0.195 1376\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "/Users/anuraglimdi/anaconda3/envs/sns_env/lib/python3.9/site-packages/numpy/core/fromnumeric.py:3419: RuntimeWarning: Mean of empty slice.\n",
      "  return _methods._mean(a, axis=axis, dtype=dtype,\n",
      "/Users/anuraglimdi/anaconda3/envs/sns_env/lib/python3.9/site-packages/numpy/core/_methods.py:180: RuntimeWarning: invalid value encountered in true_divide\n",
      "  ret = um.true_divide(\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "REL11364 -1.0 anc -0.002 3457\n",
      "REL11364 -1.0 anc -0.032 3716\n",
      "37 Number of genes  calculated in anc but not calculated in REL11364 27\n",
      "REL11336\n",
      "REL11336 -1.0 anc -0.094 86\n",
      "REL11336 -1.0 anc -0.028 335\n",
      "REL11336 -1.0 anc -0.151 835\n",
      "REL11336 -1.0 anc 0.002 843\n",
      "REL11336 -1.0 anc -0.157 999\n",
      "REL11336 -1.0 anc -0.063 1059\n",
      "REL11336 -1.0 anc -0.117 1060\n",
      "REL11336 -1.0 anc -0.151 1071\n",
      "REL11336 -1.0 anc -0.004 1138\n",
      "REL11336 -1.0 anc -0.118 1172\n",
      "REL11336 -1.0 anc -0.197 1174\n",
      "REL11336 -1.0 anc 0.013 1293\n",
      "REL11336 -1.0 anc -0.076 1686\n",
      "REL11336 -1.0 anc -0.168 1939\n",
      "REL11336 -1.0 anc -0.091 2101\n",
      "REL11336 -1.0 anc 0.003 2206\n",
      "REL11336 -1.0 anc -0.167 2244\n",
      "REL11336 -1.0 anc -0.135 2326\n",
      "REL11336 -0.353 anc -0.059 2763\n",
      "REL11336 -1.0 anc -0.127 2937\n",
      "REL11336 -1.0 anc -0.198 3138\n",
      "REL11336 -1.0 anc 0.004 3432\n",
      "REL11336 -1.0 anc -0.002 3433\n",
      "REL11336 -1.0 anc -0.056 3435\n",
      "REL11336 -1.0 anc -0.047 3436\n",
      "REL11336 -1.0 anc -0.108 3802\n",
      "REL11336 -1.0 anc -0.028 3886\n",
      "REL11336 -1.0 anc -0.003 3911\n",
      "67 Number of genes  calculated in anc but not calculated in REL11336 39\n",
      "REL11339\n",
      "REL11339 -1.0 anc -0.094 86\n",
      "REL11339 -1.0 anc 0.015 235\n",
      "REL11339 -0.378 anc -0.107 390\n",
      "REL11339 -1.0 anc -0.063 1059\n",
      "REL11339 -1.0 anc -0.117 1060\n",
      "REL11339 -0.348 anc -0.076 1686\n",
      "REL11339 -1.0 anc -0.168 1939\n",
      "REL11339 -0.362 anc -0.172 2036\n",
      "REL11339 -1.0 anc -0.014 2065\n",
      "REL11339 -1.0 anc -0.002 3433\n",
      "REL11339 -0.313 anc -0.056 3435\n",
      "REL11339 -1.0 anc -0.047 3436\n",
      "REL11339 -0.438 anc -0.152 3492\n",
      "REL11339 -1.0 anc -0.042 3512\n",
      "REL11339 -1.0 anc -0.082 3582\n",
      "REL11339 -1.0 anc -0.004 3899\n",
      "REL11339 -1.0 anc -0.003 3911\n",
      "50 Number of genes  calculated in anc but not calculated in REL11339 33\n",
      "REL11389\n",
      "REL11389 -1.0 anc -0.094 86\n",
      "REL11389 -1.0 anc -0.009 165\n",
      "REL11389 -1.0 anc -0.026 232\n",
      "REL11389 -1.0 anc -0.107 390\n",
      "REL11389 -1.0 anc -0.11 751\n",
      "REL11389 -1.0 anc -0.157 999\n",
      "REL11389 -1.0 anc -0.004 1138\n",
      "REL11389 -1.0 anc -0.081 1417\n",
      "REL11389 -1.0 anc -0.079 1440\n",
      "REL11389 -1.0 anc -0.059 1838\n",
      "REL11389 -1.0 anc -0.094 1850\n",
      "REL11389 -1.0 anc -0.044 1958\n",
      "REL11389 -1.0 anc -0.088 2574\n",
      "REL11389 -1.0 anc -0.124 2785\n",
      "REL11389 -1.0 anc -0.128 2872\n",
      "REL11389 -1.0 anc -0.001 3202\n",
      "REL11389 -1.0 anc 0.004 3432\n",
      "REL11389 -1.0 anc -0.002 3433\n",
      "REL11389 -1.0 anc -0.056 3435\n",
      "REL11389 -1.0 anc -0.047 3436\n",
      "REL11389 -1.0 anc -0.034 3460\n",
      "REL11389 -1.0 anc -0.023 3761\n",
      "REL11389 -1.0 anc -0.004 3899\n",
      "REL11389 -1.0 anc -0.003 3911\n",
      "80 Number of genes  calculated in anc but not calculated in REL11389 56\n",
      "REL11392\n",
      "REL11392 -1.0 anc -0.052 756\n",
      "REL11392 -1.0 anc 0.002 843\n",
      "REL11392 -1.0 anc -0.019 1070\n",
      "REL11392 -1.0 anc -0.195 1376\n",
      "REL11392 -1.0 anc -0.002 2041\n",
      "REL11392 -1.0 anc -0.128 2872\n",
      "REL11392 -1.0 anc -0.076 2916\n",
      "REL11392 -1.0 anc -0.001 3202\n",
      "REL11392 -0.386 anc -0.146 3423\n",
      "REL11392 -1.0 anc -0.002 3433\n",
      "REL11392 -1.0 anc -0.056 3435\n",
      "REL11392 -1.0 anc -0.047 3436\n",
      "49 Number of genes  calculated in anc but not calculated in REL11392 37\n",
      "REL11342\n",
      "REL11342 -1.0 anc -0.054 71\n",
      "REL11342 -1.0 anc -0.094 86\n",
      "REL11342 -0.351 anc -0.009 165\n",
      "REL11342 -1.0 anc -0.03 296\n",
      "REL11342 -0.403 anc -0.135 347\n",
      "REL11342 -1.0 anc 0.027 544\n",
      "REL11342 -0.338 anc 0.014 545\n",
      "REL11342 -1.0 anc 0.002 843\n",
      "REL11342 -1.0 anc -0.035 928\n",
      "REL11342 -1.0 anc -0.063 1059\n",
      "REL11342 -1.0 anc -0.004 1138\n",
      "REL11342 -1.0 anc -0.044 1958\n",
      "REL11342 -1.0 anc 0.032 2890\n",
      "REL11342 -1.0 anc -0.076 2916\n",
      "REL11342 -1.0 anc -0.05 3135\n",
      "REL11342 -1.0 anc -0.146 3423\n",
      "REL11342 -1.0 anc 0.004 3432\n",
      "REL11342 -1.0 anc -0.002 3433\n",
      "REL11342 -1.0 anc -0.056 3435\n",
      "REL11342 -1.0 anc -0.047 3436\n",
      "REL11342 -0.409 anc 0.0 3437\n",
      "REL11342 -1.0 anc -0.014 3524\n",
      "REL11342 -0.323 anc -0.032 3716\n",
      "REL11342 -1.0 anc -0.004 3899\n",
      "63 Number of genes  calculated in anc but not calculated in REL11342 39\n",
      "REL11345\n",
      "REL11345 -1.0 anc -0.026 232\n",
      "REL11345 -0.359 anc 0.015 235\n",
      "REL11345 -1.0 anc -0.135 347\n",
      "REL11345 -0.328 anc -0.07 349\n",
      "REL11345 -1.0 anc -0.107 390\n",
      "REL11345 -1.0 anc -0.188 579\n",
      "REL11345 -0.44 anc 0.002 843\n",
      "REL11345 -1.0 anc -0.004 1243\n",
      "REL11345 -0.311 anc -0.195 1376\n",
      "REL11345 -1.0 anc -0.081 1417\n",
      "REL11345 -1.0 anc -0.061 2033\n",
      "REL11345 -1.0 anc -0.091 2101\n",
      "REL11345 -1.0 anc -0.05 2330\n",
      "REL11345 -1.0 anc 0.032 2890\n",
      "REL11345 -1.0 anc -0.034 3460\n",
      "REL11345 -1.0 anc -0.032 3716\n",
      "REL11345 -0.405 anc -0.004 3899\n",
      "REL11345 -1.0 anc -0.003 3911\n",
      "63 Number of genes  calculated in anc but not calculated in REL11345 45\n",
      "REL11348\n",
      "REL11348 -0.382 anc 0.011 36\n",
      "REL11348 -1.0 anc -0.094 86\n",
      "REL11348 -1.0 anc -0.026 232\n",
      "REL11348 -1.0 anc 0.015 235\n",
      "REL11348 -1.0 anc -0.03 296\n",
      "REL11348 -0.465 anc 0.014 303\n",
      "REL11348 -0.575 anc 0.015 339\n",
      "REL11348 -1.0 anc -0.07 349\n",
      "REL11348 -0.388 anc 0.011 350\n",
      "REL11348 -1.0 anc -0.029 425\n",
      "REL11348 -0.361 anc -0.04 451\n",
      "REL11348 -0.449 anc -0.115 540\n",
      "REL11348 -0.383 anc -0.11 751\n",
      "REL11348 -0.514 anc 0.024 753\n",
      "REL11348 -1.0 anc -0.052 756\n",
      "REL11348 -1.0 anc -0.083 764\n",
      "REL11348 -1.0 anc -0.013 871\n",
      "REL11348 -0.378 anc -0.005 914\n",
      "REL11348 -0.501 anc -0.12 1026\n",
      "REL11348 -1.0 anc -0.004 1138\n",
      "REL11348 -0.487 anc -0.006 1157\n",
      "REL11348 -0.367 anc -0.002 1177\n",
      "REL11348 -1.0 anc -0.027 1341\n",
      "REL11348 -1.0 anc -0.112 1349\n",
      "REL11348 -1.0 anc -0.171 1378\n",
      "REL11348 -1.0 anc 0.013 1387\n",
      "REL11348 -1.0 anc -0.023 1394\n",
      "REL11348 -0.508 anc -0.081 1417\n",
      "REL11348 -0.346 anc -0.05 1434\n",
      "REL11348 -1.0 anc -0.079 1440\n",
      "REL11348 -1.0 anc -0.001 1608\n",
      "REL11348 -0.301 anc -0.016 1744\n",
      "REL11348 -1.0 anc -0.016 1861\n",
      "REL11348 -0.323 anc -0.168 1939\n",
      "REL11348 -0.325 anc -0.044 1958\n",
      "REL11348 -0.503 anc -0.023 2079\n",
      "REL11348 -0.374 anc 0.016 2196\n",
      "REL11348 -1.0 anc -0.189 2341\n",
      "REL11348 -1.0 anc -0.007 2401\n",
      "REL11348 -0.329 anc -0.034 2542\n",
      "REL11348 -1.0 anc -0.088 2574\n",
      "REL11348 -0.587 anc -0.013 2712\n",
      "REL11348 -1.0 anc -0.092 2716\n",
      "REL11348 -1.0 anc -0.124 2785\n",
      "REL11348 -0.384 anc -0.128 2872\n",
      "REL11348 -1.0 anc 0.032 2890\n",
      "REL11348 -0.329 anc -0.076 2916\n",
      "REL11348 -0.367 anc -0.002 2918\n",
      "REL11348 -1.0 anc 0.015 2985\n",
      "REL11348 -0.316 anc -0.097 3091\n",
      "REL11348 -1.0 anc -0.05 3135\n",
      "REL11348 -1.0 anc 0.023 3152\n",
      "REL11348 -1.0 anc -0.041 3160\n",
      "REL11348 -0.558 anc -0.028 3261\n",
      "REL11348 -1.0 anc -0.03 3272\n",
      "REL11348 -1.0 anc -0.036 3274\n",
      "REL11348 -0.366 anc -0.146 3423\n",
      "REL11348 -1.0 anc -0.047 3436\n",
      "REL11348 -0.766 anc 0.007 3577\n",
      "REL11348 -0.462 anc -0.157 3760\n",
      "REL11348 -0.453 anc 0.015 3781\n",
      "REL11348 -0.306 anc 0.023 3811\n",
      "REL11348 -1.0 anc -0.004 3899\n",
      "REL11348 -1.0 anc -0.003 3911\n",
      "REL11348 -1.0 anc -0.024 3937\n",
      "REL11348 -1.0 anc -0.035 3951\n",
      "138 Number of genes  calculated in anc but not calculated in REL11348 72\n",
      "REL11367\n",
      "REL11367 -1.0 anc -0.04 451\n",
      "REL11367 -1.0 anc -0.028 470\n",
      "REL11367 -1.0 anc -0.004 617\n",
      "REL11367 -1.0 anc -0.12 1026\n",
      "REL11367 -1.0 anc -0.019 1034\n",
      "REL11367 -1.0 anc -0.004 1138\n",
      "REL11367 -1.0 anc -0.006 1157\n",
      "REL11367 -1.0 anc -0.03 1351\n",
      "REL11367 -1.0 anc -0.004 1388\n",
      "REL11367 -0.321 anc -0.023 1394\n",
      "REL11367 -1.0 anc -0.076 1686\n",
      "REL11367 -1.0 anc -0.082 1782\n",
      "REL11367 -1.0 anc -0.016 1861\n",
      "REL11367 -1.0 anc -0.168 1939\n",
      "REL11367 -1.0 anc 0.031 2238\n",
      "REL11367 -1.0 anc -0.038 2519\n",
      "REL11367 -1.0 anc 0.007 2724\n",
      "REL11367 -1.0 anc -0.024 2850\n",
      "REL11367 -1.0 anc 0.032 2890\n",
      "REL11367 -1.0 anc -0.076 2916\n",
      "REL11367 -1.0 anc -0.109 3053\n",
      "REL11367 -1.0 anc -0.05 3135\n",
      "REL11367 -1.0 anc -0.001 3202\n",
      "REL11367 -1.0 anc -0.139 3278\n",
      "REL11367 -1.0 anc -0.056 3435\n",
      "REL11367 -1.0 anc -0.047 3436\n",
      "REL11367 -1.0 anc -0.034 3460\n",
      "REL11367 -1.0 anc -0.023 3761\n",
      "REL11367 -1.0 anc 0.023 3811\n",
      "REL11367 -1.0 anc -0.028 3886\n",
      "REL11367 -1.0 anc -0.035 3951\n",
      "117 Number of genes  calculated in anc but not calculated in REL11367 86\n",
      "REL11370\n",
      "REL11370 -1.0 anc -0.094 86\n",
      "REL11370 -1.0 anc -0.009 165\n",
      "REL11370 -1.0 anc -0.026 232\n",
      "REL11370 -1.0 anc 0.015 235\n",
      "REL11370 -1.0 anc -0.047 569\n",
      "REL11370 -1.0 anc -0.078 666\n",
      "REL11370 -1.0 anc 0.001 712\n",
      "REL11370 -1.0 anc 0.002 843\n",
      "REL11370 -1.0 anc -0.063 1059\n",
      "REL11370 -1.0 anc -0.004 1138\n",
      "REL11370 -1.0 anc -0.13 1239\n",
      "REL11370 -1.0 anc 0.004 1416\n",
      "REL11370 -1.0 anc -0.016 1861\n",
      "REL11370 -0.311 anc -0.168 1939\n",
      "REL11370 -1.0 anc -0.018 2323\n",
      "REL11370 -1.0 anc -0.003 2340\n",
      "REL11370 -1.0 anc -0.088 2574\n",
      "REL11370 -1.0 anc -0.001 2623\n",
      "REL11370 -0.433 anc 0.032 2890\n",
      "REL11370 -0.355 anc -0.109 3053\n",
      "REL11370 -1.0 anc -0.05 3135\n",
      "REL11370 -0.405 anc -0.198 3138\n",
      "REL11370 -1.0 anc -0.034 3460\n",
      "REL11370 -1.0 anc -0.152 3492\n",
      "REL11370 -1.0 anc -0.157 3760\n",
      "REL11370 -0.305 anc -0.004 3899\n",
      "REL11370 -1.0 anc -0.003 3911\n",
      "REL11370 -1.0 anc -0.035 3951\n",
      "95 Number of genes  calculated in anc but not calculated in REL11370 67\n"
     ]
    }
   ],
   "source": [
    "for k in range(1,len(libraries)):\n",
    "    anc_num = 0\n",
    "    #correction factors for fitness estimation.\n",
    "    pseudogenes_analyzed = np.where((s_inverse_pseudo[:,k,0]>-0.05)&(s_inverse_pseudo[:,k,1]>-0.05))[0]\n",
    "    corrections_ev = np.mean(s_inverse_pseudo[pseudogenes_analyzed,k,:],axis=0)\n",
    "    if k!=3:\n",
    "        count2=0\n",
    "        print(libraries[k])\n",
    "        #the following statement means: gene is not calculated in the evolved clone, but calculated in the ancestor\n",
    "        #with the relaxed thresholds for number of TA sites etc., and that the gene is not lost over evolution\n",
    "        ess_ev = np.where((np.mean(s_inverse_var[:,k,:],axis=1)==-1)&(np.mean(s_inverse_var_relaxed[:,anc_num,:],axis=1)>non_ess_cutoff)&(exclude_genes[k,:]==0))[0]\n",
    "        \n",
    "        for gene in ess_ev:\n",
    "            s_relaxed, std_relaxed = s_gene_pairwise(counts_all_green[k,:,:], counts_all_red[k,:,:],k,anc_num,gene)\n",
    "            #updating the relaxed fitness estimates array for the evolved (gene was -1 in evolved):\n",
    "            s_inverse_var_relaxed[gene,k,:] = s_relaxed - corrections_ev*(np.mean(s_relaxed)!=-1)\n",
    "            if np.mean(s_relaxed)>cutoffs[k]:\n",
    "                count2+=1\n",
    "            else:\n",
    "                print(libraries[k],np.round(np.mean(s_relaxed),3),'anc',np.round(np.mean(s_inverse_var_relaxed[gene,anc_num,:]),3),gene)\n",
    "        \n",
    "        print(np.size(ess_ev),f'Number of genes  calculated in anc but not calculated in {libraries[k]}',\n",
    "              count2)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 83,
   "id": "golden-israel",
   "metadata": {},
   "outputs": [],
   "source": [
    "fname = savepath+f'fitness_genes_relaxed_thresholds_updated'\n",
    "# np.save(fname,s_inverse_var_relaxed)"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "veterinary-strip",
   "metadata": {},
   "source": [
    "### Outputting additional data:\n",
    "\n",
    "- trajectories for rffG and rffH\n",
    "- p_vals_fdr (just for making the figure on defining essentiality in LB, not anything more)\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 17,
   "id": "public-sarah",
   "metadata": {},
   "outputs": [],
   "source": [
    "loc_rffg = locations[list(names).index('rffG'),:]\n",
    "loc_rffh = locations[list(names).index('rffH'),:]\n",
    "# np.save(savepath+'rffH_trajectories_red', counts_all_red[:,:,search_gene(locations,ta_sites,list(names).index('rffH'))])\n",
    "# np.save(savepath+'rffG_trajectories_red', counts_all_red[:,:,search_gene(locations,ta_sites,list(names).index('rffG'))])\n",
    "# np.save(savepath+'rffH_trajectories_green', counts_all_green[:,:,search_gene(locations,ta_sites,list(names).index('rffH'))])\n",
    "# np.save(savepath+'rffG_trajectories_green', counts_all_green[:,:,search_gene(locations,ta_sites,list(names).index('rffG'))])"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 14,
   "id": "5f34f8f9",
   "metadata": {},
   "outputs": [],
   "source": [
    "# np.save(savepath+'greensum', greensum)\n",
    "# np.save(savepath+'redsum', redsum)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 240,
   "id": "37097ea6",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "(14, 5, 20)"
      ]
     },
     "execution_count": 240,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "counts_all_red[:,:,search_gene(locations,ta_sites,list(names).index('rffH'))].shape"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "62d32ae7",
   "metadata": {},
   "source": [
    "### p-vals fdr calculation for estimating gene essentiality in LB"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 89,
   "id": "original-customs",
   "metadata": {},
   "outputs": [],
   "source": [
    "likelihood_noness_gamma = np.ones([len(names),len(libraries)])*-1\n",
    "#parameters for the fitting of gamma distribution:\n",
    "fit_domain = 0.3\n",
    "\n",
    "for k in range(len(libraries)):\n",
    "    #this is the data that will go into the fitting functions\n",
    "    data = fraction_t0[:,k]\n",
    "    a, loc, scale_noness = stats.gamma.fit(data[data>=fit_domain])\n",
    "    mu, var = stats.norm.fit(data[(data>=fit_domain)])\n",
    "    likelihood_noness_gamma[data>-1,k] = stats.gamma.cdf(data[data>-1],a,loc,scale_noness)\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 91,
   "id": "convinced-settlement",
   "metadata": {},
   "outputs": [],
   "source": [
    "pvals_fdr = np.ones([len(names),len(libraries)])*-1\n",
    "#printing all the preliminary results for essentiality classification\n",
    "for k in range(0, len(libraries)):\n",
    "#     print(libraries[k])\n",
    "    data = likelihood_noness_gamma[:,k]\n",
    "    pvals = data[(data>-1)]\n",
    "    genes_analyzed = np.where(data>-1)[0]\n",
    "    rejected,corrected_pvals,sidak,bonf = multipletests(pvals, alpha=0.05, method='fdr_bh', is_sorted=False, returnsorted=False)\n",
    "    pvals_fdr[genes_analyzed,k] = corrected_pvals\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 94,
   "id": "exterior-crawford",
   "metadata": {},
   "outputs": [],
   "source": [
    "fname = savepath+f'pvals_fdr_{site_thresh}_ta_min_{ta_min}.txt'\n",
    "# np.savetxt(fname, pvals_fdr)"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "62438403",
   "metadata": {},
   "source": [
    "## Fig S4:  excluding Ara+4 and Ara-2\n",
    "\n",
    "### Panel B: Making all the trajectories for REL606 and Ara-2\n",
    "(Doing this in the current notebook so I don't have to load all the trajectories all over in the final_analysis_plotting.ipynb)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 14,
   "id": "dd8e0d01",
   "metadata": {},
   "outputs": [],
   "source": [
    "import matplotlib as mpl"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 15,
   "id": "31d2301a",
   "metadata": {},
   "outputs": [],
   "source": [
    "anc_sum = np.zeros([len(names),5,2])\n",
    "aram2_sum = np.zeros([len(names),5,2])"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 18,
   "id": "2e5ee4b2",
   "metadata": {},
   "outputs": [],
   "source": [
    "for gene in range(len(names)):\n",
    "    search_area = search_gene(locations,ta_sites,gene)\n",
    "    anc_sum[gene,:,0] = np.sum(counts_all_green[0,:,search_area],axis=0)\n",
    "    anc_sum[gene,:,1] = np.sum(counts_all_red[0,:,search_area],axis=0)\n",
    "    aram2_sum[gene,:,0] = np.sum(counts_all_green[3,:,search_area],axis=0)\n",
    "    aram2_sum[gene,:,1] = np.sum(counts_all_red[3,:,search_area],axis=0)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 227,
   "id": "8938e8f0",
   "metadata": {},
   "outputs": [],
   "source": [
    "sns.set_theme()\n",
    "sns.set_context('paper')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 19,
   "id": "638f0dd0",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "image/png": "iVBORw0KGgoAAAANSUhEUgAAAVgAAACmCAYAAACfg5edAAAAOXRFWHRTb2Z0d2FyZQBNYXRwbG90bGliIHZlcnNpb24zLjUuMSwgaHR0cHM6Ly9tYXRwbG90bGliLm9yZy/YYfK9AAAACXBIWXMAAAsTAAALEwEAmpwYAABtEUlEQVR4nO39eZwtVXnvj7/XUFV76D59JuZzQEAmQxQxXqPGazQQlARBg3AigWAc8opRjEaiqMEJEAIkudGrXoi+DKiArxs0mFz1GxR/QdRoUFCUIByVeThzD3vvGtZavz9Wrerqpvt0N3SfsT6vV5/TvXcNa9de9alnPc/neR7hnHM0aNCgQYNFh9zZA2jQoEGDPRUNwTZo0KDBEqEh2AYNGjRYIjQE26BBgwZLhIZgGzRo0GCJ0BBsgwYNGiwRGoLdg5DnOb/1W7/FG9/4xp09lAa7MZZyHv34xz9m3bp1nHrqqZxyyin8y7/8y6KfY1eC3tkDaLB4+Pd//3eOPvpo7rrrLtavX8/hhx++s4fUYDfEUs0j5xznnXcel1xyCS960Yt47LHHePWrX81znvMcnvGMZyzKOXY1NBbsHoTrrruO3/md3+Hkk0/mn/7pnwD4z//8T9atW8f555/Paaedxu///u9z++23AzAxMcEFF1zASSedxMknn8zf/u3f0uSdNJhtHr3qVa9i3bp1nHLKKWRZxkUXXcRrX/taTj75ZF75yldW82o2ZFnGn//5n/OiF70IgP3335+VK1fy2GOPLfln2mlwDfYI3Hvvve7Xfu3X3ObNm92dd97pnv3sZ7vNmze7733ve+6YY45xP/vZz5xzzn360592Z511lnPOuUsuucS94x3vcEVRuDRN3VlnneW+973v7cyP0WAnY3vz6Oijj3YPPfSQc865H/7wh+5tb3ubM8Y455z7P//n/7g//dM/XdC5rr/+evfSl77U9fv9Rf8cuwoaF8Eeguuuu46XvexlrFixghUrVrBmzRq++MUvctxxx3HggQdyzDHHAPCsZz2LL33pSwB85zvf4YILLkAphVKKz33uczvzIzTYBbC9eXTAAQdw0EEHAfDc5z6XkZERrr/+eh588EH+8z//k263O+/zXHXVVVxzzTX84z/+I61Wa6k+zk5HQ7B7AHq9Hv/yL/9CHMe8/OUvB2B8fJzPfe5zHHvssVMmsBCicgNorRFCVO89+uijtFotVqxYsWM/QINdAnPNo06nU237rW99i4svvpjXv/71/M7v/A6HHXYYN910E1/72tf45Cc/OePxP/CBD3Dsscfynve8h/vuu4/rr7+eNWvW7JDPtrPQEOwegK985SssX76cr3/96yilABgdHeVlL3sZmzdvnnW/F77whXzpS1/iBS94AUVRcN555/GmN72J3/3d391RQ2+wC2Eh8+i2227jZS97Ga973esYDAZcffXVGGN4xStewSte8YpZz3HeeecxGAy4/vrrpxD2noomyLUH4LrrruP1r399dVMALFu2jLPPPpvPfvazs+731re+lSiKOPXUUznttNN46Utf2pDrXoyFzKN169bx/e9/n1NOOYVXv/rVrF27loceeghr7azH/9GPfsTXv/517r//fv7wD/+QU089lVNPPZVbb711qT7STodwrgkbN2jQoMFSoLFgGzRo0GCJ0BBsgwYNGiwRGoJt0KBBgyVCQ7ANGjRosERoCLZBgwYNlgh7FcG+4Q1v2NlDaLAHo5lfDaZjryLYLVu27OwhNNiD0cyvBtOxVxFsgwYNGuxIzItgrbUYY/iv//ovsixb6jE1aNCgwR6BOWsRXH755axdu5ZHHnmEn/70p6xevZrLLrtsR4ytQYMGDWaEtY6sMFNfnCEn1ZUvOlf+4LDW/zjclNf9/35jY8LhHNb58wkBq0ZadJJo3uOck2Bvv/12zj//fM4++2yuvfZa/viP/3jeB2/QoEGDxUaWG3712CgTgwJHjSDL38t/cA6scyWp+n2t86+DKLej3K+k4hqZOgdWOLAAEpzh2YevXlyCtdby4x//mDVr1pBl2XarMzVo0KDBUiLNC37+0Fbuf2KcltYEG1VQEqK1k0SLJ08pQSCwwiEcSCnK98s9RXkA61l10hB2TB7cgpQURbGg8c5JsKeeeiof+chHuOSSS7j88ss555xzFnSCBg0aNFgMDLKCu+/fwv0bxomlZKQbISR4+pyEw/Nh4E78JuDKv0W5vQSJQAiBkA7hBDIcT4CSojyG/1sAK4bbCxrznAR71llncdZZZwHwvve9b0EHb9CgQYPFwCAruOsXm3lg0ziRFCwfThjuxBUJSuFJECjJ0JOlUgqJt2KlkJ5Mg9Fa+x0hkKLctyTgcBzK16UQRFrNOsaZMCfBfvnLX+aqq64iTdPqtW984xuzbv/zn/+crVu3smrVqqaraYMGDZ42+mnOnes388jmHrESrF7e4cBVbfYd6SBkIFNREWIgyV0BcxLs1VdfzSc/+UkOOOCAWbfJsoyrrrqKr33ta6xatYrVq1czOjrK448/zsknn8y55567R/fdadCgwdJgop/xw3s38cRon5YW7LO8y5pVXfZf1UHKXYNEt4c5CXbt2rUccsgh293mwgsv5JRTTuEtb3kLUk5Ka51z/Md//AcXXnghf/M3f/P0R9ugQYO9BmMTKT+8bxOPb+vRjTX7ruxw8Ooh9l3R3mUs1LkwJ8G2Wi3e+MY3cswxx1Qf6p3vfOeUbS699NIZ9xVC8NKXvpSXvvSlizDUBg0a7C3YOp7yX/duYOPWHsvaMfuvHuKQfbqsGllYkGlnY06CXQg5Pv7441x++eVs2bKFk046iaOOOornPOc5T2uADRo02LuweXTAD+7ZwJbxPss7CQfs0+UZ+w0z0k129tAWjDlTZU855RR6vR4//vGPGR0d5fd+7/dm3fav//qv+YM/+AOyLOM3fuM3uPjiixd1sA0aNNiz8cTWCb77s8fZOj5geTdhzX5DPPOAZbslucI8CPbCCy/kwQcf5MUvfjEPP/ww73//+2fdNk1TXvjCFyKE4LDDDiNJds+L0qBBgx2PRzaM8d27n2C0N2DVSJuD9xvmmQcuo9uOd/bQnjLmdBHcf//9fP7znwfghBNOYN26dbNuG8cxt956K9Za7rjjDuJ4970wDRo02HF44LFRbl+/kf4gZ/9VQxy0ssNhBy5bsO50V8OcFmyapvT7fQAGgwHGmFm3/chHPsKNN97Ili1b+MxnPsMHP/jBRRtogwYN9kysf3gb/3XfRnppzpp9hjlkn2GeuWZkVnJ1zrFtPGXz6ICJQY4JhQZ2QcxpwZ5zzjmceuqpHHHEEdx333287W1vm3Xb/fffn7/7u79b1AE2aNBgz8U992/mzvu3kBUFh+23jANWdnnG/sOzyrCy3LBhW5+tExlSCJSAOFZ0Ek2iFUmkSGK1y8i45iTYV73qVfzP//k/efDBB1mzZg0rVqyYddtPfepT/OM//uOUpIJvf/vbizPSGfDd736XL33pSwwGA97ylrdw9NFHL9m5GjRosHhwznHXLzdx1wNbcMZxxIEjHLS6y0Grh2Ylx/F+xuNbekykhuXdmE4S0Utzstywucw0jSNJrBWtWJNEkjiSJNGcNLdkmPXMn/jEJ3jLW97CO9/5zid94CuvvHLGfb761a9y66230m7vGK1av9/nsssu4+677+bb3/52Q7ANGuwGsNZy5/qN3P3gNiSOw9cs55B9uuy3sjvL9o5NowM2bBsAjgNWdFi5LEFJibUJaV4wyCyDvCDNDBODnIlBgZKQxIpIq9K6lSSxQskd18hlVoJ9+ctfDrDdoNZ0HHTQQUuaEvvZz36W7373uwAcd9xx/Nmf/Rm9Xo9rr72Wd73rXUt23gYNGiwOrLX84J4nuO/hMSINhx+4gmfsNzRrAsEgK3hia48tEzlDiWK/FV2GaqoCKQXtJKKdAEySbZYb0tyQFYZ+ahBApH2xlkjLHeZOmJVgjzjiCLIs45prruHv/u7vcM5hreXNb34z11xzzYz75HnOKaecwpFHHlkNejZr96ng3HPP5dxzz63+3rJlC1dccQXnnXceq1atWrTzNGjQYPFhjOV7P32MXzw2TjtSPHPNMp6x/zDLh2Y2yraNpzy6pUeeW/Zd1mL18tacqoIk0oR62MZaBpkhzQxpYcgLyyArGKSScZkTa4VSgri0bmOtiKPFVS3MSrD//M//zKc+9Sk2btzIK17xCgCklDzvec+b9WBvetObnvJA7rzzTq644gquvfZarLV88IMf5J577iGOYy666KIZ6yF89KMfZfPmzVx55ZWccMIJ1TgbNGiwayHLC779k8d4aOMEnZbkiDUjHLb/MoY7T5ZyGmt5fHOfTWMDtJKs3XeIkW68YEtTSUm3Jem2IpxznmhzwyA3FMaSFYa8EAwGBqEg1gopxaK6E2Yl2DPOOIMzzjiDz3/+81U92LnwrGc9i6uvvpoNGzbw27/92xx11FHz2u/qq6/mpptuqny3N998M1mWccMNN3DHHXdw6aWX8slPfvJJ+zUFZBo02PWRZgX/8ZNHeWTjBCPdmCPXLufQ/Ydpz9B6pZ/mPLRxgt6gYHk3Zv9VnUUJUgkhaCWaVqIZAfKiJNvME621jiz3EtRBWqCVREpBpCRxVBJutHB3wpz0/K//+q/zPth73/te1q5dy69+9StWr1497wLdBx98MB/72Meqv2+//XZe8pKXAN7Xetddd817DNNxww038JrXvIbXvOY1Td/6BouOZn5tH2O9jG/+6BEeemKMlctaHH3wCp554MiTyNU5x6ZtA9Y/OkaaGQ5a3WXtvsNLpgCItGKoHbN6pMX+KzqsWpYw0k1oJZo4UhhrKYylnxaM9VI2j6U8tqXHIF3kljGdTodLLrmEQw89tCpFeOaZZ8647datWzn99NO56aabOP7443HOzbjddJx00kk89NBD1d/j4+MMDQ1VfyulKIoCrRd+sc8888xqvK95zWsWvH+DBttDM79mx7bxAf+/Hz/Khm191uwzxGEHLOPQ/YfRaqqfszCWhzdOsGU8pdvSHLSqQ6e147JApwbKmBIoy43FOYexBuHkgmvQzslYz33ucwHYtGnTvA64fv16AB577LEptWEXgqGhISYmJqq/rbVPiVwbNGiwc7Bh6wTf/ukTbBodcMi+wzzzgBEO2X/4SQQ13st4YMM4eWHZZ6TF/is7c/o9nXPYmvEWOnJNtox5eqqA7QXKfNbY/ANhc7LWW9/6Vr7zne/w0EMP8exnP5tDDz101m3f97738d73vpf169dz3nnn8YEPfGDeA6nj+OOP55ZbbuHkk0/mjjvu4Mgjj3xKx2nQoMGOx6Mbx/nOz55gy0Sfww9YzjMPXMZBq4emkKtzjsc293hi24BICQ49YBnLZgh4TUedXKUQte6xrvo9/LIYxFsPlD0VzEmwf/u3f8tjjz3G+vXriaKIq666ir/927+dcdujjjqKG2644SkNpI4TTzyR2267jXXr1uGc45JLLnnax2zQoMHS41ePbeN7d29gvJ9z9Frvb91vZWcKuaVZwf1PjNEbGF+ScN8uWs292rXW4XAIBIO0IC18DQJR9uwWQlRuSVtr2/0kT2Ug4DAkIaa/VfvbTfl9WTemFc9/NT3nlrfffjuf//znOfvss3n1q1/NddddN+u2L3nJS9i8eTMrVqxg69atxHHM6tWr+cAHPsCLX/zi7Z5nzZo1fPGLXwS8HOzDH/7wvD9EgwYNdj7ufXAr/3XvBvppwbOesZzD9x95UnuXTaMDHt7o3X9rVndZNdKa07KsW63OOjaPp4z1M6x1U8gRN2nFChyI8m8hPDk7fyzhwOLJ2JX7WQfWgnMWcBjnydlafwpfT8Zx5Jrli0uwxhjSNEUIgTFmu37V5z//+bz1rW/lsMMO44EHHuDjH/84f/7nf875558/J8E2aNBg94S1lv9+YAs/Wr+ZNC849vBVPHP/ZaxcNplAYIzlwScm2DqR0o4VB+83NKNMazrq5DoYFGwcS9m0rU9WGLRWWOvfDwRqS7J0bqrbwP9fkqx/BeFp2BMzIEOLbkDI4GCY7FQrBRTF7NUEZ8KcBPvHf/zHvOY1r2Hz5s289rWvnZJJNR2PPfYYhx12GOClV48++iiHHHIISu3eNR0bNHiqsLVSek816LsrwjnHeD9n87YBD24aZ/0jowjgOYet5vADlrF8eLLY/kQ/44ENE2S5Yd/lLQ5Y1Z2XPzS4BExh2bCtzwMbxtkylhJHknYS09XKE6IQyLJlt8P5tt1OIKUnSyUETjikVJ4opfffCuGJUwm/Z70FOM4hpABENVYhYGiBvtg5CfaVr3wlL3rRi7j//vtZu3btdqtp7bPPPlxxxRU897nP5Uc/+hGrV6/mtttuI4qemoN4V8ZcErT5StSWEvNZeu3sMezu55sLha1dYzt/60cyv8+xIzk76EK3jA14fGufTdsGbJnIsM7RiiXPWruKZxwwPCU767HNEzy2pY9WkkP3H2bZPFq/eFmUj95v2tbjF49NsGk8JZKw3/I2B6zqsnwoIdL+KklZkiWeBIMvVky7hmG+B2t28vdJny1A2E2UW9WDZAudX3MS7A9+8AP6/T7OOd75znfy9re/nVNOOWXGbf/mb/6GG264gf/4j//gyCOP5G1vexs/+9nPZg2K7WxY73R50utP9onvfLJsMDfCDbArWYp6Dt3kfGtF2yfNyoXt/1TgnCMrLIPUsGU8ZetYytaJjG3jAwrriKVg1bI2nZZm9bIWa1Z1GWp7YyorDPc/NsbEoGCkk3Dwfl3UHIEs5xz9tGDreMq2iZSN2zI2jPaRUnDIqg7POGAZI0MJUoCzU8kvwFjH5B08lxHk/w/EHMYgnKhcCS5sAFgsWi5sNT4nwV5++eVcccUVfOhDH+K6667jL/7iL2Yl2CRJOOecc6a8FnS0uzxq39T0W2IxrKL5HGM+RL6jLLRd6aEy01hme21XIleY2y3wVIY7aYnNfA3m+91VBFObU9a6MoXUMNbL2DqRMdrLfJGUzLP5so6PpI8MxQy1NMOd2Be8TjS5sYyOpzy8uQfA2jlabYcaAdsmMraOD+jlliwrGOsXGGc5ZN8uz9h3GSPDvhaBsw5jvZ9VAKL6DOX/tWPXb5Xp98327qPplm8Y51z7zYQ5CTZJElatWoXWmn322YcsyxZ0gl0ZUsopPrKlxGIR1q5EfAvCdkhhptdn226KwHyGyS6l3KXcBPX5NZ+p5r2OTxfz/PzlZrmZFNNPDHJGeznjg4LCGNK0wACRFKwaadHSkm5JqJ1E02351FIHmMLw8IYJtkzktBPNofsP0ZnBZzmFVCdS0txirCOJFLL8fpcPx+w30mbfFW20kjhryY0PaIlSkiWl8F0N5iHxerpzYqaH0XwwJ8EODQ3x+te/nte97nV8/vOf54ADDph121/84hdVkGt3gLV2hxHsTNgRZLkQi6aOmSZS/bXpVtT0rac/8Wcah5tGlkKIJ5HLfMdRP8+uhPFBvt33BSDF9gli0gfo/5czXpPwf30lNjMZOOcLmwwyy6As4zfWy+j1c/IyKp8XFmu97jOOFJ1I0Uo0SazpxJpuW1faVSFgvJfyqycmGBSW/UYSDljdBSEojEVJiRCQ5oZt4ynbejmDzOCAREtWDcfEsWTjlpTcOoY7MWv36dJpRVhjyHOL8RorT6pq0kNt8cQuKYNaM/hJgzxre5jvY02yMJKdk2D/1//6XzzwwAM885nP5Oc//zmvfe1rZ932fe9733Z1srsatmfBhvq3T+eGnS5SXsh+2yMjVz7Byz+efIC5JsDk4/jJr/Fk8oSpliPUCDVEWJnm8Zrt8047p5CTN0vYI1ih4frPNaHrN9WO6qYxb1hTjk3MeEkcYJgMftUfFFPJstx+hmNYN/Mcnv71+or/lrSw5IWll+akqSEtLEIIirKqFELSSjTtlvRVpJIIKQWtWNNpaYy1bBzte5LOLVlmmUhzOknEYfsOsWzIB7KEg17qLeLRvm/tIqWkFUn2HWkx1I5IYsmGrQMefGIC5wQHrGix34ou1lrSzGABZy1KOrRW4MAYr1P118WiJBgEhfOz0D+0JoNfM16baQ+fJ/09y5RbdAt206ZN3HLLLXzta1+rXnvrW98647YLKQyzKyDLMnq93pTXphPudCtrptenvzfTNlWGyXYs5vkQ8vZ8eoGQZiLn8LM9i3DKZ531LJPHC59F1MZUz6bBeSH4XOd8ssUxExPN8FCYhpDDXi8UtDORFYb/7/sPYK1FIIgiRVQ26ou0RCpBFGsiJbwkyLlyuSvKAI6/7ZUUgLcCAxG46vVJWGu91Kh8ubCGNLcUhSM3PqCbF5Y095arAIyz/tJa/Pi0oBMpMI5tmZ9PUgiEhNxAWlhcGRy2ZUApUpIVnRYH7D/MUDsmy3K2jGdsm8jJjMVZR6QUq4ZbLB+K6bYihBCM9zPuuX8Lvdwy1IpYu+8QkRKkufEEakFrh4411hiyzFA4hytD/iH11jiBlv66+gQD4ZMLHIhy/GHb7a3C/DWczTXl53qk9eJasG9/+9t54QtfuF3XQMBCC8PsbIyOjjIxMeEJw1qE9L4eSpKok+J0Ugm/TyenkJAhCBN+knxs2fI8kMpMaoXZyCZYdLMRuZelMDVYN5+JMI18jbVTLMp64EhOI9Lp455+fcK+ddKt/74961Rs73PMsJ8QAq01UqldhmBjrZgY9OkNCqyxpQi+FM67IIIXIBxCqErkLoVAl0QsPauWiheQWmKF13niQCrvh3Sl5SaEJ+o892FwhycHa70iwLmwraMwFmPBOBDCgXSA8q87SLRAK+VJXglfI1UIdKkjjZRCS18jdeOmHk9sG6ObxLS6bZLI+2ZHOprlQ3Hpi/UrxvFBzmObemwY9fKtA1e0WTXSIS8MuXHgJFp5d0CeW7I8xziHMQ5ZhvydtWBFdR0La3HWBw2ral2u1LIGsvW7lhrYaXN4hu9v+soKYIhiQbLTOQm22+3yjne8Y14HW0hhmF0BGzdu5PHHHtvZw2iwiIiThP33339nDwOALCv46f2P0JsYICR+3VoSoAOMBWtsGR0HnLdQLQ6Jd8tIIbHCebVBmRoqLAglvDUr/LLYAVJJTyrl+YvcYIqCLIOiJFJKeVNp0KIUqIiSuEBLaCWKJPIErrQkipQnNcAWBhEeEIKK1Ps5jHSXc8DqYdauhsMOHGb/5S2iSCGUJM0sxhRsGU15ZOsEg6xgRTfiwFVtIqUY76WAQGuBlJY8B4TzLgJT+IeCVBjrkMIBEmepXAjCCU/ChSPN8jL7yqGU8P5ZAU74FYLB64yFmPx/uougXpim7h6TcmFV/ebc+ogjjuDf/u3fOOaYYyrGn404F1IYZlfAdPdAg90fadm+eVdAVhhu+d44fcDbb1N91ar2WkDprkXgSdGW/tmwj63tM91ZpARo7TcaGCrPrsITqMX7RZWAOAKlS4IV3vKLtD//Noy3ft3keUVtEOUKvHpNWNAKokM3YnuOJ57IcbZgy1jOUCeh09JkuWHDtj7jEyntdswBKzqsGEowVmCBSHnL2BqLNV7P6qxfBSmtsAaEyJFCkBeWopRqeR4U5aoTwOAsFM6XGjTGeRLVvq40rkAiEbIkaahSYcNnktOs2zriaGF+/jkJ9u677+buu++u/hZCzNr0cCGFYXYFjI2O7uwhNFhkbNm8eWcPoUKnFRFG05/vTo659PHb33c20YKb9ntW/iwi9tto2bx2Awce3GUszdi6bcDQUEw7gn4BsdKMLG+xerhDrBRj/Rxd8lgfkGVtAeEcSioKYzzhIijzWLEuOEOo1QzwVq1/UMjSDRBI0hNxYRxp7slVKeuzwESZ7SXck/zb4F0JSoZgmSfghdalnnPra6+9dt4HW0hhmAYN9nQUswRM9lQ8Dmx5EH5tbILxgybYtmKc9tYuSitWjyxj/31iOlIx3hsghPfdSqXICosUoGRBJLxFa2yOkoIojlDOoiVoJYgiTay9ukFIUVXAksHHwWQSghSglMSV/ltjvfrAOocxFiV8hwKlJt0EgUidFVCzZIM9u2gEe9555/EP//AP/NZv/daT3vv2t7894z7TC8O8/vWvX9BgGjTYkzBXmuyeiAz40Vb4761wSHucfdeMs2a/YbpaMppIjJEMdSJaESRKEknL8o4iN4bCSQrnVQnLYkmkNEpJYu2TCSKlUEoiZE0BYAGhSsFKUK94cqxcMVIQx5MyOWO8sqIoA47GhoC1RCuvmAgQpU/blhGyolikINc//MM/ALOT6Ux4wQteUBWGWbNmDStXrpz3vg0a7GnYm1dwfeC/+zB+L2gzRmYycgf75QVtOjDUIcWRa0VWQKwj2kognEVGEXEkUThakfIBPSm9LK10Ysug9BEl2bqgegGvivUMa6wjDbIuvBoiOLklfofCBTWFoZfacltRVtZyKCk9ZwtBe7FdBAvBG97wBg4++GDOOOOMhlwbNGjAQ8DgF7B2a0r/gIfYus9KHhrrsk8yzv77jTDUitCtCCFjhJa02wnWgs0dRkOeW2Tms8Gc9ZpaBzipKne1dZO1YKuAf5n1JYRXEfi6sv4tWXMHqNK1EApxy1KqaB1YYREGXGFQZabYQilzUQn2S1/6Ej/5yU+48cYbufLKKznxxBP5sz/7s8U8RYMGDXYzbAQ2boZtY7BmYjOrVmym3+4yZjK6UYtVK9oMt2O0jmi1NDHQSjS6UBjj6BuDtV5lgFZoKYiFRQmHVoqoDGx5WbEPWjm8Pljg1RKutEjBq90sk4GwSd/rpKbdWodxjrwoMIW3YA1uwan1s65hLrjgAgCuv/76BR3wiCOO4LjjjmPFihX813/914L2fSrYtGlT0y65QYPdAL/I4Qe/gjvvhnsenuC+Xz3ALx97hPseeJSHH9/Gps3jPL61z2PbUh7d1OPhjRNsHO/jigItDVpCW1ja0pFogRQK64TPVCu8oM0KsEJiEaTG0Ssco33D2IRf/vdSSz815Jml1y+YmMgZ72VM9P3//X5GluU+ucg4tFBEyhEJ/7NoqbJ33303l112GV//+td55JFHprz3zne+c8Z9LrjgAu68805OOukkPvShD7FmzZoFDWahcM7xj//4jxx00EFLep4GDRosDvrALwbw+P1wyGZYvd84YyM9xrOcFe2ElUMdRrpDyDgi146Wi7CFJI4jpCiIS39spL2ESkmJddJrfA1QLvtFGfTSsrRYJeSF9em0UniLthQF+95cXqplpfQJGcZW1qeQCuHA2YVbsLMS7Mc//nF++MMf8q1vfWveGVknnngil1xyyYJZfr747Gc/y3e/+10AjjvuOEZGRjjllFP4zGc+syTna9Dg6SBbYP+mvQkTwC/GYOMY7L/SsrGzheUrYdnQECuGxli9eohV7RGKyELicJkDocm0J0aZ+iwzpRyxdCjlcE7irCNzoEu3ARDyEBBYcilwhQ9cSSlRZWadENJLuuxkGnNROnmdK3DOoRUIsUg+2DVr1rBmzRqe//znMz4+zn333ccznvEMjjnmmFkPduCBB3L66afz+OOPs3r1ai6++GJ+7dd+bUED2h7OPffcKT3B3vrWt/Lf//3f/OQnP+GrX/0qr3zlKxftXA0aPF3EuulFtz0Myp8tm+HgcTACemPj9FelbOplrGiPsmqoy9DQEN12TKeVkOW+vkeiFdnAECpnKCm85ap8+UepFFiHEz4wZoXAFK7yr7qylIEsC+tIKXCutGrL1FpZWrc+IcFjqLXImVw333wzX/nKV3jOc57Dpz/9aV75ylfyhje8YcZtL774Yi6++GKOPvpo7r77bj70oQ/N24d75513csUVV3DttddireWDH/wg99xzD3Ecc9FFF3HIIYc8aZ+Pf/zjALzrXe9qyLVBg90UObA+g8cehP01rO7lDI9sY6OAh4c20223WDU8wnCrxXC7jZKQxDGxslihfJGmstKLtMLXcHAOGcsquwt8gRgHOGu89rXMCSs7ciMj7xTQCLSUuEIQK1GpDxwsuIHrnAT7r//6r3zhC19Aa02e56xbt25WgnXOcfTRRwNwzDHHzDvr4eqrr+amm26qngw333wzWZZxww03cMcdd3DppZfyyU9+ctb9r7jiinmdp0GDBrsuJoD1BWx4HPbZCquWg8IxPtZndGJAK9GsbA8x3GnRilreL6slsQAdRwiXY42pyja6Ca8ekFLgpCCSkMQ+eUFphXK+uSLGZ3r1c4N00meCKV+pQEnpK5fhkEoyGAxYtmzZvD/TnAzonKuIMoqi7WYxaK255ZZb+I3f+A1+8IMfEMfxrNvWcfDBB/Oxj32Mv/qrvwJ8TYOXvOQlgPe13nXXXfM6zky44YYbuOGGGwDYsmXLlPd+84Uv5HulT7dBg6eC7c0vgN9S8CPjyaPB/DAKjKYw+jis2Ar7rAApHEU/x7g+W/s9lnWGiCOB1hHCCRIlSSLtCVWV1bMcSOVryTpbFnaZsESRQsoCLQVS+MIvFsD4AjFpnhFp5YNjhUOpUgLmwJjhBX2WOQn2ec97Hueddx7Pe97zuP3227fbxPDiiy/msssu48orr+Twww/nIx/5yLwGcdJJJ/HQQw9Vf4+Pj0+p6amUoiiKBecBgy/4HYp+T5dzCSE4/vjjMdby0IMP7jZ1bBvMDr2DW8Rvb34B/NNnzqTfH/Do5gGPb5tg27Y+mzePM5r2KPKM3MLY2DbGewNyB5nFr1kj6CYJwjkiBXluiKOYQa8sGxMJOjJC4bCuwMTaF882ljTzonqTex1mpKCTaApnaMcaqRVYS6wjsl4fq70n0jrhK3hZgcHXFnHWkhdlbr8FfOVCjPEFuPOy/OFP74HFLvy5AchSSDfAPjl0R0CNDVCxYKsdY3nSInMpUgpSJYh0RCIlsdIo7Yt6FyLF4ijK+gRYR5pSll8UREojtUQrhdI+8JW0Y7Qsaxx0NdZ694LeToeE2TAnY7373e/mW9/6FuvXr+c1r3kNv/3bvz3rtgcddFCVYvt0MDQ0xMTE5DPfWvuUyHU+EFIy1Omw7377gRD0ej2yLMMUxZTtpFJIIWi1WlVhblFGIo0xaK0ZDAZYa6vK+oUx5Lt5k0ilfMZMKPZdb7Pjyrqgs6FeXFsqVRUcD+/J8vq5sj6qVr4ASAj/yrKdjC0LeVtrKYrCF1uePFA1BiEEhx9++KJ+/sVAq5Wwdj/NsnbE2PIhNo602TraZ6yXMz5IGRnqoqygl6c45yiMYyIb4JyjIxSd4S6RUFjpSwNaYennFmtzbGHpdtooqVHaR86V8AGboigonMPYApxAKEekQQuNRNBWMbawvkpUS5fl/BxKKZQEWwZ7+nmOs75odyiUUuQ5hQzlAjVHHrSJb35vE/cscgXQbUDPQG8TtDfBvvtAZ9jRaWW+sL3WKKURhcClObGUxLGiozVaSqTWaKF8YRfhu0pQJgwIqZGRIC4TEmRhEFhfx0BKEAphBFqA1BqLJc+332NtOubFWr/927+9XWJdbBx//PHccsstnHzyydxxxx0ceeSRS3o+rTUjIyPgHBs2biSOY4qiYDAYEEcR/X4frXVFBJFSJN0uWimMMSitcdbSaberHlO2nKihupgut7HWlnUqja91qVTVliMgEEogs1BD1NmyuntRTGkpEiguVDCTZXHkLMsqMgp9qwI5ykBmTGawqPLzBPKr33CBKJVSVaeG0LWgeviV18daW1biL7NppnUnqMbpu9RVnRjC9XFlqxmc89Fg5yjK6xWOEcjdx3zLzxfGvItBCIGSiuUjLTptQydRLB/uMDo+zqaxFuMTA3qDlJaWSCMQ5HS7CUVhKTJBLy1Y1nLsM7KMWIGMIyggzwpSoUizAUpCpCK0kmSpIYkcSsd+CewsE1nZx0pAJAROCSKhQVhkWZJPI4lj/30prZAItFBo7XWmWEueF2SFpTBlWqr0Ra1XDrVI2ppVdzzOdx5e3OuX47PBVgNbx71/VAkwuSHpGF+jNfa9w5SKyHBIoSGK0EiUlCQ6RgpHpATGeRI1wrfqsQIwBhVpnBVliq3DuRwrhJ/zLFG5wp2BE088kdtuu41169bhnOOSSy5Z8nMmSUKn22Uky+j1+6SDAQB5njM0NISxljiKKEpCLXJf/FdHEQKIWq2K9PIsI9La6+tarUpzJ4QgKo8RiKQoioqAAjFVxZXLbay1VQQ0abU8qZbkZmoECp5kAzHWbcs8zyvyDKTr/UqezOI4RitVkXsgwerhULbVCcRY1Cz88LmDxVmHrD9ASrKtNzQMRKm0Lns/yclrUZ5resue0BYlEHH9nEu10nm6kFLgrKiCLEms6bQiRoZztoz1GJ3I2bxtgryfM7CKjtLoxJC3YDwV9ArDgxsmWDmcMGxg5fIh1EgXIRxpLyE1BikV7VgipCQz/iEonCOKFKvwc8Y4cKHOqXDgDAaJzY0X30tfLtDhx5vhyy5K7SPxSavFsJYI4YNBuS187j6aww/RLEsSkjsf4Jb1i3v9HN5lMNGH8T6sGoFVK3wgK1Up4xoiZSjsKJEAqQEl6SRtYq3RWqGkRse+zQ3WeqWAcxgryJ3v1CAkSCdQQoL0CgUFGOHndu/A+cWVAuacjRs3bmT16tXb3eYHP/jBrO89//nPn9dA1qxZwxe/+EXALw0//OEPz2u/xUS73faWJWWxXa2x1jIxMUG302EwGHiCE6IsLOF/ojiubnYp/CQMS19RLmlVSV5ZliGlJIqiJ5NVbTldkShlS4xAwrXzukDANQvY2MlGddZavzSvkXDwqwViDuQVyNEYA0ohkqTqK+ac7/4qS3eBEKL6zGGcwaKFya6w4XdZulSsm+zdFSxNrVTlEqiW+uWxQoM5Vf4fyLXqoRYeRoFoy4fBrgopfDuYSEuk1ERaEWtJK1KMtDOGuzHbJvqMbx2wLSvop45urFm7qs1YljLoCbb1+oyNa8ZSx4phzVAroTvcIikLSlvpy/oNaf9AM1hc4ci8EwZbFChtwUpkFBHpxPfwkr47q9YKYyzOSt8dAYsWAh1pcGVLF+E7IAgpEEIjECzv5Dy0GcS++/Dy5wtGuvdz849hfJGvYQ/vos63QTYGcQtaQ9BuQZRkoCDR3ncrjKXfn0BLiLUgimJQ4ISmo2NiJUhUhIiEr0MrBRKFVmUxbgXOGV8qURokasFdi+ecjW9729tYuXIlp59+Oi996UtnLMEWOhc88MAD5HnOr//6r/Ozn/2Mbre7oILdOxtKKbpd3zLYlkRrjWGo2yVNU9rtNv3BgG6nQ5L41sRFUaCVQkeR716pFHnuc5njKEKUQRellF+yQ0U4riSWdrtdWYB5aRlXy+rgqyzJpE6Wtlw2h4dCcDtUboWalSvwxBXHMba0ZIMlPaUpId4KDA+DOmkFYqz8pjWil1JOIe0plqzwd2R98R4s2kC8MEnMxhhfPakk/rrbQpQPMVmuHKhdK1H6yHdVCOFTOB2+Yn8SW5SMiSNJHEva7ZihWLO1HdMdS9m6pc8gdzwxnrGioxkaiegXit5EVrbczkl0xqp9LENRzHCiyRzk1rdUSZQi0RqrBAiDcgITKQonKIzD5AXWQUsKhlsxOonJi8zn38e+/xVOgYRIKnRUXmtRfjcCIuX9vlopkkjx8MYJlDyA/3Fsm6R9L9/6T8Pji3wdB3i3gbOQ9ED0oRXDihUw1IG4LeniEFpgnC/UYqUhV5blUQcplW+howSpsGgREUcRLSGIpTeetPTFuGMpcdLf1+glcBFcd911rF+/nv/7f/8vn/zkJ3nhC1/I6aefztq1a6ttQt+tN7/5zXziE59Aa40xhje/+c0LGsyuACkl7XbbL917PfI8xzlHnCQIIeh2u/T7fdp4t4IurVxnLQUQK0USx5V1ZfPc+zNLcgsEYq31vtTS7xhpjZWy1JaUy+RaEEmVX2z4gqWUiCia0UVQFEW19LY1wg3L6XpLDFEuwetBvcIYFFTjrSzwWmAvoO7XDWSLqHXWrVmtYnInv721uNKdEYJbodNs8KsG8hVC+OtVuhOCHzu4aQJhG2MWVBB5R6MWk0MKSRQ5lIyII00kc7QStNsJY1FKdziht23A5n5Kf0IghwraSeQzxCwMipyigEefGGe4pRhZNsTIUMRwEpHlhsJatHMMt+Ky2Z/3uythoWzB3UsNE8ZRFJa2G9COJKKjywec14IWVlA4S575zgGRKJBOQiQoXPnQcI44iVi73zDLWuNE0QqOU0fRjh/gtu+Pc+8it0oz+Hncw/9iUsgfg7EEnthgacewbNgxNGTRERRKQgEb8wFKOu9O0Zo40mTW0c/6OAuRN3HRsUYLsDisBOkckYrZunUr++6777zHOS863nfffVm7di0//elP+fnPf87FF1/MMcccw9vf/vYp223YsGHyAhjD5l2oP9J8IaVEa0271arIwtZ8fVJK4ijyZcyspdPp0K5ZTUVREEcR1jmSkpRNGT0PZBRHUbW0zcqoZJZlfilMGSCT0i/9S2s3Ky1bBxWpB0u3+inHF0WRVzOUkftAslVr7pKYqmi9ELgkqazf1rSHgJWycg2Eh4mquQvCMr/uP64HvIL/OBBsCFjJMvBXt4bDd1D5n0uidqVLQcBkJNf5FMi8KKrPE0XRbmTFSqwFqSASINqRr+AfWSIlaQ8U4yqik7XYum2Mfga9fkYrkbgopouAliW1grxv2Li5x0QfVi4bphOD1pLUSAZjOcMtxVDLV/U3VoDWtLSl0/bfwSBzpBayDDpYIgmRVqhIoXEoobDOkBeCNHcYHLpwaFcgrBfyC+cLpiwbGSJp5yRa0mrFaH5J64fb+Mkii4GD6ljgCbcHbEpBp5AA3S2QdGDFELQSS6ItSRtoC6SD1FnSLINIEFvo6MQnzjqLFY5cC4rCIGwZu9ALb4M1J8G+/e1v59577+VVr3oVl19+Ofvttx/gNX/TCfb000/n937v9zjyyCO57777eNvb3ragwewq0FoTJ0nZYM1/gYUxlVWWJAl5UVDkeSXN6rTb6Cjyy+By+W6MIdLaW50lWVRR7tJ3GyzIIDyq/KdlAEzVov2UpBvIMwR7ghvBb+L/r1uWU47BpF+0itiXxBuUAyEYF5bvpigqsg9ugzrZFuX+RWldKuWDA9QUBnWLOigYgnujer/8P7gEvAJIVyZfeCjJ8hrXx1HV9ah9zl0VUoqyTYkvCm3LVrGRAJEolJE+4h0nRIkm7kniWDMx1mein2GkRNoCkUQMbEHLaZKuQRCxrd9nkI7Tbce024q28g/mJ0YFT4wJlrVi2rEjKhwDYf33JQRDHcUyAYPMMCgcmXXIPCfKIdEaIR3tJEJrRzuSGByZMeS5L1RtTOFrBEQSjKMdRxy4sksrkshjDkNFDzD035v47iNzXp4FwzHV15vgu/iOWv/G1nGIgeEWtIYhjh2tKKPV9dMryQUDlTBwGdb4EJ92BVHqZWhDQ12UE76b7wLn15wEe8YZZ3DcccfR7XZ54oknqtdn6hh71llnceqpp/KLX/xit28Zo0tiDL5Wsszf/EVBFMcVGaWDAVJK75tVCl0GZIKPNM9zkiRBloWBgxY0+ExV6VcMS/BQNFhA5YcMvs8sz731agzWmClEVQXVSksyBIp0acVWx69Zi3X3xvRAk5QSJ7xVEkg7HD+QejheCJzVA2qVSyD4TEtJlio/a+U2CEGrEGirWfwBdXIOY6wrKKIoqo61Kwe56ggPUe8RKpv3CYi1RAnn+0gVBi0kWipaaU5LaVqdAWNjGYVRWGlo5WAUGBmTWMGqZTGOmDzLydMCq2JQEmVztJZsGzdsQ9LuWNqRROb+Go5JRaQErUjQjrxbqDCOwkgGWUqkJP3cEqHQkXdVdLVGJGCcYCItyHODMT65wUlJFCtWjnQRQpCog4mVQqkn+PaDS3tt0/IHoIO/11JgYgB64BMvkhi6GuIYhpc5VDSgFYGMSg+dtRjp7+fxfo4VjrYWU9Qz88Gcs/FHP/oRt956K+95z3u46KKLOPbYY3nzm988STw13HvvvXzgAx9gbGyMU045hSOOOIKXvexlCxrQrgJZ+kyZdvMPrCXPMuIkIYoitFL0+v1KcjU8PEwrSTyZaI0wxku6SvLVWiOkJM+yaomvSmsvWG3WestCloQbrDsdJE9Kle0zfIGLsLSH0ldpLaLc1xRFZeGqGgFWPmCoZF0hoBR8oYEYXY1k62QbSC+4JKY/PILFaaeNL/hbRXnuOhmL2jmn+2zDuOvWbiD78N7uYMEC1YPKOocSoiJZr58WgERE0BaSKDL0JKhYocahk7QYHZ9gkEmcdsSxpUhhICyRjdBYtBZESYwqo+BECeBoK9+nKksLilwSSYFOFMoY+n3LuJRE2lvMsZJIa+kkCbkpyDJDZgsiG9HLUhSSOFFo4YicT1YYpAVKCaRzKCWRzrJi+RBCS47Vvr+Wco/z3YcWvWv4jOiVP+DJrg1gwPU96SYxTOTQHYKO8i4FraA9JJHSIVxOnuZ0WzFSLNy3PyfBfvOb3+TGG28EfCPEdevWzRq8uuiii/joRz/K+9//fk4//XTe+MY37rYEC55MlPYtLAR4EgTyoiDLMh/kiiI6wGAwAOd8Blrpf7XWVkvtYOVZ8FZhIJaSEOIk8SqDmt9VlMGxgKrYr3NQyr6C9RmIq+6CgElZF3g/L0w2jAsWZx4CceXfFanWfKpBhRCOX2lRS8Lz2T+yspyDD7cu2arrXuukGc5XfcZybHUyD9eq0vgGa3aaBby7WLBQt2KDi6Z0HZQkK50ELM4oRoYV8SBHC8EgzYEuHZszPlaQO4NqKUYkpL0MIzQOSdobEBkYWq4ZiqRPGRUC7QSqMEjhcEpS5A4rHDryEq2ssOR5ykAppBa0yImEpRNp3zXAWWxuEQoGfR/wcrFE557YrXG4Ml5LpGgZw4pui0hK3OEH+8wq/Qg//pVPINhRKICx8idAZ/5nZBS6HWiNQRRBtCkjbkGkodWG8X5BrBT9fn9B55xzNgohyLKMOI6riPr2cMghhyCEYOXKlXS73QUNZldDIB1TFEjlq+/E5U0hSxeA1hoRxyAEaZpijaHf72OsrdQIuvRzAThjJiP35e+qlHbVl+zGGFwZDBIwY2Q8+ISCaiDsa8vlv5I+xxqoCChYeTYQfiBUa8nLwJkJ1ilMEfK7UqEQyHZKqmtJmnVJlSwt88ptwKQUa7q1WQ9wAU/aZoplWhJtndyD+2Cu+bkrIQS8ghUL3i1kyu9QSUkECOH9tZ2Wz9SSQiB1QdoHvUwxkaXkKVjjGBrqYAFrCgojKfKCLWM54yon0ZpuSyCVJhcC6wTKgsYgghBfFGQWbG58FlMu2JamSKVIZEq3EyGsX4EUxncCkNJiehm50mgr0QryXBBpAVlBIRVSOpZ1YpIDVoIALRU6fpCf/Bwe3YnfQVH+DKAydzt4P+4Q3pptaWh1LHHXMj6+MGXvnAS7bt06TjnlFI488kh+8Ytf8MY3vnHWbUdGRrj++uvp9/v827/924LKeu2qCFZsIIcsy4ijiEGaesIqySwQR5b5HOngAkiSpAr+1K05rVRFDEWpmw3JDSFCH6xeAPK8IrWQ/RV8rUmSVBZjcB3kee5TAcWkUF8pVS3nXc0POr22gC01utOtzyCdCtuE4FLdsp3ur5WlVVtPWqjSfGvWZ7nTlOyzuo+2Olf4O/i5S19sIPTdxUUQMD3gBZOWrbG+rkUsoSgsxlkirVg+LOkPJD2l6KcFsZYMtKFfWFxRIJxDR5qWEuSJpJUassKS2hzb18TKkCTe/5sXksIJXJ7TaVlipelGAhnFFEjywiDxmuhBFDGxNaWtodvSKKEwQGEcuRVErsDhi6No5chzSRQJpLDkhW8iWNiCtauHURLQkji+n5/8N9y/MNfmkiK4FbaAlycYiFJobYH773+Y3/qt+R9rToJ97Wtfy+/8zu/w4IMPsnbt2u0Gri655BI+9alPsWLFCu666y4uvvji+Y9kF0VYdgZy0EphtaYtpV8ulDe/kpJ2mRab5TmFMYg8R5W+V5vnRFH0pGi9MQZdWrAhiBbILPgnQ+aVZNI3HCRMRVFUpBaOr4qismqNMV7XWp4vJAqEZbaojaUeAKtIvBY8C3IrhECLSW2rLc9DjTwrpULp07XluerL+2A9h+s8nVCBKQqH+ndSJ+qwfXDJ7G6oB7zCg0cSalJ44tXa6ziFdjjj6HZitCrQUtIf+O81LgwTucAaS1EIcmFKJUaEVI6Ws5isIJUWl0e+5Qr4ojFS0i9gIvOugySWtCR0tMRKTdJ2ZLkBY8kKQdYriCJBWzti4du39FKDy1KcVCgsrdi7KrSSdGJJljuckGTOsGpZB7d2HwpjQD1EZz3cvdhpX4uIvPx5cIEBujkJ9u677+aGG24gTSeVwh/96Edn3LbdbvP7v//7la/vgQceYPny5Qsb0S6IYPnlee5rDwhBnuc+s6vfrzKogrwrZEkFn2q32618mkIIXOl2COmyIqSLBguu/L/ankmZWJB+qfDDZAZXUCVoravIetgnFK8JqoYgHYPJ9NpAZKE6WMgGC77aOkHW/aYhQysQcJW4UCYE1LO/6qQNU4NWlXVf1j6gvBZBIoabTMmtuxDq1vBCdYq7AqYHvMJrkyRLRbK2TGhx1vnW1loQKcFEmiOkIFKKwhrSzDKWGYTTCGdQ+Cr+DoW1kFtDmg2IdYTUikhpTJ6BkggL/Z4hixyRFnTimHYkaCUxWRyTFQV5VpCnGXkmkFqQKIHG4LRXn2QG0r4hyXPacYvMOGIJrdjXNBDOsXxZl6MO3h/pQPAw8SOOH2+AXdnJs+gE+573vIc/+qM/Yv/995/zYG9+85vJsoyRkZHqJghtXXZlzKfWbFh+BsJRpc+x1WrRL1UEwSKLkwRZBo4GpYwrSRKK0loNVmtIHqgvseuBIAFVERRlTPWQK0pdKs5VVmsYW15atMHyDlawMYbBYIApCvrWl3+Mo2hKMCmcv/5ZRUn2lTVdC8rULVigKuMYEhzqQTyBr7FQnUOIKfUL6scOOt9w3WEyYYH68aZrf5/6FNjpmB7wgplJVkqJpiRZ59BKMtRVJJFktJ/RSw2J0kTK0W1HjPVT8kiDdVhjMS1DkVoGeY5Dkw4sUhgynYMGlZW+8khgTUFWQJYXDOKYVuTrJrQTSaZ9fyyDpSgMvdTLwHAQOUdLCDIL/dRRmAEtLbBxRBRplLRYrYlNzshQwqEH74PFIdUTJHHO3Q9PDUTtSrhzgVb2nAS7evVqXvva187rYGma8rnPfW5hI9gFMJ/ASCCsPMsqi86VBOeszwippEXOoctsrlaSVJlH7VYLK8tOlmIybdWFiLsQla83LK9tmeEVukkUReHdCWUgxJVEHsYYLExrDFnpt4zimCiKfDAty3zJQ2NIS6s0WLuqLDITbvSgXaXuo4UpViMANcu1sk5rpB/2tbUKXRU51+Rb9e8jz/NKtla5VUpXTeWuKI8dZHDOOWTN7bA7YaaAV3h9JpINflkrHMI64kQzIiWtqGB8UBAJ/xDTsSLPDekgAyGxTmNjyzI6pHmKyS1ZYTAOFBrjDGlRIAt/XiJJ5BTjRY9+WZ1qqK2IJMRtRWEkmZIUhcNYg7OOHIVDkCQOYRW93DDILUMIjHHESnjCVxJpItpxwoH7rMRQoNmCcxkPPMKi1zBYDCw0IDcnwR500EFcddVVHHPMMdVN8FuzeHl/4zd+g1tvvXVK0eMDDzxwgUPa8Zhv5FlKWZFrVOa/F8bQarUq69EEP6fwJQCLPK/0qKFQjAFff1KpqvhKEXSj1Ipbl9ZuURTVsj1Yw0FeVX6AKXn8db+sNYa8zAhL4ti7MEo1QbAyszStygTWaxXUFQN+yJMpv3VLt77ErZNw+BxBnB06UwTCDEVnqs8MT7Kmg7tBCIEs/blVQkEovyhElXSwO6kIpmOmgBfMTLLgU2GLwpa58pDEClmmuA4yQ2ENLeHQVkHLp0IXucFpTURBK2qRG1vOW4kwBmcNhY3JbOk6KiyFK8pMQ8tEf4KtPcFQpFg21CJSijjSQI4wytdaNQaEL5kYCclIS9NPc/ppQSEdWeTHKaxFR4IVKsZag3GryfICY7aQRJbofnhohuu0O2FOgs3znF/+8pf88pe/rF6bjWA3bdrEJZdcUqkHhBDz7iq7O6CyYoNcTUyK5gPphawpISWiTBgA7yA3RUGv36fdbk/Kqpwvyxes03qGUyDyEMCp0lTxy22tlA8S4P2fIlhxpQWc574JnCzPXeR55SMOBG6NwYVAUY3AqqWqm0zZrS/FpwfCAoLv1kzzj4aAXL0gti1dENNlV3XJlhQCFcpB1s4Zyj7Wz7G76WBnwvSAV0CwcB1PJllrffUsHGgl0VqglSA3ktxYtHLEUjA+niIib0XmQpMoX/9VK4UzBX2nsSoidpYEH3AttMYKgbUFJiuwNqcYaDYOLGMDQxJpIiGJWvjSf1gyvBZWCsu2zCGdQAjf9saogoSISEdoCamBNIduJ8E5S9pfhSkcRb6ZNWtAPwy/2n2fmXMT7Ec/+lF++ctf8sADD3DUUUdtt5LML3/5S7761a8u6gB3NQRdrAWiMhhjnZdhtdptxsfHKfIcrRSR1hUBgifZIs8ZCEFSdk0wRYGJIl+zIIooSjlWQFhWhwpSgXiBKooeJFo6LKnL/UOh72rMzsttVCkRC6RUBEuw5lsNn1UIgWbSMgyujHppwmBB16P7wY8agn9OSlRJkKEouKxZxvWAVSDZujIBav7YmnshPASCVR6K3eyumCngFVCl1DKVgKWURMKRGwt4azdJNJFx5GFlZKG9QtIbFOSFZTAoQEjakUBpSVpEtHEUaUYvgwKFUt4XW9blxrVbuAKMLchNTn8io8hyVBKRWIVNyrlnykQJoejGjiwP8jzHeL+gn1tGuhYRJ7SU832yCsdwt43ZBwprKHBkdgv2ANCPw31m+pXaPTAnwX7uc5/j3//939m2bRuvfvWruf/++7nwwgtn3PbII4/kjjvu4FnPelb12nw7y+5OUKXsypTBIltG3qMootPpMDExQVomZ4QEjUBmwf+ZlzKq3Dm0MRRaE5dyr+n9vkL7mbrUKhwvJCiEGrQ4h83zSk87peKWc1WJxBCMEkJMitvxyzZgipX4pCQFpgWYnNds1jsw4BxFLRkhbBskb0HjWy9/OMVyDdZyjXCD9R7Sh+v62PCw2RVbxiwUMwW8qvdqKbX14iNCCGKtvMsAB8YHBJWUGONwwtc4XdaJSAuH1oJBajH4ZoZt5XBSkug27U5BkUOGIx8YBrnBSoXNDIkqUDoiibpEDDBSkhcFaZZjnUEjUErgjK+jGglBHGty40gEKA3jvZyN2xxDSU6n02aopSG1OCfYZ7iNVprCOTKTIeUErRhaG+GuXVjGNRvmJNh/+7d/4wtf+ALnnHMOf/zHf8wf/MEfzLrtD37wA771rW9Vfwsh+MY3vrEoA92VUEXmyzz/ejAmKcv+pYOB920yqU+1pbUZOhaEClp52WcryzLa7ba3jKctl0Nxl0oXWiOS4JOsNLPTC29D5RMOrWMq90Qc+2OFG7m0nupLe0qruS7kry/NhRClnnJS7hUKlk/ZpxxLyC4LATVbPqhmItNAwCEoFqzZuq81yMPq6bm7M2YLeAXMRrIw6TKwVUaywJX1B6x/MpJEAiUFWhXkmSHNLU5JIudACqSLiLuCjjEUscJaQZblbOsVpE5jsxylB+hEIXNfeDtLFM5YrJReTWAEopSDoXwha6EjrItItGLbtpzRrMC4AdbEJEqSAsZJRjqaZ65dXcrQNrLxsc1IB78u4O4xn3m1u2BOgp2uUdyeRfqVr3xlkYY1P9x3331cc801SCl53etet+TNEeuQUvov2rkpUqgQTDKl2D9Ygbr0ZeVB/1paKUWpAnDl71npF42iaEoL6ikSqbB8rhFcHMf+/KW2NpQ7LHcGpZBAq92mKEm23viwLhMzpTWslKraxKhSUVCfD8ECrVfpqshNCN/zqHbN6u+HmgLT6SMQclEUlesh+Jyn+IZL1Ek37L8nYLaAV/392UhWSumDYrZMZXWeUK3zUi0nQElBK1LESqIjwyAzZE4iC4eOwBqwTiKkQ+EYHooZHorJsoItY75ASjHw9fycksRCEUvIBWSZI46hFSkv80oNVkiSWCAQdGJJtDJha0/S7xt6gx7LuhGdSJE6QZYL2q0W+61cTpZa8txQ2G2sBJ4dw32bYHSHfRNPD3MS7O///u9z1lln8cgjj/CmN72JE044YdZtr7/++iclJfy///f/FmekM+C6665j33335YknnuCggw5asvPMhODrC5H8QFCBuNqdDr2JCS+HKtvNSCmrZoiyrE8Q/LRFnmOkJB0MKEp/a8KktVf5OIOvsrZkrwv0gaofF1AFyEITw6IoqmZ4piim5PsHv2goKB78qbJ0IdTrCITjB2u57qOt1AM1WVYY85QEgZprICgfnPM5+FIIRJnhVifR4BoJx6m3nNnTMFvAq3q/VhxmplqlUvoC25ayZ5sBoUs9M2VxGePoRopO4uilBWlWlp4sz29s2QMrs2iliOOIA/eNSdOcXj+iZwzjYwNya2h1FZEBhCHLJXlmiCLpW38XBeMDSxL7RAcdaRIsoivpjxu29XL6kWM4FmgV+XtGOw7cdxlaOZTUbHh8E1LDEQIe3AixV3tVxbbBV8jaEVW65os5CfaP/uiPeOELX8jPf/5zDj30UI4++uhZt73mmmu46qqrfAvsJcBnP/tZvvvd7wJw3HHH8fDDD/O2t72Nn/3sZ3z5y1/mrLPOWpLzzgYhxGRlKiF8hauygEsgu4nxcQpjGKSpT6Utl/8CqmV0iKAnYrJgTBoCV1r7FjSlPzbUL1Cl5RwsyZCuK0pLU+Lb1wio/LMhAUDUMquC6yH4RpVSyLKkYhXAYtItMl23WncL1IMudSVEeC0s+4NVWreGKwu5XO5Tji0Qat1yrdwfgbTLceZlwsieYsVuL+AVMKU4zAzbBImXc75zgjVlIFM6rBVI6SgKQDi6iUYrQS+1FM5LrFotfy2LwlE4R5pbtBMkcUSrJRjKLW3h2JqDEw4nClTkmysOjCUbFCQ6YjiR2ERRWIhljrFgVIwzBd2hFtkgpz/IcEYTJxbhFBZBrAXLOkMcstKghOWJx7eQD8HBwOp9wRa+C8O2cd/ocNCHvPAB5RRPvqPsvOywOQm2nom1fv16br75Zt761rfOuO1RRx3FAQccsGSBhnPPPZdzzz23+vt973sfnU6nyhzb0QjR/TzPMUEepZRf9peJASHoFXysgcza7bavWRDIT2uE9e1aQiAsy7JKEhakXJWPsfRzVlZgCDCV2WGh/UxdwiRKi8iWRBR8skEPG7SpKnRhwBNoXcdaWY01KzZci+nL9/rfdSIO8qsQmAp+3kqFUG4XCo5X5yit4ur4ZWZX8M/Gcbzby7SmY3sBr4BAsqE4zHT4OeDzRYR0GFMqEfybCGFLS9WTbCsW9NLc1381DiW8xjZ2jlwJcuv7eSXKdzBwQ21ia9HW0bcxvX4OQiGsd5MNCosR0NKKREuEaNESEiEKBrkizb0LKMJhTUZ/wj84IymJY+i0FcaNsF9hSQuDEKP0Wt6yljF029BKwDgwAkwPhAWjoZfCxIS3avMx6APb8AT8VLB8gdvPK5ML/M3ys5/9bMpNNR2/+Zu/yQknnMDatWurCXHNNdfMayB33nknV1xxBddeey3WWj74wQ9yzz33EMcxF110EYcccsiT9lm3bh3vf//70Vpz/vnnz+s8i42q71UpX9K1Ii5SSnTZI2qQpuRFgSiKqsNBEkivrBcgK82hrQJAoS2NCXULaiRknUO6yboAQSNqS6KtZE92stpUsFTD95iUdWjDEl1K6SsylYGz4NYoaiQsygSI6dKqeqlCqHWOrRFrIGfK94IrIFi7UxQF06xWM23uBSs71FWYieR3d8wV8AqYXoELJlcIrtTPVseUgJUIYREIdKJ9Rpex9HNHN5EMt2L6qiBLCyyCtLBEWpDEXu2SG69h1YXFmAKcpN1N0Lkh0aXW1kSMZ5Y0zSjSnNTCWDGg2xIImSCE9L2/WooidvTTiEHqGBSWInekoiB2Ed12RBRL4mglsZY8Eids27YJpTWJ8vdEscxntkkB2TKLUlA4hysMubFIoRj0DVoBFnoZjG2F8RxGR2EzZcnCObB27k2mYF7lCuvYXrnCG264gb//+79neHh4QYO4+uqruemmm6qe4zfffDNZlnHDDTdwxx13cOmll/LJT37ySfv9+q//On/zN3+zoHMtBaRSGGsrn2O9PUvV28r6dNq8TOucUly79KFV7oUoQmldWY5FUZCXZRDjkPZaCyCFDKspIn47Wbg6FNRuldZz0KLmYTxysiZtvWh2aANTpe7ayfTUkIVWdxsEtUQge1cSaBhPUAVMUUiUgay65RWOafEJEqJ2TeuEG6RrdeLeU9wDdcwV8IKp2V7TH0QAAm/FhhWPtQ5jS4UJDiJJS8EgNYynhm6ivMtAKtI8pzC+LKGzvkNCK45oO8d4mpMZiZAWZyxJJIijhKxwRNbQUTkTOmY8ikl7E2ipGZ2wCNEjijXtSNOOE1qRItK5dwmU1vZgYBjkA8YmJENJTKelyIeXMbCWjBFWdLogJDESoX19ZA0MigKJv29QCiV801CrNYOBvz65GcCBljTLGducsbWPb/q4DXoD2GpgK5NKn4DDVi/su5uTYOsZXBs2bODRR2fPxt1vv/349V//9QVP8oMPPpiPfexj/NVf/RUAt99+Oy95yUsA72u96667FnS8Om644QZuuOEGALZs2fKUj7M91ANcMBmIqWrCtlpefO8caZqSpilJ2XIm+FXDfpXWtPxfa02WZb4eZ1kNK4oi2u32FDnUdNSDSECVCRaIufLBWp+KGjoSWOfQUKkHgtg/BKiAKWRWl4hNzzwCqqSEQIK1DXzfpxppVj5bNzXtNlyLuh84EHkI4AWrfEfrYHfE/AJPkNMTDJ60Tc3f6v+efVspBUJICmOxZXKCk5Ik8STbSw2dRNGKJVJGFMaWrcAFzlkK/IqrHWuccQyswEpJYR0SS6TAZY5UaVpCEYmcCdXF2QJbFEzYiLRvyJxDygEis8SthCTR5MYSKUskBTIfIu2ljPczWrGgNRSz3A4z3Gr5z4BDauW7MziJcY4o9wE55RxOWfK+JdOWNM+JYt/q3KoOeWEQUYtWAgcpS9rvM2Ey8syyaRNkFtIxeGwMQr/sY35tYd/bnARbTypIkqQiwZmQZRmnnnoqRxxxRPXFXnnllXMO4qSTTuKhhyazjsfHxxkaGqr+DjnsT8W/duaZZ3LmmWcCvhPuUqISypeuAkrpVL0odiCJIJUKf9c7v1aR+RD9b7UqC9OU1udgMKiyv1StZoGqWZzhuCGbqrL0apKyQIAhm4vgapCyIlwhBKKuWKj5RusugXq6bbCeg1+6nhhQWdfCJy2EegjBJxtkWWHc1ecuXRchiaJqLV4r5ThTr7ilxI6aX8GKLQ3+WRH8rfOBEIKolpwgylbirQQGqaU3KGgnmjiSSOnHkGc5qRXe6s0NMpYorRhWgpb2BJsV3tXg+29ZeoUn464sSIuEAZquydBtGB/k9EYBBJ3M0ooFSaQwQqOUYFUimNCCbb0BhfPysnYnoRg4OkmCcY5BP8fiGFhLhEBJhZaSPCvIUodVApyvfIdNcALSomBESoQrcM5QOEFvqEt7okevcLRGUvoTfcQKx8oBmML7cg87dL8FfW9zMta1114774P96Z/+6YJOPhuGhoZ8b6sSlZ9tF0awnPI8r2RDYVlblH7XTqdDmqZVVN8F67Gs0hXF8ZT207KWPBDHMZHW9AcDrzJIU0xRoI2hU7pWggUYyKfevyovkxgog2lhe0rLUDBJ7KFKly79vsF1IIP1Wy79p+tfBVQVr1zNwgyQoixgU5NYBd9paMBoSl9w6BQR9LCOycpjwRVT1ScoVw5SPLkj7Z6EuWRbTxVaS4yxGJ9n4Ek2FqS5oZ8VtGOFVj4AquMIURiskxQGMN41kCOJtUNLgYwlifbdEpyzJFqRGYexkuHIEmWGicw3ZOzGkkKDdTnOwdjAMSgEcZQinKIvBFoKlnUScmswOSSxY0Omyb2gl9awIhKwQnhVhLMKnAVVasuNwGSmlP9BISzORYDEWciMYyLLWFZYVNIiMyn9rE0/Gca6jJHBAGkVG7aMs2LFioVd27k2eNWrXsXExARJklT61rAcnJ6l9cgji9P0/Pjjj+eWW27h5JNP5o477tihCQRPB5U1FyxHqEoEhpTWOI5994NOpyKINMs82eT5ZGCntILDUh3nsFLSarW8hVn6ZIMvN4oi4ijyxbnzvJJCWWMqn1wIVNX9mNXyvBSnB9IrrK0UDCHQVdVzDUGsUjqGc5VVWVcNQK1rrBAUNR9uXacb1BhSCAo5WXKxKDXGpgwgFvXaBCUBK62Jg6qifBDtqZhvwOupQCmJsK6sd+GLxoCXZfUyQ6wsWisoHCrWmNygtSDLfKEZi8UYSYFFC0mkBFEExnpZYVRYBoOcgZHoKGZl5OjlEqwmz8Hbnt46HxsU9HuWJAYtJSrSxFoijMAKkCgOiDRCOrTURFJggDQtKISkGORkOFQuiXSHViKIlznywmBwuMI//CeyAmNypNQMxxabWJyRCKMYanv533ieMegaTJYzNNQh0gubX3MS7HOf+1xOO+00nvvc53LPPffw6U9/mosuumjGbdevXw/4G/fuu+9m+fLlnHbaaQsaEMCJJ57Ibbfdxrp163DOcckllyz4GDsD9WCRDX5NpRBxTF4Uk00Soaq+BZOWpyutsaAJNbXltiij+8454jKYZIqiknSFZbfS2re1Ka3IkNEViD4EsKYU0C61p6EKVtCght/Bu39CEZU4iqbIqkKlLOcmK4xVJRfFZFpvUEYAlQshXDf/kpdbmZoiIbgFKAN0pkzKqLtadBRVpQ53hh92R2I+Aa+nc2yQWGdB+LbbkXOkmZc1CXw3DescTvvMQJ341YcpFyph7hYOMM77gR0IKWi1E9rOMN73RNeJvV9ZUJRVtQQSw1Ai6SZtJlKLkJZESpyEBEcufJbZUCTpdjR5AWNZwaBXgDG+5oIWJLRwyhArSWotvdRhC4GKJFI7TOFoRRHEikQpnOp4jTBl0RwrKFzGkGuRFjm5ERTjGcuGF9ZncE6CXb9+Pc997nMBr3N99NFHZ7US/vIv/7L63Tm3IJfBmjVr+OIXvwj4G+XDH/7wvPfdlRBkW1McYSXpVFInpaoi1/WKUKEbQSC2UA8W56DczhQFtgwsAWXXUVG1hIniGJkklYQrBKQE3vIMEfwqc6pUCtQVApWEqgzUhYBUcGeEZXu9iEuwKh1MCXpJratAXN0PGyx1hy/ULmoPEmCyWlYV9Z7U4yZx7LPfSndAXmakOefodDpL+fXuEphPwOupQkoBVk5pI9RKBKZwGMAVhkh7CzU3DikcrVgjrdfKermeo7AOZx1S+nmiEF4eJiNGpCIzvtB3OihI2jE6L5DCMJpCOoBW4ugk0OvDBAUdrUiSGJnljKa5r4/QtyAkwkG3G2EKiUEQCT+PB5llkAtyLFr69N2sMAxy3wlipBOhpSCKNFqBVoLCKgalwZIVEVnuiDO/kkqVXnCltjkJdnh4mL//+7/n2c9+Nrfffvt2C2iHXlzgFQf1wNXeBFkG5YCqR5YQPksrWGb1BICQyaXL1NtQgDu4GoAqGUBHUeV3DBF8nO+yWlhLnmVTLOUojiuiCj21gla26ooQrESoSDcqO+lOKQhuJwvJBP+yVJMts0NyQNDehrKFQY1QSbyk79tUKQmCSyME0yjblJfbG2P8QwuqVjlA5S8O+4OP5Aft9p6K+Qa8ngp8Gu3k95lE5QNfCGzh6xjkhSWKFFo4ClcSrRTgyuYXQqCEQETgnMVar0nFOSRlcFVJtLXQiihyL6cabiuGEkkvt2wdL5ACpLLkhWC0AFUYWlrQbiVM9CdwqSWJ9GQShZC0dURmfEnG3Phl/lCicVKRF751eewMQigKBMZIMuvwz3GDkgYpJImCWGt0Iunl3mrfXLrjFoI5CfbKK6/kC1/4ArfeeitHHXUU73jHO2bd9hWveEUln2m1WrzhDW9Y0GD2FNRlWw6q+gFxFDHo96v36umlgfziKMJqjSpJZVBad0JKtJyseSqLorKIcY6k1fJlD4vCZ/WUiQ2J9UXA4ziuZFLVUl7KyYaFIaof5FeUlnFZn1YLUVmzYTzB5zpFvlVG8/OyyA1JUpFjUCYE/3A4R6vV8sqGMtAlhC+KExIkwD+0wvmDbzg8KCp3CuzywdDFwlIEvEJxGPA+WFcmKGjlg0aFlpjcIrSgyC06kkTWkUpB7sDhGzAqNaksMcY3WzTGGwWWMrXXOCKlUMLgijKV1YGME4ZVQUsrtg0MmZHEIicvCgYDAS1FTMFwu02BwGHD44BYgcUQK4USjm4cIYQmd5YiL2jFEUooWokiNRZjLEpQ1nGwFNbXBmnpCKmhMGCdYChyCJNTDLdZIL/OTbBJkjAyMkKv1+PQQw9ldHR01tbd3/zmN2tf1p4p+l4wyqW0g6pOQEibDXn508sAgrfgbBngqrKk8MVfrPPVsYSUk0kDZZHvkACQ57nP0CpVDf3BwPf1Kgko1K8NFmFlXZaKBlsSLkxtaqhKd4QrCTGQbeUmKLWYouZ7rcu08tKyr1QAeOVFPSHBhP3lZHGaIHcLLpZQJzfIz4wxPhBW7E7F7J46FjPgFVYC4N0PwbcbLEN/naVvBRNJTOEQ0mELi9S+LbcSkERe/eEo281YX4wFIXwmFgohfMZVYRy5tYCiM6Qpcq+JztIcqQTtVkwrcaRZTr/QjI4OaHUkeV4wkApVONpKIrSqXFrGOj9+HO1W4lUOzmBySbfTIhLeF9xJImLluz4YL4tBlNfA/0g0oJUnfuMgMxFZFNNqLexaz0sHu++++/Kd73yHY489lne/+91cffXVM2771a9+tRK0X3755bzhDW/Y663YvFaYu8o6Kt0GdU3qlIh+7Yapp6LWb4SQzz9d/B+CTkopup0O1jlGR0crjWheFFUdgEGZ8BBIP0iypmRfyckiLfVzh0CTVApTBpqCZhW8u0iWbo6wXyDUuFZTQYBPlWVaZS4hyErfaghehWOEpAghJVmaVlY1QdWwl2AxAl7Baq0yvaaRdb0sopfqQYEjLyyFcETGByaLMkoapHLGiLKKF0i8n1QpX91La4nW0HIKIQVFYbBKYYwlVZI0K8iMQ4uSaE1OrLtM9AtiJVEKxvuO1DhWtbW3tnG+FkFRILVPVhDCqwyGWpIokt5fiwMHTgqQmpZ0KK2qjC3nLEVmyawldQItBLGGdsvfb53OwlQEc5qYDzzwAG9/+9uJ45iXv/zljI3N3lD3M5/5DC960Yu46aab+Na3vsUtt9yyoMHsaZjiY3WTBTuCxSpLi1OVEfJ6sRMb/I7lPoGcBZPBolDvoNvt0ul2iZOk8p3iJlMmkzhmxfLlvo1NSYRSSt+LqZRj1WvZhg62oRNtVMqhAtFXPtLyMyqtMUXBoGxfHggSfADKlBZvUBeEpICi7FGWDgaMjY0x0et5q7smP3POVVlvrVaLTrtdJVr0+/3qgd7v9ej1+3u0gmAmTAa8FrZ2DbUdXFnK0Gd2bSfrC1EpTbQWtGKJQJBZS5FblCwry1lA+L5g7VjTijWRlggJhXUY58B5X71S3mXkCxkJkkQx1I4Y7kS0IonFt5tBxwy1IzqxIndeobCqo1izsoXSMJoZtqaW1ICMY4SSREoylGg6LU2SREghiUqfcqS9xZ0Zw8A4+llBkRuM8V0VdKLotGNaWuKw5A4GZffdha6Q5rRgjTFs3rwZIQTj4+PbXfaHLJput0scx1OSBfZG1GVbdZIMfbamo3Kr+Md8ZaXW652GTKg8yxBloCovrclgaTomO8IGV4G1lk63W0Xws1KXW5eKhXoHwSKsxlhG96tOCGXmV5ZlaKXQpevBAflgQNJqIYWoCDqoH8Iyvh4MDcE6Wx7Tysk+YiGINzo6OpnYwGRlrqDEkEIQt1qkWbZgIfjujoUGvGZzB8znPLbsB0Ypu0oiSV5YJozF9cvUViF8soIug7Pl2HzNgzA7oTC1vnLUxi58rdiWE0hlyPJSjSAcK5e1UFrw2LYBW9OMpJcjtW8hvrwdo5TPlLAIhBJopVHSW9HW+Y4iAFJ5l0DkBIU1OLwFHJefSwJKgZaKxJXF63OH1HLBD7I5CfYd73gHf/iHf8iGDRs488wzed/73jfrtmvWrOEP/uAP+Ou//ms+/vGP8+xnP3tBg9kTEbKpQm59sOSmWwtV6ilM5v3XSCW8F9JgQ7HvoDqohP1lQkCSJOR5PuVBF7aNk6Qq1FIlAZRL++CjDecNQTrpBzmloHZ4eFAbV8jkGgwGFMbQ7XanaGSzmhYXqNwRnU6ncknYGpEbN9n1IPhyg543WPDWWvLBgKIo9goVwXTMN+BVD2LJGebgnOcpyRxRapZlWV1L5fRKzasPM4iyR9skGSkZUnjDeYPbi0pyBqCEROKQsUIWAoVlfJBhEUQaDljZYVlL8YvHt+GcY3W3Tasdg4O0MBS5QQqvKHC2wCqJVl4lEKkIIcqynGVUTVlVrvQEmQWF9Q8TgrFSukgk5YNskQn20Ucf5etf/zqbN29mxYoV2/1SLr30UiYmJuh2uxx77LHss88+CxrMHotSjhSi5tUyukRVK5VaD6vgjw3ZUyEYFXYqLbigOS3KRIa8toTpdrsYYxgZGaEoiiqlNi8KX1axJLo8z+nW9KNpWbc2EGVUpsxGZZUvoCphWE928MtHX6w7L/uVVYW+ywdAnCST8rPyuoQqY0XZTgf8yikt25TjHJ2SqAVl7dByHCEoF67P9spp7qmYK+D1VK3WmRBqz4bjhnoGLeVoJ/67s2VsIbgWxLTFWl0GFvy+dbKFUM/D3wddGTPRyxkdePXIyLIOx7Z9RmOrFZHnhvG0QBtHq6VASJQELSSoUNydchzepaIECCeQEQgRURjvyzYOnPFbaS3LB4MfX5YakmSRM7m++MUv8qpXvWpW5QD4QNgf/dEfceSRR9LtdgEqcr377ru57rrrdtvEgcVAsGLzkFdfI9jKkistwO1VSgqok23IlhJRhC0j+aaUaoX03G6369UFSeInZYjeZ1lVWQuo/KOBtFutFp1OpzoOzvcDC+NJkgQbRRX5FUVRybDarRZSKS/BCqm6eQ6lMgDh+4iNjY6iy/KNQgjSUJax9LmGGrdVEoFzdKCy0oObIxQn39483ZMxW8BrriDWU0EgWU/cFudCe5qaO0B4v+uT5i1U23nFQb1S7eR2wvlqX0kk0dISScnoRMq2Xo510Em86D/LDeMDHwxNkphYCyItkUKWfuWQDFm6JorQ8h2MdBi8flcI339Mlg+H1Ph08VhLkjhCa0mns/B28HMSbJZlnHbaaRx66KHVjTi9QtY73/lO/v7v/5677rqLQw89lNWrVzM6Osrdd9/Ns5/9bP7iL/5iwQPb41D6NIMlUUX87WTZv0kPFdt1qIWOBiHgBVSFVmLh8/ltmvqnbinVqle1otzPT9C8uglChhQwJTDXSpLK2q1H6at01tK3Ct690el0fAucVotWq0W/12Oi1/N+Vecqsp6YmKhqLSQlmbbb7UnrXfhyjK2y3OOUZa3wumJRWtahWtnejHqGF9RUH0/BHTAXqlY1ZU+woDSYfr7J+cZUKi2JT1Zx9rojLLwSHuYSKR3LZcKWsQEbxjLktj6RkhBpEg3dRJNEPqCm1JMNlckkl6nHtja4vQDrfbBKCgrjymI3BpUa2rGknzuUFvj8yflhToJ917veNedBli9fzgc/+EHGx8e588472bJlC6tWrapaujSYoRh2iSf1kAp3R21ZNx3TK1QBfrntXLWUl1L6TgjG0Ov3iaPIV+sKSoBSqhXHcTW2MI4pkVI32Z3WleOtt3/Js6wqbCOEoN/vM+j32bR5M1JKlod2PkJU6blFntNqtaak2rZbLX89ymLi4fNMLyZebzne7/er8wbVRJIkCy74vqeg3gQRnr47YC74IumloQClJVg2WnQzF/2GSdthKgnONM5JklbSa8AjJYjHMx7cWBArw0HDCcPdGKXkdgPwkxLIqceWwkGZVAGUDw2ItCOOFYM0Iyss28Z9Wu/Qqva8rk3ArAT7iU98gre85S38j//xP3jiiSfYd9995zzY0NAQL37xixc0gL0JVWHtmjtge5MivDOnXzG4DGqELKWvvBXy/PNyCR+sC+N81pgs/aDgi6zAtIkvBFEcV0v4cOygBAhtW0JVK2stvV7P+4StJW23p/TKSpKk0uECT8rtrlvaM12fug/bFEWlI+73fF/RNE33WoIFyiacQXu/dOQa4LWxoZklk+4IxCxEunDUyTGOI/ZdGYH10f6Vy5+eATedeKWsxTyAREuMcfQHOWO5pSgyoDXv489KsN/73vd4y1veAngrdr69tRpsH4EwFqLXnG9GXH2rEFyKoojx8XGAqnX4kzCDpVyXUkHNUi7RSpLKnRAIujAGrRTDQ0P0JiYqCzZMYqWnFsuYLa11e5HaeuJF3G5XZBsE7mqWY+4tCIGjHQklJdLN3HBxqdCKl1bvHOas1gqtIUk0IlK01CJlctUn+UKlCQ1mRwhm7YjzBAyVgcfQ82w+SFpTn9KhFnAdM8VTszJIJUrJVqvdrtQOTwXz3c863xUh3sEdDRp47AhreXfErLN3erpmg70b823FErbbtnUrURTt1cv1Bg1mJdif/vSnVcHr++67r/pdCMH111+/I8fYoEGDBrsEFmprzkqwN91009MdS4MGDRrs1ZiVYA866KAdOY4GDRo02OMg3F4UwXrBC17wpAfHli1bdniBkOacO/acK1as4NOf/vSSn7uZX3vvOWedY24vx6tf/ermnM05lwx7yzVozjkzmpYDDRo0aLBEaAi2QYMGDZYIez3Bnnnmmc05m3MuGfaWa9Ccc2bsVUGuBg0aNNiR2Ost2AYNGjRYKuy1BGut5cILL+TMM8/k7LPP5v77798h573zzjs5++yzd8i58jzn/PPP53Wvex2nn3463/jGN5b8nMYYLrjgAtatW8dZZ53FAw88sOTnDNi0aRMvfelLWb9+/Q4752zYWfMLmjm2VHgq82uvJdibb76ZLMu44YYb+Mu//EsuvfTSJT/n1Vdfzfvf//4ZC6csBW666SaWL1/OF77wBa6++mo+8pGPLPk5Qyfh66+/nvPOO4+PfvSjS35O8Df6hRdeSKs1/1JyS4mdMb+gmWNLhac6v/Zagr399tt5yUteAsBxxx3HXXfdteTnPPjgg/nYxz625OcJeMUrXsHb3/726u8d0dL6hBNOqG6yRx55ZIc1ILzssstYt27dvOoW7wjsjPkFzRxbKjzV+bXXEuz4+DhDQ0PV30qpBfc8XyhOOumkp1y276mg2+0yNDTE+Pg455133g5r3aO15t3vfjcf+chHOOmkk5b8fDfeeCMrV66sCG1XwM6YX9DMsaXA05lfey3BDg0NVe2swfvMduTE3FF49NFHOeecczj11FM55ZRTdth5L7vsMr7+9a/z13/91/TKbgNLhX/+53/mO9/5DmeffTZ333037373u9mwYcOSnnMu7C3zC/b8OfZ05tee+Y3PA8cffzy33HILJ598MnfccQdHHnnkzh7SomPjxo38yZ/8CRdeeCEvfOELd8g5v/zlL/P444/zp3/6p7TbbYQQS75s/PznP1/9fvbZZ/PBD35wp7eM3xvmF+wdc+zpzK+9lmBPPPFEbrvttqrO7SWXXLKzh7To+NSnPsXo6Cif+MQn+MQnPgH4IMhSBoJ+93d/lwsuuICzzjqLoih473vfO+9i3XsS9ob5Bc0cmwtNokGDBg0aLBH2Wh9sgwYNGiw1GoJt0KBBgyVCQ7ANGjRosERoCLZBgwYNlggNwTZo0KDBEmGvlWntKrj00kv56U9/yoYNGxgMBqxdu5YVK1Zw7LHH8pu/+Zs8+9nPXpTz/Mu//AudTocTTzyxei1NU175ylfyzW9+82kf/0Mf+hB//ud/vsNSYxvMD8382slY9KY1DZ4S/vmf/9ldfvnlS3LsiYkJ9yd/8idPen0wGLiXvexli3KO9evXu/e85z2LcqwGi49mfu0cNBbsLor3vOc9nHzyyWzcuJFbbrmFwWDAhg0bOOecc/jGN77Bvffey1/91V9xwgkn8NWvfpXPfvazSCl53vOex7ve9a4px/rKV77Ci1/8YgAmJiZ417vexejoKAcffHC1zfe//30+/vGPAzAYDLjsssv4/ve/z69+9Sve/e53Y4zhtNNO4wtf+ALnn38+4+PjDAYDzj//fF7wghdw2GGH8Ytf/GKndPpssHA082vHoPHB7gaYmJjg6quv5k1vehPXXXcdH//4x/nwhz/MjTfeyNatW/nYxz7GZz/7Wa677joef/xxbrvttin7f//73+eoo44C4Etf+hJHHnkkn//851m3bl21zb333svll1/ONddcw8tf/nK+9rWv8Xu/93t84xvfwBjDrbfeygte8AIee+wxNm7cyKc+9SmuvPJKBoNBdYzDDjuMH/7whzvmojRYNDTza+nQWLC7AY455hgAhoeHOfzwwxFCMDIyQpqmPPDAA2zevJk3v/nNgL9ZHnzwwSn7b9myhVWrVgF+ooeqQM95znOqAiT77bcfF198MZ1Oh8cff5zjjz+eoaEhnv/85/Ptb3+bG2+8kbe85S0cccQRnHXWWbzzne+kKIophZ332Wcftm7dutSXo8Eio5lfS4eGYHcDCCFmfW/NmjUccMABfOYznyGKIm688cbqhglYuXIlY2NjgLcC7rjjDk444QR+9rOfVSX03v/+93PzzTczNDTEu9/9blyZQX3GGWdw9dVXs2XLFo4++mjuueceJiYmuOqqq3jiiSdYt24dL3vZywDYtm1bdaM12H3QzK+lQ0OwuzlWrlzJueeey9lnn40xhoMOOohXvvKVU7Z5wQtewJ133snzn/98zjrrLC644AL+8A//kMMOO4woigA49dRTOeOMM1i2bBmrV6/miSeeALwVcv/993PWWWcB8IxnPIP//b//N1/+8peJoojzzjuvOs/dd9/9JP9cg90bzfx6mtjZUbYGS4/x8XF3zjnnPKV9jTHujDPOcGNjY9vd7t5773Xvfe97n9I5GuzeaObX7GiCXHsBut0up512Gl//+tcXtN+DDz7Iq1/9ak499dQp1flnwrXXXjuldUiDvQfN/JodTbnCBg0aNFgiNBZsgwYNGiwRGoJt0KBBgyVCQ7ANGjRosERoCLZBgwYNlggNwTZo0KDBEqEh2AYNGjRYIvz/ARcjNLfFxLVfAAAAAElFTkSuQmCC\n",
      "text/plain": [
       "<Figure size 360x144 with 2 Axes>"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    }
   ],
   "source": [
    "#making the plot:\n",
    "with sns.axes_style('ticks'):\n",
    "    fig,ax = plt.subplots(1,2,figsize=(5,2),sharey=True)\n",
    "    for gene in range(len(names)):\n",
    "        if anc_sum[gene,0,0]/greensum[0,0] >= 10**-4:\n",
    "            ax[0].plot(anc_sum[gene,:,0]/greensum[0], alpha=0.02, color=mpl.colors.hex2color('#c0bfbf'))\n",
    "            ax[0].plot(anc_sum[gene,:,1]/redsum[0], alpha=0.02,color=mpl.colors.hex2color('#c0bfbf'))\n",
    "        if aram2_sum[gene,0,0]/greensum[3,0] >= 10**-4:    \n",
    "            ax[1].plot(aram2_sum[gene,:,0]/greensum[3], alpha=0.01,color=mpl.colors.hex2color('#225ea8'))\n",
    "            ax[1].plot(aram2_sum[gene,:,1]/redsum[3], alpha=0.01,color=mpl.colors.hex2color('#225ea8'))\n",
    "            if aram2_sum[gene,4,0]/greensum[3,0] >= 10**-2.5:\n",
    "                ax[1].plot(aram2_sum[gene,:,0]/greensum[3], alpha=0.1,color=mpl.colors.hex2color('#225ea8'))\n",
    "                ax[1].plot(aram2_sum[gene,:,1]/redsum[3], alpha=0.1,color=mpl.colors.hex2color('#225ea8'))\n",
    "\n",
    "\n",
    "    ax[0].set(xticks=[0,1,2,3,4], xlabel='Time (days)', yscale='log',ylabel='Frequency of insertions\\n(summed over gene)', title='Anc')\n",
    "    ax[1].set(xticks=[0,1,2,3,4], xlabel='Time (days)', yscale='log', title='Ara2')\n",
    "    sns.despine()\n",
    "    fig.savefig(repo +'/Analysis/Plots_for_paper/trajectories_comp_bluegray.png', format='png', dpi=600, bbox_inches='tight')\n",
    "    plt.show()\n"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "bf862204",
   "metadata": {},
   "source": [
    "### Counting number of reads mapped with and without UMIs for each library"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 252,
   "id": "3ab96f12",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "['REL606',\n",
       " 'REL607',\n",
       " 'REL11330',\n",
       " 'REL11333',\n",
       " 'REL11364',\n",
       " 'REL11336',\n",
       " 'REL11339',\n",
       " 'REL11389',\n",
       " 'REL11392',\n",
       " 'REL11342',\n",
       " 'REL11345',\n",
       " 'REL11348',\n",
       " 'REL11367',\n",
       " 'REL11370']"
      ]
     },
     "execution_count": 252,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "libraries"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 253,
   "id": "1a50b14f",
   "metadata": {},
   "outputs": [],
   "source": [
    "order_readcounts = [2,3,5,6,9,10,11,4,12,13,7,8,0,1]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 255,
   "id": "b81852a4",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "(14, 5, 211995)"
      ]
     },
     "execution_count": 255,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "counts_all_green.shape"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "a5bb2855",
   "metadata": {},
   "source": [
    "### With UMIs\n",
    "\n",
    "#### I want to print the number of reads at every timepoint after processing, and after UMI correction to compare how many of our reads are actually used in the sequencing analysis"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 258,
   "id": "ef853c80",
   "metadata": {
    "scrolled": true
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "13486281.0\n",
      "17694625.0\n",
      "5133499.0\n",
      "16804732.0\n",
      "17619749.0\n",
      "15663983.0\n",
      "17874299.0\n",
      "19921062.0\n",
      "18603464.0\n",
      "12414855.0\n",
      "16597219.0\n",
      "42985069.0\n",
      "13055480.0\n",
      "15259340.0\n",
      "12646038.0\n",
      "14167768.0\n",
      "11923541.0\n",
      "14827004.0\n",
      "11307015.0\n",
      "16818436.0\n",
      "14904000.0\n",
      "19211907.0\n",
      "15849856.0\n",
      "14016416.0\n",
      "13755600.0\n",
      "15532521.0\n",
      "14150381.0\n",
      "10931386.0\n",
      "16439265.0\n",
      "13551304.0\n",
      "14190097.0\n",
      "17847663.0\n",
      "14376095.0\n",
      "14831844.0\n",
      "15499952.0\n",
      "14079835.0\n",
      "11687571.0\n",
      "16141495.0\n",
      "14380740.0\n",
      "13915157.0\n",
      "16296161.0\n",
      "17180928.0\n",
      "12730881.0\n",
      "16917776.0\n",
      "18411299.0\n",
      "10170080.0\n",
      "16194566.0\n",
      "15964627.0\n",
      "13394732.0\n",
      "13691101.0\n",
      "10105339.0\n",
      "20258996.0\n",
      "15344550.0\n",
      "15806226.0\n",
      "7890716.0\n",
      "20042959.0\n",
      "17705836.0\n",
      "15337935.0\n",
      "19625762.0\n",
      "10345061.0\n",
      "14150387.0\n",
      "17228960.0\n",
      "17055849.0\n",
      "11308160.0\n",
      "16626514.0\n",
      "14990406.0\n",
      "13240654.0\n",
      "15308771.0\n",
      "13666208.0\n",
      "19924923.0\n",
      "14220291.0\n",
      "13527508.0\n",
      "11883549.0\n",
      "21852240.0\n",
      "22256813.0\n",
      "19084482.0\n",
      "20377811.0\n",
      "17640486.0\n",
      "19796610.0\n",
      "20061950.0\n",
      "19623047.0\n",
      "4685148.0\n",
      "18825950.0\n",
      "17441450.0\n",
      "16982798.0\n",
      "24364592.0\n",
      "16992465.0\n",
      "18147717.0\n",
      "16816223.0\n",
      "15080123.0\n",
      "10019733.0\n",
      "26945352.0\n",
      "20870358.0\n",
      "20883963.0\n",
      "26208644.0\n",
      "18515764.0\n",
      "20048107.0\n",
      "18635793.0\n",
      "18020675.0\n",
      "13551337.0\n",
      "18418822.0\n",
      "17442558.0\n",
      "19545920.0\n",
      "20391839.0\n",
      "16994635.0\n",
      "14812117.0\n",
      "17768516.0\n",
      "18669903.0\n",
      "15071951.0\n",
      "16444999.0\n",
      "10926749.0\n",
      "12387024.0\n",
      "15528916.0\n",
      "14560311.0\n",
      "18645080.0\n",
      "14105362.0\n",
      "16380071.0\n",
      "14002401.0\n",
      "21143836.0\n",
      "11391849.0\n",
      "11110020.0\n",
      "12970273.0\n",
      "12663821.0\n",
      "19697626.0\n",
      "30983049.0\n",
      "25037708.0\n"
     ]
    }
   ],
   "source": [
    "#first, UMI corrected counts\n",
    "for pop in order_readcounts:\n",
    "    print(greensum[pop,0])\n",
    "    print(greensum[pop,1])\n",
    "    print(redsum[pop,1])\n",
    "    print(greensum[pop,2])\n",
    "    print(redsum[pop,2])\n",
    "    print(greensum[pop,3])\n",
    "    print(redsum[pop,3])\n",
    "    print(greensum[pop,4])\n",
    "    print(redsum[pop,4])\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 256,
   "id": "5e84afe5",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "array([15071951., 16444999., 12387024., 14560311., 14105362.])"
      ]
     },
     "execution_count": 256,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "np.sum(counts_all_green[0,:,:],axis=1)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 257,
   "id": "21b60c1b",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "array([[15071951., 16444999., 12387024., 14560311., 14105362.],\n",
       "       [14002401., 21143836., 11110020., 12663821., 30983049.],\n",
       "       [13486281., 17694625., 16804732., 15663983., 19921062.],\n",
       "       [12414855., 16597219., 13055480., 12646038., 11923541.],\n",
       "       [11308160., 16626514., 13240654., 13666208., 14220291.],\n",
       "       [11307015., 16818436., 19211907., 14016416., 15532521.],\n",
       "       [10931386., 16439265., 14190097., 14376095., 15499952.],\n",
       "       [10019733., 26945352., 20883963., 18515764., 18635793.],\n",
       "       [13551337., 18418822., 19545920., 16994635., 17768516.],\n",
       "       [11687571., 16141495., 13915157., 17180928., 16917776.],\n",
       "       [10170080., 16194566., 13394732., 10105339., 15344550.],\n",
       "       [ 7890716., 20042959., 15337935., 10345061., 17228960.],\n",
       "       [11883549., 21852240., 19084482., 17640486., 20061950.],\n",
       "       [ 4685148., 18825950., 16982798., 16992465., 16816223.]])"
      ]
     },
     "execution_count": 257,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "greensum"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "9853112f",
   "metadata": {},
   "source": [
    "### Without UMIs"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 259,
   "id": "b4d7e408",
   "metadata": {},
   "outputs": [],
   "source": [
    "#all the information from the fitness assay condensed into a couple of matrices\n",
    "counts_all_green_unc = np.zeros([len(libraries), 5, len(ta_sites)])\n",
    "counts_all_red_unc = np.zeros([len(libraries), 5, len(ta_sites)])\n",
    "\n",
    "for k in range(len(libraries)):\n",
    "    #loading the new file where I have the counts for each TA site for all time points\n",
    "    gname = directory+'green_'+libraries[k]+'_merged_all_TAsites.txt'\n",
    "    greendata = np.loadtxt(gname)\n",
    "    rname = directory+'red_'+libraries[k]+'_merged_all_TAsites.txt'\n",
    "    reddata = np.loadtxt(rname)\n",
    "    #now for extracting the UMI corrected coutns\n",
    "    counts_all_green_unc[k,:,:] = greendata[1:10:2,:]\n",
    "    counts_all_red_unc[k,:,:] = reddata[1:10:2,:]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 260,
   "id": "1d69bc59",
   "metadata": {},
   "outputs": [],
   "source": [
    "greensum_unc = np.sum(counts_all_green_unc,axis=2)\n",
    "redsum_unc = np.sum(counts_all_red_unc,axis=2)\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 261,
   "id": "18489a07",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "16036648.0\n",
      "21457324.0\n",
      "7454636.0\n",
      "20666011.0\n",
      "21884163.0\n",
      "20405469.0\n",
      "21423973.0\n",
      "24736298.0\n",
      "22698662.0\n",
      "13782304.0\n",
      "20521795.0\n",
      "52973495.0\n",
      "15628150.0\n",
      "18561220.0\n",
      "19115126.0\n",
      "19241617.0\n",
      "17515829.0\n",
      "20690618.0\n",
      "13262758.0\n",
      "20731516.0\n",
      "17064327.0\n",
      "25580190.0\n",
      "20227014.0\n",
      "17113378.0\n",
      "16599006.0\n",
      "18614253.0\n",
      "17040862.0\n",
      "12856547.0\n",
      "19882407.0\n",
      "16918427.0\n",
      "17533034.0\n",
      "22532722.0\n",
      "17533823.0\n",
      "18283477.0\n",
      "18029149.0\n",
      "16684039.0\n",
      "13654979.0\n",
      "18962524.0\n",
      "17617796.0\n",
      "17518070.0\n",
      "20245096.0\n",
      "21405335.0\n",
      "15107879.0\n",
      "20121697.0\n",
      "22989921.0\n",
      "12151265.0\n",
      "20064540.0\n",
      "19182593.0\n",
      "16624868.0\n",
      "19075316.0\n",
      "13546494.0\n",
      "26129163.0\n",
      "17764462.0\n",
      "19685847.0\n",
      "9637957.0\n",
      "30151748.0\n",
      "23149678.0\n",
      "20691306.0\n",
      "26072939.0\n",
      "12849259.0\n",
      "20115798.0\n",
      "22232771.0\n",
      "21995997.0\n",
      "12972727.0\n",
      "19994460.0\n",
      "17506622.0\n",
      "15886264.0\n",
      "18795651.0\n",
      "16621209.0\n",
      "24430186.0\n",
      "16929145.0\n",
      "16064049.0\n",
      "14478695.0\n",
      "27089326.0\n",
      "27936004.0\n",
      "24186300.0\n",
      "29024493.0\n",
      "21856104.0\n",
      "24718150.0\n",
      "24257872.0\n",
      "24234638.0\n",
      "5907163.0\n",
      "25319694.0\n",
      "22529343.0\n",
      "23485343.0\n",
      "34738245.0\n",
      "23489661.0\n",
      "24694884.0\n",
      "23603905.0\n",
      "19607032.0\n",
      "10218028.0\n",
      "28226637.0\n",
      "21505300.0\n",
      "21746000.0\n",
      "27459953.0\n",
      "19280257.0\n",
      "20862759.0\n",
      "19159720.0\n",
      "18571097.0\n",
      "15988813.0\n",
      "21907363.0\n",
      "19895528.0\n",
      "24430701.0\n",
      "24869302.0\n",
      "20632218.0\n",
      "16991370.0\n",
      "20130835.0\n",
      "22024464.0\n",
      "18453444.0\n",
      "20641391.0\n",
      "13680499.0\n",
      "16215957.0\n",
      "21896583.0\n",
      "18683725.0\n",
      "24123282.0\n",
      "16912009.0\n",
      "19483941.0\n",
      "15999782.0\n",
      "26971985.0\n",
      "14243905.0\n",
      "14329834.0\n",
      "18452597.0\n",
      "16114244.0\n",
      "24523548.0\n",
      "40659382.0\n",
      "32041631.0\n"
     ]
    }
   ],
   "source": [
    "#first, UMI corrected counts\n",
    "for pop in order_readcounts:\n",
    "    print(greensum_unc[pop,0])\n",
    "    print(greensum_unc[pop,1])\n",
    "    print(redsum_unc[pop,1])\n",
    "    print(greensum_unc[pop,2])\n",
    "    print(redsum_unc[pop,2])\n",
    "    print(greensum_unc[pop,3])\n",
    "    print(redsum_unc[pop,3])\n",
    "    print(greensum_unc[pop,4])\n",
    "    print(redsum_unc[pop,4])"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "035d70a7",
   "metadata": {},
   "outputs": [],
   "source": []
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3 (ipykernel)",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.10.4"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 5
}
