{
 "cells": [
  {
   "cell_type": "markdown",
   "id": "792dc44b",
   "metadata": {},
   "source": [
    "## Comparing fitness estimates with and without UMI correction\n",
    "\n",
    "Generating a supplementary figure in response to reviewer 2's comments on the companion paper."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 1,
   "id": "c852d17b",
   "metadata": {},
   "outputs": [],
   "source": [
    "import numpy as np\n",
    "from scipy import stats\n",
    "import matplotlib.pyplot as plt\n",
    "from Bio.SeqIO.FastaIO import SimpleFastaParser\n",
    "import pandas as pd\n",
    "import seaborn as sns"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "id": "adbf41cf",
   "metadata": {},
   "outputs": [],
   "source": [
    "import pathlib\n",
    "import os"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "id": "9cf07a43",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "/Users/anuraglimdi/Desktop/TnSeq_Paper/LTEE-TnSeq_Paper/Analysis/Part_3_TnSeq_analysis\n"
     ]
    }
   ],
   "source": [
    "#current working directory\n",
    "cwd = os.getcwd()\n",
    "print(cwd)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 10,
   "id": "11fc7e13",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "/Users/anuraglimdi/Desktop/TnSeq_Paper/LTEE-TnSeq_Paper\n"
     ]
    }
   ],
   "source": [
    "#use the pathlib.Path function to get the parent directories-> goal is to navigate to directory with the metadata\n",
    "# and the breseq output data\n",
    "path = pathlib.Path(cwd)\n",
    "repo = str(path.parents[1])\n",
    "print(path.parents[1]) #this should be the base directory for the github repository: the exact path will differ for \n",
    "#each unique user"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 13,
   "id": "afbbb149",
   "metadata": {},
   "outputs": [],
   "source": [
    "metadata_path = repo+'/Metadata/'\n",
    "#data_path = repo+'/Data/Mutant_Trajectories/'"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "bf371f16",
   "metadata": {},
   "source": [
    "### Loading the relevant metadata"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 14,
   "id": "3d375066",
   "metadata": {},
   "outputs": [],
   "source": [
    "#opening the pandas file with all the metadata!\n",
    "all_data = pd.read_csv(metadata_path+\"all_metadata_REL606.txt\", sep=\"\\t\")\n",
    "names = all_data.iloc[:,0]\n",
    "gene_start = all_data.iloc[:,3]\n",
    "gene_end = all_data.iloc[:,4]\n",
    "strand = all_data.iloc[:,5]\n",
    "locations = np.transpose(np.vstack([gene_start,gene_end,strand]))\n",
    "k12_tags = all_data.iloc[:,2]\n",
    "uniprot_rel606 = all_data.iloc[:,6]\n",
    "locations_pseudogenes = np.loadtxt(metadata_path+'pseudogenes_locations_REL606.txt')\n",
    "#list of genes to be excluded from analysis as they lie within large deletions\n",
    "exclude_genes = np.loadtxt(repo+\"/Analysis/Part_2_WGS_analysis/output/Deleted_genes_REL606_k12annotated.txt\")\n",
    "exclude_pseudogenes = np.loadtxt(repo+\"/Analysis/Part_2_WGS_analysis/output/Deleted_pseudogenes_REL606_k12annotated.txt\")\n",
    "n_pseudo = exclude_pseudogenes.shape[1]\n",
    "#fractions of the gene at the 5' and 3' ends to be excluded from analysis because they insertions there may not actually\n",
    "#be disruptive to protein function\n",
    "frac5p = 0.1\n",
    "frac3p = 0.25"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 15,
   "id": "77164fb8",
   "metadata": {},
   "outputs": [],
   "source": [
    "with open(metadata_path+\"rel606_reference.fasta\") as in_handle:\n",
    "    for title, seq in SimpleFastaParser(in_handle):\n",
    "        ta_sites = [m.start(0) for m in re.finditer('TA', seq)]\n",
    "ta_sites = np.array(ta_sites)\n",
    "\n",
    "#counting how many TA sites are present in each gene\n",
    "ta_gene = np.zeros(len(names))\n",
    "for i in range(0,len(names)):\n",
    "    start = locations[i, 0]\n",
    "    end = locations[i, 1]\n",
    "    length = end - start\n",
    "    #if the gene is on the forward strand\n",
    "    if locations[i,2]==1:\n",
    "        #counting sites only in the middle 80% of the gene, excluding 10% at each end\n",
    "        ta_gene[i] = np.sum((ta_sites > start+length*frac5p)&(ta_sites < end - length*frac3p))\n",
    "    elif locations[i,2]==-1:\n",
    "        ta_gene[i] = np.sum((ta_sites < start+length*frac5p)&(ta_sites > end - length*frac3p))"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "e356961b",
   "metadata": {},
   "source": [
    "### Loading the TnSeq data"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 24,
   "id": "e563837e",
   "metadata": {},
   "outputs": [],
   "source": [
    "directory = repo+'/Data/Mutant_Trajectories/'\n",
    "data_green = np.loadtxt(directory+'green_REL606_merged_all_TAsites.txt')\n",
    "data_red = np.loadtxt(directory+'red_REL606_merged_all_TAsites.txt')\n",
    "# with UMI correction\n",
    "counts_green_UMI = data_green[2:11:2,:]\n",
    "counts_green = data_green[1:10:2,:]\n",
    "counts_red_UMI = data_red[2:11:2,:]\n",
    "counts_red = data_red[1:10:2,:]"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "371fcc16",
   "metadata": {},
   "source": [
    "### Defining parameters for the analysis + any functions required"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 22,
   "id": "5809830e",
   "metadata": {},
   "outputs": [],
   "source": [
    "def search_gene(locations,ta_sites,i):\n",
    "    start = locations[i, 0]\n",
    "    end = locations[i, 1]\n",
    "    length = end - start\n",
    "    #if the gene is on the forward strand\n",
    "    if locations[i,2]==1:\n",
    "        search_area = (ta_sites > start+length*frac5p)&(ta_sites < end - length*frac3p)\n",
    "    #if the gene is on the reverse strand\n",
    "    elif locations[i,2]==-1:\n",
    "        search_area = (ta_sites < start+length*frac5p)&(ta_sites > end - length*frac3p)\n",
    "    return search_area\n",
    "\n",
    "#following the weights definition from \n",
    "def inverse_var_weight(n0,n1):\n",
    "    \n",
    "    return (((1+n0)**-1 + (1+n1)**-1)/np.log(2)**2)**-1"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 23,
   "id": "b8fb963c",
   "metadata": {},
   "outputs": [],
   "source": [
    "#calculating fitnesses\n",
    "def fitness_estimate(counts_red, counts_green, locs, genes_lost, initial_depth, min_sites, min_frac, max_reads, t_start, t_end):\n",
    "    \"\"\"\n",
    "    data:\n",
    "    - counts_red, counts_green: trajectories for both replicates\n",
    "    - exclude: filtering out genes that are lost/have too few sites from analysis as they're indicated as -1\n",
    "    params:\n",
    "    - locs: start and end points of genes\n",
    "    - initial_depth: exclude sites below a certain threshold in fitness estimation\n",
    "    - min_sites: minimum number of sites required for estimating fitness for each replicate\n",
    "    - min_frac: fraction of TA sites that need to considered in estimating fitness\n",
    "    - site_thresh: minimum number of reads for a trajectory to be included (strictly greater than)\n",
    "    - max_reads: maximum reads that can be used for inverse variance estimate\n",
    "    \n",
    "    \"\"\"\n",
    "    green = np.sum(counts_green,axis=1)\n",
    "    red = np.sum(counts_red, axis=1)\n",
    "    #fitness effect (best estimate)\n",
    "    fitness_inverse_var = np.zeros([len(genes_lost),2])-1\n",
    "    #three different metrics of measurement error\n",
    "    stderr_fitness = np.zeros([len(genes_lost)])-1\n",
    "    stderr_fitness_inverse_var = np.zeros([len(genes_lost)])-1\n",
    "    stderr_fitness_hybrid = np.zeros([len(genes_lost)])-1\n",
    "    #defining time interval based on gene essentiality status:\n",
    "    time_gens = np.linspace(t_start,t_end,t_end-t_start+1)*6.64\n",
    "    \n",
    "    #iterating over all genes that are not lost over evolution\n",
    "    for gene in np.where(genes_lost==0)[0] :\n",
    "        #TA sites within this gene\n",
    "        search_area = search_gene(locations=locs, ta_sites=ta_sites, i=gene)\n",
    "        #weights\n",
    "        weights_green = []\n",
    "        weights_red = []\n",
    "        #fitnesses\n",
    "        s_green = []\n",
    "        s_red = []\n",
    "        \n",
    "        ta = np.sum(search_area==True)\n",
    "        #now, iterating over sites with at least tm1_depth reads at t-1:\n",
    "        for j in np.where((search_area==True))[0]:\n",
    "            if (counts_green[t_start,j]/green[t_start]*10**7>initial_depth) & (counts_red[t_start,j]/red[t_start]*10**7>initial_depth):\n",
    "                #extracting the trajectories\n",
    "                traj_green = counts_green[t_start:t_end+1,j]/green[t_start:t_end+1]*10**7\n",
    "                traj_red = counts_red[t_start:t_end+1,j]/red[t_start:t_end+1]*10**7\n",
    "                n_0 = traj_green[0]\n",
    "                n_after_g = traj_green[1]\n",
    "                n_after_r = traj_red[1]\n",
    "                #this part is different:\n",
    "                #i'll find the first occurrence of zero if at all, and then fit a trajectory upto that point\n",
    "                #green replicate first\n",
    "                if 0 not in traj_green:\n",
    "                    sg = np.polyfit(time_gens, np.log(traj_green), 1)[0]\n",
    "                    wg = min(inverse_var_weight(n_0, n_after_g), inverse_var_weight(max_reads, max_reads))\n",
    "                    #sotring the data\n",
    "                    weights_green.append(wg)\n",
    "                    s_green.append(sg)\n",
    "                else:\n",
    "                    #first zero in the gren dataset:\n",
    "                    ind_g = int(np.where(traj_green==0)[0][0])\n",
    "                    counts_g = traj_green[:ind_g+1].copy()\n",
    "                    #add a pseudocount to the first occurrence of zero\n",
    "                    counts_g[-1] = 1\n",
    "                    sg = np.polyfit(time_gens[:ind_g+1], np.log(counts_g), 1)[0]\n",
    "                    wg = min(inverse_var_weight(n_0, n_after_g), inverse_var_weight(max_reads, max_reads))\n",
    "                    #storing the data\n",
    "                    weights_green.append(wg)\n",
    "                    s_green.append(sg)\n",
    "                \n",
    "                #now the red replicate\n",
    "                if 0 not in traj_red:\n",
    "                    sg = np.polyfit(time_gens, np.log(traj_red), 1)[0]\n",
    "                    wg = min(inverse_var_weight(n_0, n_after_r), inverse_var_weight(max_reads, max_reads))\n",
    "                    #sotring the data\n",
    "                    weights_red.append(wg)\n",
    "                    s_red.append(sg)\n",
    "                else:\n",
    "                    #first zero in the red dataset:\n",
    "                    ind_g = int(np.where(traj_red==0)[0][0])\n",
    "                    counts_g = traj_red[:ind_g+1].copy()\n",
    "                    #add a pseudocount to the first occurrence of zero\n",
    "                    counts_g[-1] = 1\n",
    "                    sg = np.polyfit(time_gens[:ind_g+1], np.log(counts_g), 1)[0]\n",
    "                    wg = min(inverse_var_weight(n_0, n_after_r), inverse_var_weight(max_reads, max_reads))\n",
    "                    #storing the data\n",
    "                    weights_red.append(wg)\n",
    "                    s_red.append(sg)\n",
    "                \n",
    "        #converting to numpy arrays\n",
    "        s_green = np.array(s_green)\n",
    "        s_red = np.array(s_red)\n",
    "        weights_green = np.array(weights_green)\n",
    "        weights_red = np.array(weights_red)\n",
    "        \n",
    "        #pooling over all sites, require minimum number of sites, and fraction of sites represented\n",
    "        if min(len(s_green),len(s_red))>=min_sites and min(len(s_green),len(s_red))/ta>=min_frac:\n",
    "\n",
    "            #weighing based on inverse variacne:\n",
    "            fitness_inverse_var[gene,0] = np.average(s_green, weights=weights_green)\n",
    "            fitness_inverse_var[gene,1] = np.average(s_red, weights=weights_red)\n",
    "            \n",
    "            #standard error in fitness estimates:\n",
    "            stderr_fitness[gene] = np.std(np.hstack([s_green,s_red]))/np.sqrt(len(s_green)+len(s_red)-1)\n",
    "            #-1 is for the unbiased estimator\n",
    "            \n",
    "            #inverse variance weighted standard error measurement\n",
    "            stacked_s = np.hstack([s_green,s_red])\n",
    "            stacked_w = np.hstack([weights_green,weights_red])\n",
    "            \n",
    "            average = np.average(stacked_s, weights=stacked_w)\n",
    "            sem_inv = np.sqrt(np.average((stacked_s-average)**2, weights=stacked_w)/(len(stacked_s)-1))\n",
    "            sem_hybrid = np.sqrt(np.average((stacked_s-average)**2)/(len(stacked_s)-1))\n",
    "            \n",
    "            #saving the inverse variance wieghted measurements.\n",
    "            stderr_fitness_inverse_var[gene] = sem_inv\n",
    "            #saving the hyrid measurement errors\n",
    "            stderr_fitness_hybrid[gene] = sem_hybrid\n",
    "            \n",
    "    return fitness_inverse_var, stderr_fitness, stderr_fitness_inverse_var, stderr_fitness_hybrid"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 25,
   "id": "a88ac37a",
   "metadata": {},
   "outputs": [],
   "source": [
    "#parameters for filtering out noisy data and selecting which time points to use in analysis\n",
    "min_sites = 2\n",
    "min_frac = 0.2\n",
    "#timepoints to compare:\n",
    "# change this depending on the analysis you want to do!\n",
    "t0 = 0\n",
    "t1 = 4\n",
    "time_range = np.linspace(t0,t1,t1-t0+1)*6.64\n",
    "tm1_depth = 5"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "c6930687",
   "metadata": {},
   "source": [
    "Running the analysis with and without UMI correction"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 27,
   "id": "d9887aba",
   "metadata": {},
   "outputs": [],
   "source": [
    "fitness_UMI, pop_error_UMI, pop_error_inv_UMI, pop_error_hybrid_UMI = fitness_estimate(counts_green_UMI, counts_red_UMI, \n",
    "                                                           genes_lost=exclude_genes[0,:], locs=locations, initial_depth=tm1_depth,\n",
    "                                                           min_frac=min_frac,max_reads=100,\n",
    "                                                           min_sites=min_sites, t_start=t0, t_end=t1)\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 28,
   "id": "ccfeb770",
   "metadata": {},
   "outputs": [],
   "source": [
    "fitness, pop_error, pop_error_inv, pop_error_hybrid = fitness_estimate(counts_green, counts_red, \n",
    "                                                           genes_lost=exclude_genes[0,:], locs=locations, initial_depth=tm1_depth,\n",
    "                                                           min_frac=min_frac,max_reads=100,\n",
    "                                                           min_sites=min_sites, t_start=t0, t_end=t1)\n"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "5b9e6f7d",
   "metadata": {},
   "source": [
    "### Looking at the correlation of fitness estimates"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 42,
   "id": "02e766e1",
   "metadata": {},
   "outputs": [],
   "source": [
    "sns.set_theme()\n",
    "sns.set_context('paper')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 43,
   "id": "a4abb9a1",
   "metadata": {},
   "outputs": [],
   "source": [
    "genes_comp = np.where((np.mean(fitness, axis=1)>-1)&(np.mean(fitness_UMI, axis=1)>-1))\n",
    "r2 = stats.pearsonr(np.mean(fitness[genes_comp], axis=1), np.mean(fitness_UMI[genes_comp], axis=1))[0]**2"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 48,
   "id": "756e9580",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "image/png": "iVBORw0KGgoAAAANSUhEUgAAAKQAAACmCAYAAABUfSRKAAAAOXRFWHRTb2Z0d2FyZQBNYXRwbG90bGliIHZlcnNpb24zLjUuMSwgaHR0cHM6Ly9tYXRwbG90bGliLm9yZy/YYfK9AAAACXBIWXMAAAsTAAALEwEAmpwYAAAgxklEQVR4nO2deXjM1/7HX5PJnqDZCJFGUImlUYIWRVNpcB91iyA0UVS1RVQvIqE3GsS+1FbLrUtRVGvJbWtJlFha6iaEIqJJbpGIJCKLbM0s398f+WUqMjOdRGYyw/f1PJ7HnPOd8/kk3s75nnM+53MkgiAIiIgYCWYN7YCIyKOIghQxKkRBihgVoiBFjApRkCJGhShIEaNCFKQWfvnlF8LDw6uVjRs3joyMDA4cOICXlxe///57tee9vLzIyMhQ+11DkpWVxZkzZwD4+OOPG8yP2iIK8glo1aoVsbGxqs9Hjx7Fzc2tAT36k3PnzpGUlATA6tWrG9aZWmDe0A6YMn5+fsTHxzNp0iSUSiUZGRk8//zzGp9XKBTMmzePmzdvolAomD9/Pm5uboSFhVFaWoqFhQULFiwAYPLkyTRq1Ijhw4ezY8cOXF1dadmyJePGjSMyMhKZTIaLiwvR0dFYWlrWaHfLli1UVFTQrVs35s6dy4kTJ7h69SrR0dFIpVJcXV1ZtGgR33//PfHx8ZSUlJCbm8uiRYvo1KmToX6FNRAFWQckEgkAdnZ2ODg4kJmZye3bt+nevTs///yzxu/FxcWhVCrZt28fN27cICkpie+++46AgAACAwM5f/48S5cuZfbs2eTn53PgwAHMzc1Zv349K1eupE2bNkybNo0PP/yQ7t27s2fPHnbt2oW7u3uNdidNmkRmZiY9e/ZU2f/0009Zvnw5np6erF27lq+//ho7OzuUSiVbt27l+++/Z9++fQ0qSHHI1oK1tTUymaxaWWlpKVZWVqrPAQEBxMXFcfToUQYNGqS1vdu3b+Pj4wOAt7c3QUFBpKWl0aVLFwB8fX1JTU0F4Pnnn8fc/M/+wtPTE4DU1FTWrFlDSEgIMTExZGdnq21XHfn5+ap2fH19+e233wBo164dAK6urlRUVOjwm9EfoiC10KZNG65evUphYSEAOTk5FBYW4uTkpHqmf//+nDhxguzsbNzd3bW216pVK3799VcA/ve//xEZGYmnpyeXLl0CICEhQTXkm5lV/6ep+uzp6cknn3zCzp07mTx5Mr169VLbrkQi4fEwhSZNmqgmYQkJCXh4eAB/9vjGgDhka8He3p6PPvqIcePGYW1tjSAILFiwoNo/oL29Pba2tvTo0eMv2/P39+fMmTOMGTMGpVJJZGQkrq6uhIeHc/DgQQRBIDo6WmsbYWFhREVFUVZWhlQqZenSpTRv3rxGuwqFgs2bN6t6Tqgcsqtm/k2bNmXp0qUcOXKkjr8d/SARo31EjAlxyBYxKkRBihgVoiBFjApRkCJGhShIEaPCZAXp5eXV0C6I6AGTFaTI04koSBGjQhSkiFEhClLEqBAFKWJUiIIUqVdkcgW3soqQyRV1+r4Y7SNSb5SWVzBt5SnyH/5BUwcb1s54DQtzaa3a0CrI9evXa6ybOnVqrQyZGsePHyc+Pp68vDzefvttXn311YZ2yaiRyRVMWxlP9oMyAHLyy7ibW4JH88a1akfrkO3m5oabmxtJSUk8fPiQF154gYqKClJSUuruuZGxd+9eevfuzZAhQ/D39+fQoUNAZeziwoULWbJkCYcPH25YJ40cmVxBwvVsHhSWq8ocGlnRwsWu1m1p7SGHDh0KwHfffUdERAQAAwYMYPz48bU2ZKykpKQwdepURo8ezZUrV3jvvfd46623VPUbN27k7bff1tqGUqnk008/JSUlBUtLSxYuXKiKxgaoqKggIiKCO3fuYG9vT2RkJK1atVJb3qJFC7XPGit/DtPlCICFVIJjE2vWzuhX6+EadHyHLCkpISkpiY4dO5KYmEh5eflff8lEuHnzJgMGDACgZcuWWFhYACAIAitWrKBv37507NhRaxvHjx+noqKCr7/+mqSkJJYsWcLGjRtV9fv27cPW1pZ9+/aRnp7OggUL2Lp1q9pyPz8/tc8aI2kZ+cxadwaZvDLG29LCjJljfOnWoVmdxAg6CjI6Opq1a9eSlpaGh4cHn332WZ2MGSM3b97E09MTQRDYtWuX6lD9zp07OXfuHA8fPuTWrVuMHj2ao0eP8tVXX1X7/qxZs0hMTKRPnz4AvPTSS1y9erXaM6mpqfTt2xeA1q1bk5aWprHc3d1d7bPGRGFxGV9+d5W4hLvVyh0aWT+RGEFHQbZt25a5c+dy584dWrdujaOjY50NGhNZWVmUlJQwadIksrOz8fLyIjQ0FICxY8cyduzYas8PHDiQgQMH1mjn66+/xt7eXvVZKpUil8tVpwbbt2/PyZMn8ff35/Lly2RnZ6NQKNSWe3t7q31WKq37P3J9cvteIVOWx9cotzCX1HmYfhSdBLl//35iYmIoLCxk0KBBFBYWMnv27FoZksvlzJw5k5ycHHx8fKqlGTl06BC7du2iUaNGLFmyhGbNmtXup6gjKSkpdOvWjR07dlBYWMjgwYO5dOkSXbt2Vfu8ph7S3t6ekpISVZlSqax2hHX48OGkpaUxduxYunbtSseOHZFKpWrLAwMDSU9Pr/GsMfDh4u/IuK9UW/ePMV2wtbZ8Yhs6LYx/++23bN++ncaNG/PBBx/wyy+/1NpQbGwsXl5e7N69m6KiIq5cuQJUvvDv2bOHvXv3MmXKFDZt2lTrtuvKzZs36dChA1B5RHTw4MGcOnVK4/MDBw5k586d1f74+PjQtWtXTp8+DUBSUpLqnHMVv/76K76+vuzcuRN/f3/VcVl15ZqebWjenBGjUYyNbCS83LFFvdjRSZBKpRKlUqk6/vnoQXldSUpK4uWXXwagV69eXLx4EYD09HTatWuHubk5vr6+Nd6/9ElKSgrt27dXfX799de1ClITb7zxBpaWlgQFBbF48WIiIiIoKChQrdV6eHiwZ88eRo0axZo1a1Sjg7pyTc82JG/OiNFa/8UnA594qK5CpyE7ODiYESNGkJOTw+jRoxk1alStDRUXF2NnV7kuZWNjoxriHi2XSCQolTX/F65bt07rIn1dWblyZbXP3bt3V61D1gYzMzPmz59fo7zKZ0dHR7Zv316jXlO5ujJDI5MruJtbwtQVJ7U+t+2f/vUyVFehkyA7dOjAtm3buH37Ni1btuTevXu1NmRnZ0dpaSlQmY6kUaNGNcoFQaj27lVFaGioarJRhRgxrj9kcgWhK06SmVui9blt//TH+bnaL35rQ+uQnZSUxMGDBwkNDSU+Pp709HTi4+NrPaEB6NSpExcuXADg/PnzqowKrVu35saNG8hkMhITE0WhGQG37hU1iBjhLwRpZ2dHZmYm5eXlZGRkkJGRQVZWVp0SYA4aNIjk5GRGjRqFVCqloqKCXbt2YWVlRVBQEGPGjGHFihW8//77df5hROqHj1ef1lq/IvRVvYgRdEylcvfuXa5fv46/vz+ff/45w4YNw9XVVS8O6YqXl9dTtaduLGhb2gGYMaYLr/lqzoH5pOg0y545c6Zq4bdr1651GrJFjJ81e/6rVYyujrb07qzfDME6L/u88sorALzyyitqZ8Iips2CzWc5/thWYLX691/m89mv19vyjiZ0mmU3bdqUjRs34uPjw5UrV3BwcNCrUyKGZcM3iVy4maexfkXoq3i1ctJYX5/o1EMuWbIEKysr4uLisLGxYenSpfr2S8RA/HT5DkfPZ2is3zDrNYOJEXTsIW1tbWnevDmCINC5c2fkcrm+/RIxAPuPX2P7kVS1dW3cbPjsHwEG9kjHHjIqKorr169z7NgxMjMzmTVrlr79EtEzx86naRRjt7aODSJG0FGQaWlpzJgxAysrKwYPHkxBQYGe3RLRF4XFZWw5cJn132iOGZjzXi8DelQdnYZspVJJZmYmEomEBw8eqKKqRUyLwuIygufFan3GQiqp0+Gs+kInQc6dO5cZM2aQnp7O5MmTmTt3rr79EtEDn399SWu9uRk0c7Kr0+Gs+kInQZ45c4a9e/fq2xcRPSGTK9j2w1V+vp6rtn5KYCf6dnEnN7+cFi52el9r1IZO75Dnzp1TReSImBYyuYJ3Fx7ju9O/q63/ZHw3BvZsg621JR7NGzeoGEHHHvLOnTv06NGDpk2bApVxiz/++KNeHROpH/bGJZP/UKa27qNRnXm5k3FcFlqFToL89NNPxcwNJsix82nsO67+1GIzByv6dTWO4xGPopMgP//8c1GQJoJMruDWvSJSfs9j08Frap+ZMaYLvTu7NfjwrA6dBCmXyxk5ciStW7dWnatZvHixXh0TqT0yuYIpy06Qlaf5fX/J5F50bONiQK9qh06CnDlzJoDaCx1FjIe0jAKtYtwS8TrNnRsZ0KPao5Mgvb292bhxI6mpqXh4eDB58mR9+yVSS2RyBbHn/qe2ztYStswNoIm9jYG9qj06RYxPmTIFf39/unbtSkJCAkeOHOGLL74whH8aESPG/6SwuIwPl/zIw7KaSUJnvP0SvX1aGuX7ojp06iGLiopUmdA8PDw4cOCAXp0S0Q2ZXEFaZgGz1p5VWx8+tiu9OxvfTFobOgnSzMyMCxcu8NJLL3Hx4kW1R1VFDEtVgtCs++pPB34wrL3JiRF03KmJjo5m165dDB06lK+++ooFCxbo2y+RvyDl1gOy7hejUNZ843J5zoqAl9s0gFdPjk6ClEqlvPnmm/zwww+0b98eS8v6y1QgUjtkcgVXU+8T8fnPKNQcbXpvSHs2R7xhMu+Mj6OTIGfMmKHKNCGeOmw4qobpiI0/qa2fEvgiQ/q1M1kxggFPHc6fP5/Ro0cze/bsGkcgYmJiGDlyJCNGjFBlEROpjkyu4GxSJhk5xWrrmzna0L+7h9o6U8Igpw4vX76sSru3YcMG4uLiGDRoEAAKhYLt27fzzTffUF5ezoQJE1QZZEUqkckVTF56nHsP1KfSnju+G77eribdM1ZRq1OHsbGxdTp1+HgqvsTExD8dMDNj69atqpl71dakyJ9cS8/VKMYFk3rySifj3JeuCzqfOpwwYYLOje7du7faWuXly5dVSeBtbGyqxVZKJBIcHR1RKBTMnTtXrR19peMzBQqLy/jnZvUJYl2aWNCxjeGOqBoCnXZqnpTt27fj7OzM4MGDuXjxIseOHVNdMwKVwRthYWF06NCBiRMn6tTms7BTU1pewfj5xyj9o+Y7+0dBPvTr8vxT0zNWYZC7DjWl4qti/vz5+Pj46CzGZ4H7BSWEzDuiVoxLJvfCv7vnUydG+Isesj6vlouKiiI5ORkXFxdWrVrFnTt3iImJITAwkEGDBtGlSxcA7O3tq93xoomnuYdM+T2PmevUbwcufP8VOrczzKUADYFWQR48eFDjF6v2thuKp1GQMrmCxOR7RG9PUFv/tIsRdLxaTkT/yOQKpi4/wd376uMZPxrV+akXI/yFIL29vXF2dlYlpYfKPOASiYRjx47p3blniZRbDzSKsbmTrVGef9EHWgW5aNEi4uPjsbCwoH///vj5+WFjY/xBnqZGyu95RHz+s9q6xZN74eXh+FROYNSh07JPUVERcXFxnDp1CmtrawICAvD39zeEfxp5Wt4hr6XlEq5BjPMm9qBb++YG9qhh0WnZp3Hjxrz55psMGTKEoqIi1q1bp2+/TBqZXMGtrCJk8poR3I8+k5CcpVGMzk2s6fxCU325aLRoHbJLS0uJj48nNjaWnJwc+vbtS3h4uFHf39zQVEXk5OaX4eJgw9oZr9UYbv/qdKCLgw3rZ9b83rOAVkH27NmTFi1aMGDAAPr164dEIiEpKYmkpKRql5yL/Mnd3BJy88v4Q6YgN7+sWiaxqtuxyv6QqRXjGP/WdH+xJR6uDZ/SpKHQKsiJEyeqgh0yMzMN4pCp08LFDhcHG1UPWZVJrKrnvJtTjLrLXZvYWRD4RodnVohVGGQvWx8Y86Smqid8NJPYrawijfcGmklgx6emcUxV32jtIQMCAqqFgykUCpo1a8b8+fNp08Y0z2wYAgtzaY2EnwUPy9Q++1ZfT0YP8K7XCyxNGa2CjI2tmW31xo0bREVFsWPHDr059TShbTtQaoYoxseo9XlWb29v8eIkHSktr+DDpSd4UPRHjTr/7m6895aPKMbHqLUgKyoqKC5Wf65D5M/3x+caWTBl2QkKS2peoeLcxJrJgV2e+QmMOrQK8vHwM5lMxrlz5xg+fLhenTJVqmbSOfllyOUK1ByZZphfa4IHirNpTWgVpJtb9eyqVlZWvPXWW3h6eurVKVOlag2yQqZ5h2bYa21FMWpBDD+rJ2RyBQ9Ly7G2lPCH+gzKWFqYUfBQJi7vaEFM0lMPVMYynuSuhjw7zk2seVgqq7ZQLqIeUZD1wK17RdzLUy/G5+wt2RDmZxRXbpgCdTrkdfDgQb799tv69sXoURfFI5MrWL4zQe0EBmD22G5Gc+WGKVCnHtLKyor27dtz//59nJ2d69sno+HRLUCgRhQPwNnLmZojvZ1t8fJwNJS7TwW1EqQgCKSnp/O3v/1NX/4YDY+Hkc0Y41stiudaei7Ld16kqLT6DGZIH0+6t3fF3s7imY7aqSs6CXL9+vW4uLhw7949EhMTcXd3Jzo6Wt++NSiPh5EhEXBxsCHnQSmNbC3UZpNo6mDDuMEdRRE+ATq9Q549e5ZRo0aRkJDAjh07yMjQfAO9JrRlP6tizJgx3Lx5s9Zt64OqMDIrCykuDjZ4uDZmyZReNLK1JK+wZp4dh0YWrHtGg2rrE516SEEQOHPmDJ6enuTl5ZGfn18rI9qyn1Vx5MgRbty4Uat29YmFuZS1M16r9g758erT5BWpT/o0Z/zL4r50PaBTD/nuu++yf/9+PvjgA7766iumTZtWKyPasp9B5f74oUOH6N+/f63a1TdVYWQW5lLSMgrILVAvRldHG9q4PWdY555SdOoh/fz86NKlC+bm5jg5OdG5c2etz9cm+xnAl19+yejRozl69Kja9hoy+5lMriAto4Af/3tbbf2UES/Sv5uHOFTXEzoJcsaMGfz9738nNjYWd3d35syZw7/+9S+NzwcFBREUFKT6vH37dpUIS0tLVemhAR48eMCVK1d47733NAoyNDSU0NDQamVeXl66uP5E/FU2Cecm1qIY6xmdhuz8/Hxef/11srKymDp1KmVl6qOfNaEt+1lCQgKZmZmEhIRw5swZIiIiUCg0BycYkltZDzWK0aGRBRvC/EQx1jM6CVIul/PZZ5/h6+vLuXPnqKioqJWRbt26IZVKCQoKIjk5mYCAANLT01m9ejUBAQEcOHCAnTt30qdPHxYvXoxU3SkoAyOTK1i284LaOpfnrNkU7i9OYvSAToe8bt26xU8//URgYCDHjx+nQ4cODX42+0kOeak7hPV4WdLNbP65+XyN7wb5t2XkG95iz6gndHqHdHV1pbS0lOXLl9O7d2+srKz07ZfeUHeQH/j/wNpSrC3NGda3FduP1FwPNZPAUL8XRDHqEZ2G7Dlz5mBvb8/Vq1eRSqWEhYXp2y+9oe4g/93cEnLyS6mQKSkqqVArxrEDvdizcJA4TOsZnQR5//59goKCsLCwoE+fPkYz6agLj+/AuDhYI5MrsdTS67k62vCW3wuiGA2ATkO2ubk5CQkJCIJASkqKSafkq9qBuZX1kPIKGaEr4nlQWIZczUFKCRA5sQedX2gqDtMGQidBzp8/n2XLlvHgwQPWr1/PvHnz9O2X3lnxVQKZueqDaquwsJDi8pwYVGtIdBKkm5sbkZGRlJeXPxUXG93NLSH7gfr1xSqkZpXRO+KRA8OikyAjIyNJTEykWbNmqpTO//73v/Xtm95o4WJHI1sL8h+qX09t5mBF+LiXxXjGBkAnQaakpPDDDz/o2xeDIZMreKmtCycv1czo9o/RXXn1pRaiEBsInQTp5eXFb7/9xgsvvKBvf/RO1v2HTFp8Qm1dC2dbUYwNjE6CvHbtGu+//77qs0Qi4ccff9SbU/qisLhMoxjFntE40EmQYWFhqnhGgFOnTunNIX3y/dn/qS13dbQWxWgkaBXk2bNnSUpK4uDBg9WyWBw+fJh+/frp3bn6oPLoauWa496436rVWUrhnxNfoWNrZ1GMRoJWQbZt25bc3FwsLS1p2bKlaoa9bNkyQ/n3RGiLZ5QA/44Us9YaG1oFWVBQwNChQ3Fzc6u2/ljbeMiG4ta9IrViNJPA1k/8RTEaIVoFeebMGby9vdVewtm9e3e9OVUfyOQKsnKri9G5iTUT3uyEb3sXcV/aSNEqyKrI7sWLFxvEmfqg6p1x+a4EcgvKkJpJEAQBpybWrJ/lJwrRyNEa7bNhwwbV38PDw/XuzJNSlUJ51rrT3L1fgkyuRKEUkEjA0kIqTlxMAK2CfDSY3NjvqZHJFYSuOEn2g1LkiupB8Aol3C8o5+5fBFOINDxah+xHJzLGHlRxK+shOfl/npuWAFJzCYJSQGpmJuZmNBG0CjI9PZ2IiIgafwcjfK+UCEiAqr5x0eSeNLK1xsXBWszNaEJoFeSqVatUfzf29M4ero1p4WJHTn4ZTR1s8PJwUgnQo7k4kTEVnqqr5dSdJhQxLeqUQbcuaMt+du3aNUaPHs3w4cP55ptv6mzj0Vw8IqaJQQT5aPaz559/nri4uGr1y5YtY/Xq1ezdu7fWmdVEni4MIkht2c9KS0uRyWSsWbOGCRMmVIsqEnn20MstDLXJflZUVMT169dZvnw5lpaWTJ48+YmGbRHTRi+CrE32s8aNG+Pu7q66NczMzAy5XI65+Z+uaUrHZ4gMaCL6QVMaHIPcU9OpUyf+85//MHjw4BrZz2xtbbGwsODevXvY29sjk8mqiRE0p+NrqAvcn1XbhrBvkHdIbdnPACIiIpg6dSrvvPMO06dPN4RLIkbKU7UOKdo2ffsGW4cUEdEFkxXk1KlTRdtPoX2THbJFnk5MtocUeToxKUEaYj+8Lrar0NdNZNpsx8TEMHLkSEaMGMHp06frzaZcLmf69OmMGTOGJUuWVKs7dOgQgYGBjB8/nuzs7HqzCSYkyIbcD/8r26C/m8i02VYoFGzfvp3du3ezbdu2er3LJzY2Fi8vL3bv3k1RURFXrlwBUPmyd+9epkyZwqZNm+rNJpiQIBtyP7whbyLTZtvMzIytW7eqNhLqM6r/cbsXL14EKgO127Vrh7m5Ob6+vly9erXebIKBdmrqQkPuh9f3TWT6si2RSHB0dEShUDB37lwmTJjwxParKC4uxs7OTmW3pKSkRrlEIkGpVJN6+AkwWkHW9364vmzrchOZvmxD5bteWFgYL774IgMGDHhi+1XY2dmptftouSAIdf4da8Jkhmxtt4E9uh9eXFysdj9cX7b1fROZNttQOeHx8fFh4sSJ9WZTm93WrVtz48YNZDIZiYmJ9R7gYlLrkFFRUSQnJ+Pi4sKqVau4c+cOMTExfPzxx/z3v/9l6dKlCILARx99RN++fQ1mu4rw8HAmTJhAu3btDGI7MDCQQYMG0aVLFwDs7e1Vw/uTUlFRQVhYGFlZWXh5eTF48GBu3rxJcHAwBw4cYM+ePUilUlauXKkameoDkxKkyNOPyQzZIs8GoiBFjApRkCJGhShIEaNCFKSIUSEKUsSoeGYE+csvv9C7d29CQkJUf86dO8eWLVtITk4mKyuLM2fONLSbDc6+fft0fvb111+vd/tGu3WoD/r06VMjlKpnz54AHDhwgMzMTPr06dMQrhkNmzZtYuTIkQ1m/5kSpDrCw8MZOnQoW7ZsoaKigm7durF8+XI6derEjRs38Pb2Zv78+WRkZBAZGYlMJsPFxYXo6Giys7MJDw9HKpXi4ODAqlWriI+P54svvkAikdC7d2+mTZum1u66detwc3Nj2LBhxMTEcPv2bYYOHcqsWbNwcnLi1q1bjB07lhEjRhAbG8vmzZtRKpUMGjSISZMmsWXLFlUoWkhICEOGDCEkJARHR0eg8gaNS5cuUVpaysaNG9m4cSNXr15FIpEQFhZG586da7QLkJuby7Jly5g0aRIRERGqYIro6GicnJxYvnw558+fx93dXT+XHwjPCOfPnxd69eolBAcHC8HBwcKsWbMEQRCE2bNnC+fPnxf2798vrF27VhAEQfDz8xOSk5MFpVIpDBw4UMjJyRFCQ0OFCxcuCIIgCLt37xa2bNki7Nq1S1i5cqUgl8uFuLg4ITs7WwgNDRXi4+MFpVIp7N69W1AqlWr9Wbt2rbB//35BEATh0KFDwtq1a4U7d+4IvXv3FsrLy4Xs7GxhwIABgkwmE/r37y8UFhYKCoVCWLx4sZCcnCwEBwcLCoVCKC8vFwYPHizk5eUJwcHBQnx8vKr9ZcuWCYIgCCdPnhTCw8MFQRCEvLw8ITAwUG27FRUVgp+fnyAIgrBkyRLh4MGDgiAIwqlTp4R58+YJKSkpwvjx4wWlUilkZ2cLPj4+9f7v9Ez1kOqGbE14eXkhkUho1qwZf/zxB6mpqaxZswaJRIJMJqNTp07MnDmTTZs2MW7cOFxdXenatSuzZ89m06ZNbNmyhW7duqFUKpFKtWdjEx7ZvW3VqhVWVlY0bdoUhUJBfn4+Tk5ONG7cGKjs0Q8fPkznzp0xMzPDysqKdu3acfv2baAy+KEKT09PAH777TcuXrxISEgIUBmul52dXaPdR0lNTSUhIYH9+/ejVCpxcHAgNTWVjh07IpFIaNq0Ke7u7jr9LmvDMyVIbUgkkmrCeDzY1dPTk9DQULy9vTl9+jRyuZwTJ07Qo0cPpk+fzrp16zhy5Aj3799n2rRpuLi4EBISQnp6utpLS62srMjLywMgOTkZe3t7tXadnJx48OABxcXF2NvbExoaSkhICJcvX0apVCKTyUhOTqZly5Y1vm9mZqbyvV+/fsyZM4eCggK+/PJLmjdvXqPdhQsXVvt5e/Togb+/P1euXCE1NRUPDw/27NmDUqmkoKBAL3nnRUH+P23btmXz5s01wruqCAsLIyoqirKyMqRSKUuXLlVFxGzYsAFra2sWLVrElStXmDBhAs899xwtW7akTZs2HD58mPLycoYNG6Zqb+DAgUyfPp2zZ8/SvHlzlSAfx8zMjNmzZ6uCbwMCAujRowe9e/cmKCgIuVzOO++8g7Ozs8afrX///pw7d47g4GCKiop499131bbbpEkTmjdvTlRUFKGhocyZM4dt27ZRUVFBVFQUHTp0oG/fvgQGBuLm5qZ6X61PxGgfA5Cenk5iYiIjRoxoaFeMHlGQBiAnJwdnZ2fVECqiGVGQIkaF+F9WxKgQBSliVIiCFDEqREGKGBWiIEWMiv8DggNCZfjPLWIAAAAASUVORK5CYII=\n",
      "text/plain": [
       "<Figure size 144x144 with 1 Axes>"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    }
   ],
   "source": [
    "with sns.axes_style('ticks'):\n",
    "    fig,ax = plt.subplots(1,1, figsize=(2,2))\n",
    "    ax.plot(np.mean(fitness, axis=1), np.mean(fitness_UMI, axis=1), '.')\n",
    "    ax.set(xlabel='Fitness, uncorrected', ylabel='Fitness, UMI corrected', xlim=-0.6, ylim=-0.6,\n",
    "          title='UMI correction')\n",
    "    ax.text(-0.5, 0, f\"$R^{2}$={np.round(r2,3)}\", fontsize=10)\n",
    "    sns.despine()\n",
    "    fig.savefig(repo +'/Analysis/Plots_for_paper/umi_correction.pdf', format='pdf', bbox_inches='tight')\n",
    "    plt.show()"
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3 (ipykernel)",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.10.4"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 5
}
